--- /home/rossierd/soo/infrabase/build/tmp/work/usr-so3-1.0-r0/usr-so3-1.0/lib/CMakeLists.txt	2025-09-06 14:22:14.115073356 +0200
+++ ./lib/CMakeLists.txt	2025-09-06 14:25:14.373460361 +0200
@@ -3,12 +3,7 @@
 string(REPLACE "-Werror" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
 
 add_subdirectory(libc)
-set(LV_BUILD_CONF_DIR
-    ${CMAKE_CURRENT_SOURCE_DIR}
-    CACHE PATH "Directory containing lv_conf.h")
-add_subdirectory(lvgl)
 
 set(CMAKE_C_FLAGS "${ORIGINAL_C_FLAGS}")
 
-# Non external libs
-add_subdirectory(slv)
+
