--- /home/david/Projects/ChAV_TRVI/chav_trvi/build/tmp/work/linux-toradex-5.4-2.3.x-imx/git/arch/arm64/boot/dts/freescale/overlays/apalis-imx8_resistive-touch_overlay.dts	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/overlays/apalis-imx8_resistive-touch_overlay.dts	2023-08-10 09:55:58.415675165 +0200
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright 2022 Toradex
+ */
+
+// Resistive touchscreen enablement.
+
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "toradex,apalis-imx8", "toradex,apalis-imx8-v1.1";
+};
+
+&adc1 {
+	/* remove the pinctrl-adc1 to use the touchscreen */
+	pinctrl-0 = <>;
+};
+
+&touchscreen {
+	status = "okay";
+};
