--- /home/david/Projects/ChAV_TRVI/chav_trvi/build/tmp/work/linux-toradex-5.4-2.3.x-imx/git/arch/arm64/boot/dts/freescale/overlays/colibri-imx8x_atmel-mxt-adapter_overlay.dtsi	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/overlays/colibri-imx8x_atmel-mxt-adapter_overlay.dtsi	2023-08-10 09:55:58.415675165 +0200
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright 2020-2021 Toradex
+ */
+
+// Atmel MXT capacitive touch controller for 7" and 10.1" displays
+// orderable at Toradex. Compatible with a Capacitive Touch Adapter.
+
+&atmel_mxt_ts {
+	pinctrl-0 = <&pinctrl_atmel_adap>;
+	interrupt-parent = <&lsio_gpio0>;
+	interrupts = <21 IRQ_TYPE_EDGE_FALLING>;	/* SODIMM 28 */
+	reset-gpios = <&lsio_gpio0 22 GPIO_ACTIVE_LOW>;	/* SODIMM 30 */
+	status = "okay";
+};
+
+&i2c1 {
+	status = "okay";
+};
+
+&lsio_pwm0 {
+	status = "disabled";
+};
+
+&lsio_pwm1 {
+	status = "disabled";
+};
\ No newline at end of file
