--- /home/david/Projects/ChAV_TRVI/chav_trvi/build/tmp/work/linux-toradex-5.4-2.3.x-imx/git/arch/arm64/boot/dts/freescale/overlays/colibri-imx8x_panel-res-touch-7inch_overlay.dts	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/overlays/colibri-imx8x_panel-res-touch-7inch_overlay.dts	2023-08-10 09:55:58.415675165 +0200
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright 2023 Toradex
+ */
+
+// Enable the parallel RGB interface on the Colibri iMX8X with a
+// connected resistive touch 7" display.
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/dts-v1/;
+/plugin/;
+
+/ {
+        compatible = "toradex,colibri-imx8x";
+};
+
+#include "colibri-imx8x_ad7879_overlay.dtsi"
+#include "colibri-imx8x_parallel-rgb_overlay.dtsi"
+#include "display-lt161010_overlay.dtsi"
