--- /home/david/Projects/ChAV_TRVI/chav_trvi/build/tmp/work/linux-toradex-5.4-2.3.x-imx/git/arch/arm64/boot/dts/freescale/overlays/display-lt170410_overlay.dtsi	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/overlays/display-lt170410_overlay.dtsi	2023-08-10 09:55:58.419675186 +0200
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright 2022 Toradex
+ */
+
+// LT170410 panel with a resolution of 1280x800 pixel.
+
+&panel_lvds {
+	status = "okay";
+
+	data-mapping = "vesa-24";
+	width-mm = <217>;
+	height-mm = <136>;
+
+	panel-timing {
+		clock-frequency = <68930000>;
+		hactive = <1280>;
+		vactive = <800>;
+		hback-porch = <64>;
+		hfront-porch = <64>;
+		vback-porch = <5>;
+		vfront-porch = <5>;
+		hsync-len = <40>;
+		vsync-len = <6>;
+
+		hsync-active = <0>;
+		vsync-active = <0>;
+		pixelclk-active = <0>;
+	};
+};
