--- /home/david/Projects/ChAV_TRVI/chav_trvi/build/tmp/work/linux-toradex-5.4-2.3.x-imx/git/arch/arm64/boot/dts/freescale/overlays/display-vga-640x480_overlay.dtsi	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/overlays/display-vga-640x480_overlay.dtsi	2023-08-10 09:55:58.419675186 +0200
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright 2022 Toradex
+ */
+
+// VGA Signal 640x480@60Hz Industry standard timing
+
+&panel_dpi {
+	compatible = "panel-dpi";
+	status = "okay";
+
+	/* for 0.3mm pixels */
+	width-mm = <192>;
+	height-mm = <144>;
+
+	panel-timing {
+		clock-frequency = <25175000>;
+		hactive = <640>;
+		vactive = <480>;
+		hsync-len = <96>;
+		hfront-porch = <16>;
+		hback-porch = <48>;
+		vsync-len = <2>;
+		vfront-porch = <10>;
+		vback-porch = <33>;
+		hsync-active = <0>;
+		vsync-active = <0>;
+		pixelclk-active = <0>;
+	 };
+};
