--- /home/david/Projects/ChAV_TRVI/chav_trvi/build/tmp/work/linux-toradex-5.4-2.3.x-imx/git/arch/arm64/boot/dts/freescale/overlays/verdin-imx8mm_dsi-to-lvds_panel-cap-touch-10inch-lvds_overlay.dts	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/overlays/verdin-imx8mm_dsi-to-lvds_panel-cap-touch-10inch-lvds_overlay.dts	2023-08-10 09:55:58.419675186 +0200
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright 2020-2022 Toradex
+ */
+
+// Verdin DSI to LVDS Adapter with plugged in
+// Capacitive Touch Display 10.1" LVDS orderable at Toradex.
+
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "toradex,verdin-imx8mm";
+};
+
+#include "verdin-imx8_mipi-dsi-to-sn65dsi84.dtsi"
+#include "display-lt170410_sn65dsi84_overlay.dtsi"
+
+&atmel_mxt_ts {
+	status = "okay";
+};
+
+&backlight {
+	status = "okay";
+};
+
+&gpu {
+	status = "okay";
+};
+
+/* Verdin I2C_1 */
+&i2c4 {
+	status = "okay";
+};
+
+&lcdif {
+	status = "okay";
+};
+
+&panel_lvds {
+	panel-timing {
+		/*
+		 * Take the minimum pixelclock as 71.1 MHz is
+		 * not working on NXP i.MX 8M Mini
+		 */
+		clock-frequency = <68900000>;
+	};
+};
