--- /home/david/Projects/ChAV_TRVI/chav_trvi/build/tmp/work/linux-toradex-5.4-2.3.x-imx/git/arch/arm64/boot/dts/freescale/overlays/verdin-imx8mp_mezzanine_panel-lvds-dual-channel-1080p_overlay.dts	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/overlays/verdin-imx8mp_mezzanine_panel-lvds-dual-channel-1080p_overlay.dts	2023-08-10 09:55:58.419675186 +0200
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright 2020-2021 Toradex
+ */
+
+// Verdin iMX8M Plus Mezzanine with LG LP156WF1 dual-channel LVDS Panel
+
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "toradex,verdin-imx8mp";
+};
+
+&backlight_mezzanine {
+	status = "okay";
+};
+
+&gpu_2d {
+	status = "okay";
+};
+
+&gpu_3d {
+	status = "okay";
+};
+
+&lcdif2 {
+	status = "okay";
+};
+
+&ldb {
+	fsl,dual-channel;
+	status = "okay";
+};
+
+&lvds_channel0 {
+	fsl,data-mapping = "jeida";
+};
+
+&ldb_phy {
+	status = "okay";
+};
+
+&mix_gpu_ml {
+	status = "okay";
+};
+
+&ml_vipsi {
+	status = "okay";
+};
+
+&panel_lvds_mez {
+	data-mapping = "jeida-24";
+	height-mm = <194>;
+	width-mm = <345>;
+	status = "okay";
+
+	panel-timing {
+		clock-frequency = <138500000>;
+		hactive = <1920>;
+		hback-porch = <40>;
+		hfront-porch = <24>;
+		hsync-len = <16>;
+		pixelclk-active = <0>;
+		vactive = <1080>;
+		vback-porch = <23>;
+		vfront-porch = <3>;
+		vsync-len = <5>;
+	};
+};
+
+&pwm2 {
+	status = "okay";
+};
