--- /home/rossierd/soo/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/soo/drivers/Kconfig	1970-01-01 01:00:00.000000000 +0100
+++ ./soo/drivers/Kconfig	2025-08-24 11:46:06.523590591 +0200
@@ -0,0 +1,38 @@
+#
+# SOO device configuration
+#
+
+
+menu "Smart Object Oriented subsystem" 
+
+config VDUMMY_BACKEND
+        bool "SOO (realtime and no-RT) vDummy support for debugging and testing"
+
+config VUART_BACKEND
+	bool "vuart UART Backend support for SOO Mobile Entity consoles"
+	select SERIAL_CORE_CONSOLE
+	
+config VLEDS_BACKEND
+        bool "vleds LEDS backend driver"
+        default N
+        
+config VSENSELED_BACKEND
+	depends on FB_RPISENSE || ARCH_VEXPRESS
+	bool "vsenseled Sense-HAT led backend driver"
+	default n
+	help
+	  The vsenseled driver enables 5 logical LEDs that may be used to represent the joystick movement
+	  for example.
+
+config VSENSEJ_BACKEND
+	depends on FB_RPISENSE || ARCH_VEXPRESS
+	bool "vsensej Sense-HAT joystick backend driver"
+	default n
+	help
+	  The vsensej driver captures joystick movements.
+
+endmenu
+	
+
+
+
