--- /var/dev/wayland_lvgl/build/tmp/work/uboot-2022.04-r0/git/arch/arm/cpu/armv7/lowlevel_init.S	2024-03-02 13:32:12.478624607 +0100
+++ ./arch/arm/cpu/armv7/lowlevel_init.S	2024-03-02 10:09:56.710891610 +0100
@@ -22,6 +22,25 @@
 
 .pushsection .text.lowlevel_init, "ax"
 WEAK(lowlevel_init)
+
+	/* heig-vd */
+
+	/*
+	 * Examine if we are running on CPU > #0
+	 * If yes, we suspend the CPU with wfi instruction.
+	 */
+
+	 mrc p15, 0, r0, c0, c0, 5
+	 ands r0, r0, #3
+
+	 beq cont
+
+loop:
+	 wfi
+	 b loop
+
+cont:
+
 	/*
 	 * Setup a temporary stack. Global data is not available yet.
 	 */
