--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm/include/asm/domain.h	2026-04-08 18:20:04.924888760 +0200
+++ ./arch/arm/include/asm/domain.h	2026-04-08 16:21:01.114526955 +0200
@@ -112,6 +112,19 @@
 }
 #endif
 
+#ifdef CONFIG_CPU_USE_DOMAINS
+#define modify_domain(dom,type)					\
+	do {							\
+		unsigned int domain = get_domain();		\
+		domain &= ~domain_mask(dom);			\
+		domain = domain | domain_val(dom, type);	\
+		set_domain(domain);				\
+	} while (0)
+
+#else
+static inline void modify_domain(unsigned dom, unsigned type)	{ }
+#endif
+
 /*
  * Generate the T (user) versions of the LDR/STR and related
  * instructions (inline assembly)
