--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm/include/asm/uaccess.h	2026-04-08 18:20:04.928880193 +0200
+++ ./arch/arm/include/asm/uaccess.h	2026-04-08 16:21:01.138567797 +0200
@@ -53,8 +53,32 @@
 extern int __get_user_bad(void);
 extern int __put_user_bad(void);
 
+/*
+ * Note that this is actually 0x1,0000,0000
+ */
+#define KERNEL_DS	0x00000000
+
 #ifdef CONFIG_MMU
 
+#define USER_DS		TASK_SIZE
+#define get_fs()	(current_thread_info()->addr_limit)
+
+static inline void set_fs(mm_segment_t fs)
+{
+	current_thread_info()->addr_limit = fs;
+
+	/*
+	 * Prevent a mispredicted conditional call to set_fs from forwarding
+	 * the wrong address limit to access_ok under speculation.
+	 */
+	dsb(nsh);
+	isb();
+
+	modify_domain(DOMAIN_KERNEL, fs ? DOMAIN_CLIENT : DOMAIN_MANAGER);
+}
+
+#define uaccess_kernel()	(get_fs() == KERNEL_DS)
+
 /*
  * This is a type: either unsigned long, if the argument fits into
  * that type, or otherwise unsigned long long.
@@ -82,7 +106,7 @@
 	"	subshs	%1, %1, %2\n"
 	"	movlo	%0, #0\n"
 	: "+r" (safe_ptr), "=&r" (tmp)
-	: "r" (size), "r" (TASK_SIZE)
+	: "r" (size), "r" (current_thread_info()->addr_limit)
 	: "cc");
 
 	csdb();
@@ -146,7 +170,7 @@
 
 #define __get_user_check(x, p)						\
 	({								\
-		unsigned long __limit = TASK_SIZE - 1; \
+		unsigned long __limit = current_thread_info()->addr_limit - 1; \
 		register typeof(*(p)) __user *__p asm("r0") = (p);	\
 		register __inttype(x) __r2 asm("r2");			\
 		register unsigned long __l asm("r1") = __limit;		\
@@ -199,7 +223,7 @@
 
 #define __put_user_check(__pu_val, __ptr, __err, __s)			\
 	({								\
-		unsigned long __limit = TASK_SIZE - 1; \
+		unsigned long __limit = current_thread_info()->addr_limit - 1; \
 		register typeof(__pu_val) __r2 asm("r2") = __pu_val;	\
 		register const void __user *__p asm("r0") = __ptr;	\
 		register unsigned long __l asm("r1") = __limit;		\
@@ -216,6 +240,18 @@
 
 #else /* CONFIG_MMU */
 
+/*
+ * uClinux has only one addr space, so has simplified address limits.
+ */
+#define USER_DS			KERNEL_DS
+
+#define uaccess_kernel()	(true)
+#define get_fs()		(KERNEL_DS)
+
+static inline void set_fs(mm_segment_t fs)
+{
+}
+
 #define get_user(x, p)	__get_user(x, p)
 #define __put_user_check __put_user_nocheck
 
@@ -223,6 +259,9 @@
 
 #include <asm-generic/access_ok.h>
 
+#define user_addr_max() \
+	(uaccess_kernel() ? ~0UL : get_fs())
+
 #ifdef CONFIG_CPU_SPECTRE
 /*
  * When mitigating Spectre variant 1, it is not worth fixing the non-
