--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm/kernel/process.c	2026-04-08 18:20:04.933880333 +0200
+++ ./arch/arm/kernel/process.c	2026-04-08 16:21:01.156225518 +0200
@@ -114,7 +114,7 @@
 	unsigned long flags;
 	char buf[64];
 #ifndef CONFIG_CPU_V7M
-	unsigned int domain;
+	unsigned int domain, fs;
 #ifdef CONFIG_CPU_SW_DOMAIN_PAN
 	/*
 	 * Get the domain register for the parent context. In user
@@ -123,11 +123,14 @@
 	 */
 	if (user_mode(regs)) {
 		domain = DACR_UACCESS_ENABLE;
+		fs = get_fs();
 	} else {
 		domain = to_svc_pt_regs(regs)->dacr;
+		fs = to_svc_pt_regs(regs)->addr_limit;
 	}
 #else
 	domain = get_domain();
+	fs = get_fs();
 #endif
 #endif
 
@@ -163,6 +166,8 @@
 		if ((domain & domain_mask(DOMAIN_USER)) ==
 		    domain_val(DOMAIN_USER, DOMAIN_NOACCESS))
 			segment = "none";
+		else if (fs == KERNEL_DS)
+			segment = "kernel";
 		else
 			segment = "user";
 
