--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm/mach-imx/Kconfig	2026-04-08 18:20:04.945966441 +0200
+++ ./arch/arm/mach-imx/Kconfig	2026-04-08 16:21:01.742375330 +0200
@@ -28,9 +28,29 @@
 	bool
 	select PM_GENERIC_DOMAINS if PM
 
+config HAVE_IMX_GPCV2
+	bool
+	select PM_GENERIC_DOMAINS if PM
+
 config HAVE_IMX_MMDC
 	bool
 
+config HAVE_IMX_AMP
+	bool
+
+config HAVE_IMX_DDRC
+       bool
+       select HAVE_IMX_BUSFREQ
+
+config HAVE_IMX_BUSFREQ
+       bool
+
+config HAVE_IMX_MU
+	bool
+
+config HAVE_IMX_RPMSG
+	bool
+
 config HAVE_IMX_SRC
 	def_bool y if SMP
 	select ARCH_HAS_RESET_CONTROLLER
@@ -168,7 +188,13 @@
 	select ARM_ERRATA_754322
 	select ARM_ERRATA_775420
 	select PINCTRL_IMX6SX
+	select HAVE_IMX_AMP
 	select SOC_IMX6
+	select HAVE_IMX_MU
+	select IMX_MBOX
+	select HAVE_IMX_RPMSG
+	select IMX_SEMA4
+	select KEYBOARD_SNVS_PWRKEY
 
 	help
 	  This enables support for Freescale i.MX6 SoloX processor.
@@ -204,6 +230,12 @@
 	select HAVE_IMX_MMDC
 	select HAVE_IMX_SRC
 	select IMX_GPCV2
+	select HAVE_IMX_DDRC
+	select HAVE_IMX_MU
+	select IMX_MBOX
+	select HAVE_IMX_RPMSG
+	select HAVE_IMX_GPCV2
+	select KEYBOARD_SNVS_PWRKEY
 
 config SOC_IMX7D_CM4
 	bool
