--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm/mach-imx/mach-imx6sx.c	2026-04-08 18:20:04.946066302 +0200
+++ ./arch/arm/mach-imx/mach-imx6sx.c	2026-04-08 16:21:01.750468455 +0200
@@ -9,6 +9,7 @@
 #include <linux/mfd/syscon.h>
 #include <linux/mfd/syscon/imx6q-iomuxc-gpr.h>
 #include <asm/mach/arch.h>
+#include <asm/mach/map.h>
 
 #include "common.h"
 #include "cpuidle.h"
@@ -30,6 +31,7 @@
 
 static inline void imx6sx_enet_init(void)
 {
+	imx6_enet_mac_init("fsl,imx6sx-fec", "fsl,imx6sx-ocotp");
 	imx6sx_enet_clk_sel();
 }
 
@@ -52,6 +54,13 @@
 	imx6_pm_ccm_init("fsl,imx6sx-ccm");
 }
 
+static void __init imx6sx_map_io(void)
+{
+	debug_ll_io_init();
+	imx6_pm_map_io();
+	imx_busfreq_map_io();
+}
+
 static void __init imx6sx_init_late(void)
 {
 	imx6sx_cpuidle_init();
@@ -68,6 +77,7 @@
 DT_MACHINE_START(IMX6SX, "Freescale i.MX6 SoloX (Device Tree)")
 	.l2c_aux_val 	= 0,
 	.l2c_aux_mask	= ~0,
+	.map_io		= imx6sx_map_io,
 	.init_irq	= imx6sx_init_irq,
 	.init_machine	= imx6sx_init_machine,
 	.dt_compat	= imx6sx_dt_compat,
