--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm/mach-imx/platsmp.c	2026-04-08 18:20:04.946880695 +0200
+++ ./arch/arm/mach-imx/platsmp.c	2026-04-08 16:21:01.754152181 +0200
@@ -18,7 +18,7 @@
 #include "hardware.h"
 
 u32 g_diag_reg;
-static void __iomem *scu_base;
+void __iomem *scu_base;
 
 static struct map_desc scu_io_desc __initdata = {
 	/* .virtual and .pfn are run-time assigned */
@@ -47,15 +47,39 @@
 	return 0;
 }
 
+#define MXC_ARCH_CA7           0xc07
+static unsigned long __mxc_arch_type;
+
+static inline bool arm_is_ca7(void)
+{
+       return __mxc_arch_type == MXC_ARCH_CA7;
+}
 /*
  * Initialise the CPU possible map early - this describes the CPUs
  * which may be present or become present in the system.
  */
 static void __init imx_smp_init_cpus(void)
 {
+	unsigned long arch_type;
 	int i, ncores;
 
-	ncores = scu_get_core_count(scu_base);
+	asm volatile(
+		".align 4\n"
+		"mrc p15, 0, %0, c0, c0, 0\n"
+		: "=r" (arch_type)
+	);
+	/* MIDR[15:4] defines ARCH type */
+	__mxc_arch_type = (arch_type >> 4) & 0xfff;
+
+	if (arm_is_ca7()) {
+		unsigned long val;
+
+		/* CA7 core number, [25:24] of CP15 L2CTLR */
+		asm volatile("mrc p15, 1, %0, c9, c0, 2" : "=r" (val));
+		ncores = ((val >> 24) & 0x3) + 1;
+	} else {
+		ncores = scu_get_core_count(scu_base);
+	}
 
 	for (i = ncores; i < NR_CPUS; i++)
 		set_cpu_possible(i, false);
@@ -63,11 +87,15 @@
 
 void imx_smp_prepare(void)
 {
+	if (arm_is_ca7())
+		return;
 	scu_enable(scu_base);
 }
 
 static void __init imx_smp_prepare_cpus(unsigned int max_cpus)
 {
+	if (arm_is_ca7())
+		return;
 	imx_smp_prepare();
 
 	/*
@@ -88,32 +116,6 @@
 	.smp_boot_secondary	= imx_boot_secondary,
 #ifdef CONFIG_HOTPLUG_CPU
 	.cpu_die		= imx_cpu_die,
-	.cpu_kill		= imx_cpu_kill,
-#endif
-};
-
-/*
- * Initialise the CPU possible map early - this describes the CPUs
- * which may be present or become present in the system.
- */
-static void __init imx7_smp_init_cpus(void)
-{
-	struct device_node *np;
-	int i, ncores = 0;
-
-	/* The iMX7D SCU does not report core count, get it from DT */
-	for_each_of_cpu_node(np)
-		ncores++;
-
-	for (i = ncores; i < NR_CPUS; i++)
-		set_cpu_possible(i, false);
-}
-
-const struct smp_operations imx7_smp_ops __initconst = {
-	.smp_init_cpus		= imx7_smp_init_cpus,
-	.smp_boot_secondary	= imx_boot_secondary,
-#ifdef CONFIG_HOTPLUG_CPU
-	.cpu_die		= imx_cpu_die,
 	.cpu_kill		= imx_cpu_kill,
 #endif
 };
