--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi	2026-04-08 18:20:05.013194400 +0200
+++ ./arch/arm64/boot/dts/freescale/fsl-ls1028a.dtsi	2026-04-08 16:21:01.973489295 +0200
@@ -199,7 +199,8 @@
 					trip = <&core_cluster_alert>;
 					cooling-device =
 						<&cpu0 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
-						<&cpu1 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+						<&cpu1 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+						<&gpu 0 1>;
 				};
 			};
 		};
@@ -255,6 +256,12 @@
 			};
 		};
 
+		serdes: phy@1ea0000 {
+			compatible = "fsl,ls1028a-serdes", "fsl,lynx-10g";
+			reg = <0x00 0x1ea0000 0x0 0xffff>;
+			#phy-cells = <1>;
+		};
+
 		scfg: syscon@1fc0000 {
 			compatible = "fsl,ls1028a-scfg", "syscon";
 			reg = <0x0 0x1fc0000 0x0 0x10000>;
@@ -622,6 +629,7 @@
 			snps,dis_rxdet_inp3_quirk;
 			snps,quirk-frame-length-adjustment = <0x20>;
 			snps,incr-burst-type-adjustment = <1>, <4>, <8>, <16>;
+			dma-coherent;
 			status = "disabled";
 		};
 
@@ -632,6 +640,7 @@
 			snps,dis_rxdet_inp3_quirk;
 			snps,quirk-frame-length-adjustment = <0x20>;
 			snps,incr-burst-type-adjustment = <1>, <4>, <8>, <16>;
+			dma-coherent;
 			status = "disabled";
 		};
 
@@ -872,12 +881,13 @@
 
 			port {
 				dpi0_out: endpoint {
-
+					remote-endpoint = <&dp1_out>;
 				};
 			};
 		};
 
-		gpu: gpu@f0c0000 {
+		/* Remove the upstream node to add node from SDK driver */
+/*		gpu: gpu@f0c0000 {
 			compatible = "vivante,gc";
 			reg = <0x0 0xf0c0000 0x0 0x10000>;
 			interrupts = <GIC_SPI 220 IRQ_TYPE_LEVEL_HIGH>;
@@ -887,6 +897,20 @@
 			clock-names = "core", "shader", "bus";
 			#cooling-cells = <2>;
 		};
+*/
+
+		gpu: gpu@f0c0000 {
+			compatible = "fsl,ls1028a-gpu";
+			reg = <0x0 0x0f0c0000 0x0 0x10000>,
+				<0x0 0x80000000 0x0 0x80000000>,
+				<0x0 0x0 0x0 0x3000000>;
+			reg-names = "base", "phys_baseaddr",
+				"contiguous_mem";
+			interrupts = <GIC_SPI 220 IRQ_TYPE_LEVEL_HIGH>;
+
+			throttle,max_state = <1>;
+			#cooling-cells = <2>;
+		};
 
 		sai1: audio-controller@f100000 {
 			#sound-dai-cells = <0>;
@@ -1021,6 +1045,22 @@
 			clocks = <&osc_27m>;
 		};
 
+		hdptx0: display@f200000 {
+			compatible = "cdn,ls1028a-dp";
+			reg = <0x0 0xf200000 0x0 0xfffff>;
+			interrupts = <0 221 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&clockgen 2 2>, <&clockgen 2 2>, <&clockgen 2 2>,
+				 <&clockgen 2 2>, <&clockgen 2 2>, <&dpclk>;
+			clock-names = "clk_core", "pclk", "sclk",
+				      "cclk", "clk_vif", "clk_pxl";
+
+			port {
+				dp1_out: endpoint {
+					remote-endpoint = <&dpi0_out>;
+				};
+			};
+		};
+
 		tmu: tmu@1f80000 {
 			compatible = "fsl,qoriq-tmu";
 			reg = <0x0 0x1f80000 0x0 0x10000>;
@@ -1336,5 +1376,4 @@
 			status = "disabled";
 		};
 	};
-
 };
