--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/boot/dts/freescale/fsl-ls1046a-qds-sd1-lane-d-sfp+.dtsi	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/fsl-ls1046a-qds-sd1-lane-d-sfp+.dtsi	2026-04-08 16:21:01.976200761 +0200
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2024 NXP
+ */
+
+&mac9 { /* DTSEC9/10GEC1 */
+	pcsphy-handle = <&pcsphy6>;
+	pcs-handle-names = "xfi";
+	managed = "in-band-status";
+	phys = <&serdes1 3>;
+	phy-names = "serdes";
+	sfp = <&sfp1>;
+
+	/delete-property/ phy-connection-type;
+	/delete-property/ phy-handle;
+};
