--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/boot/dts/freescale/fsl-ls1088a-rdb.dts	2026-04-08 18:20:05.013882562 +0200
+++ ./arch/arm64/boot/dts/freescale/fsl-ls1088a-rdb.dts	2026-04-08 16:21:01.979286238 +0200
@@ -15,12 +15,27 @@
 / {
 	model = "LS1088A RDB Board";
 	compatible = "fsl,ls1088a-rdb", "fsl,ls1088a";
+
+	sfp_cage: sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&sfp_i2c_bus>;
+	};
+};
+
+&dpmac1 {
+	sfp = <&sfp_cage>;
+	managed = "in-band-status";
+	phys = <&serdes1 2>, <&ti_retimer 0>, <&ti_retimer 1>;
+	phy-names = "serdes", "retimer_tx", "retimer_rx";
+	pcs-handle = <&pcs1>;
 };
 
 &dpmac2 {
 	phy-handle = <&mdio2_aquantia_phy>;
 	phy-connection-type = "10gbase-r";
 	pcs-handle = <&pcs2>;
+	phys = <&serdes1 3>;
+	phy-names = "serdes";
 };
 
 &dpmac3 {
@@ -28,6 +43,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs3_0>;
+	phys = <&serdes1 0>;
+	phy-names = "serdes";
 };
 
 &dpmac4 {
@@ -35,6 +52,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs3_1>;
+	phys = <&serdes1 0>;
+	phy-names = "serdes";
 };
 
 &dpmac5 {
@@ -42,6 +61,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs3_2>;
+	phys = <&serdes1 0>;
+	phy-names = "serdes";
 };
 
 &dpmac6 {
@@ -49,6 +70,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs3_3>;
+	phys = <&serdes1 0>;
+	phy-names = "serdes";
 };
 
 &dpmac7 {
@@ -56,6 +79,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs7_0>;
+	phys = <&serdes1 1>;
+	phy-names = "serdes";
 };
 
 &dpmac8 {
@@ -63,6 +88,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs7_1>;
+	phys = <&serdes1 1>;
+	phy-names = "serdes";
 };
 
 &dpmac9 {
@@ -70,6 +97,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs7_2>;
+	phys = <&serdes1 1>;
+	phy-names = "serdes";
 };
 
 &dpmac10 {
@@ -77,6 +106,8 @@
 	phy-connection-type = "qsgmii";
 	managed = "in-band-status";
 	pcs-handle = <&pcs7_3>;
+	phys = <&serdes1 1>;
+	phy-names = "serdes";
 };
 
 &emdio1 {
@@ -170,6 +201,31 @@
 				/* IRQ_RTC_B -> IRQ0_B(CPLD) -> IRQ00(CPU), active low */
 				interrupts-extended = <&extirq 0 IRQ_TYPE_LEVEL_LOW>;
 			};
+
+			rtc@53 {
+				compatible = "nxp,pcf2131";
+				reg = <0x53>;
+				/* IRQ_RTC_B -> IRQ0_B(CPLD) -> IRQ00(CPU), active low */
+				interrupts-extended = <&extirq 0 IRQ_TYPE_LEVEL_LOW>;
+			};
+		};
+
+		i2c@5 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x5>;
+
+			ti_retimer: retimer@18 {
+				compatible = "ti,ds125df111";
+				reg = <0x18>;
+				#phy-cells = <1>;
+			};
+		};
+
+		sfp_i2c_bus: i2c@6 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <6>;
 		};
 	};
 };
@@ -203,6 +259,10 @@
 	status = "okay";
 };
 
+&pcs_mdio1 {
+	status = "okay";
+};
+
 &pcs_mdio2 {
 	status = "okay";
 };
@@ -223,7 +283,7 @@
 		#address-cells = <1>;
 		#size-cells = <1>;
 		spi-max-frequency = <50000000>;
-		spi-rx-bus-width = <4>;
+		spi-rx-bus-width = <1>;
 		spi-tx-bus-width = <1>;
 		reg = <0>;
 	};
@@ -233,7 +293,7 @@
 		#address-cells = <1>;
 		#size-cells = <1>;
 		spi-max-frequency = <50000000>;
-		spi-rx-bus-width = <4>;
+		spi-rx-bus-width = <1>;
 		spi-tx-bus-width = <1>;
 		reg = <1>;
 	};
