--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/boot/dts/freescale/fsl-ls208xa.dtsi	2026-04-08 18:20:05.013882562 +0200
+++ ./arch/arm64/boot/dts/freescale/fsl-ls208xa.dtsi	2026-04-08 16:21:01.983629572 +0200
@@ -12,6 +12,7 @@
 #include <dt-bindings/clock/fsl,qoriq-clockgen.h>
 #include <dt-bindings/thermal/thermal.h>
 #include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/gpio/gpio.h>
 
 / {
 	compatible = "fsl,ls2080a";
@@ -273,7 +274,9 @@
 
 		dcfg: dcfg@1e00000 {
 			compatible = "fsl,ls2080a-dcfg", "syscon";
-			reg = <0x0 0x1e00000 0x0 0x10000>;
+			reg = <0x0 0x1e00000 0x0 0x10000>,
+			      <0x7 0x00100000 0x0 0x10000>;
+			reg-names = "dcfg_ccsr", "dcfg_dcsr";
 			little-endian;
 		};
 
@@ -285,6 +288,18 @@
 			clock-names = "sfp";
 		};
 
+		serdes1: phy@1ea0000 {
+			compatible = "fsl,ls2088a-serdes1", "fsl,lynx-10g";
+			reg = <0x00 0x1ea0000 0x0 0xffff>;
+			#phy-cells = <1>;
+		};
+
+		serdes2: phy@1eb0000 {
+			compatible = "fsl,ls2088a-serdes2", "fsl,lynx-10g";
+			reg = <0x00 0x1eb0000 0x0 0xffff>;
+			#phy-cells = <1>;
+		};
+
 		isc: syscon@1f70000 {
 			compatible = "fsl,ls2080a-isc", "syscon";
 			reg = <0x0 0x1f70000 0x0 0x10000>;
@@ -397,9 +412,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc000000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -407,9 +422,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc010000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -417,9 +432,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc100000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -427,9 +442,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc110000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -437,9 +452,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc200000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -447,9 +462,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc210000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -457,9 +472,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc300000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -467,9 +482,9 @@
 			compatible = "arm,sp805", "arm,primecell";
 			reg = <0x0 0xc310000 0x0 0x1000>;
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>,
+					    QORIQ_CLK_PLL_DIV(32)>,
 				 <&clockgen QORIQ_CLK_PLATFORM_PLL
-					    QORIQ_CLK_PLL_DIV(4)>;
+					    QORIQ_CLK_PLL_DIV(32)>;
 			clock-names = "wdog_clk", "apb_pclk";
 		};
 
@@ -1003,7 +1018,7 @@
 
 		i2c0: i2c@2000000 {
 			status = "disabled";
-			compatible = "fsl,vf610-i2c";
+			compatible = "fsl,ls208xa-i2c", "fsl,vf610-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0x0 0x2000000 0x0 0x10000>;
@@ -1015,7 +1030,7 @@
 
 		i2c1: i2c@2010000 {
 			status = "disabled";
-			compatible = "fsl,vf610-i2c";
+			compatible = "fsl,ls208xa-i2c", "fsl,vf610-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0x0 0x2010000 0x0 0x10000>;
@@ -1023,11 +1038,12 @@
 			clock-names = "i2c";
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
 					    QORIQ_CLK_PLL_DIV(4)>;
+			scl-gpios = <&gpio2 12 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
 		};
 
 		i2c2: i2c@2020000 {
 			status = "disabled";
-			compatible = "fsl,vf610-i2c";
+			compatible = "fsl,ls208xa-i2c", "fsl,vf610-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0x0 0x2020000 0x0 0x10000>;
@@ -1035,11 +1051,12 @@
 			clock-names = "i2c";
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
 					    QORIQ_CLK_PLL_DIV(4)>;
+			scl-gpios = <&gpio3 0 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
 		};
 
 		i2c3: i2c@2030000 {
 			status = "disabled";
-			compatible = "fsl,vf610-i2c";
+			compatible = "fsl,ls208xa-i2c", "fsl,vf610-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0x0 0x2030000 0x0 0x10000>;
@@ -1047,6 +1064,7 @@
 			clock-names = "i2c";
 			clocks = <&clockgen QORIQ_CLK_PLATFORM_PLL
 					    QORIQ_CLK_PLL_DIV(4)>;
+			scl-gpios = <&gpio3 2 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
 		};
 
 		ifc: memory-controller@2240000 {
@@ -1081,8 +1099,8 @@
 		pcie1: pcie@3400000 {
 			compatible = "fsl,ls2080a-pcie", "fsl,ls2085a-pcie";
 			reg-names = "regs", "config";
-			interrupts = <0 108 0x4>; /* Level high type */
-			interrupt-names = "intr";
+			interrupts = <0 108 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "aer";
 			#address-cells = <3>;
 			#size-cells = <2>;
 			device_type = "pci";
@@ -1103,8 +1121,8 @@
 		pcie2: pcie@3500000 {
 			compatible = "fsl,ls2080a-pcie", "fsl,ls2085a-pcie";
 			reg-names = "regs", "config";
-			interrupts = <0 113 0x4>; /* Level high type */
-			interrupt-names = "intr";
+			interrupts = <0 113 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "aer";
 			#address-cells = <3>;
 			#size-cells = <2>;
 			device_type = "pci";
@@ -1125,8 +1143,8 @@
 		pcie3: pcie@3600000 {
 			compatible = "fsl,ls2080a-pcie", "fsl,ls2085a-pcie";
 			reg-names = "regs", "config";
-			interrupts = <0 118 0x4>; /* Level high type */
-			interrupt-names = "intr";
+			interrupts = <0 118 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "aer";
 			#address-cells = <3>;
 			#size-cells = <2>;
 			device_type = "pci";
@@ -1147,8 +1165,8 @@
 		pcie4: pcie@3700000 {
 			compatible = "fsl,ls2080a-pcie", "fsl,ls2085a-pcie";
 			reg-names = "regs", "config";
-			interrupts = <0 123 0x4>; /* Level high type */
-			interrupt-names = "intr";
+			interrupts = <0 123 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "aer";
 			#address-cells = <3>;
 			#size-cells = <2>;
 			device_type = "pci";
@@ -1186,34 +1204,28 @@
 			dma-coherent;
 		};
 
-		bus: bus {
-			#address-cells = <2>;
-			#size-cells = <2>;
-			compatible = "simple-bus";
-			ranges;
-			dma-ranges = <0x0 0x0 0x0 0x0 0x100 0x00000000>;
-
-			usb0: usb@3100000 {
-				compatible = "snps,dwc3";
-				reg = <0x0 0x3100000 0x0 0x10000>;
-				interrupts = <0 80 0x4>; /* Level high type */
-				dr_mode = "host";
-				snps,quirk-frame-length-adjustment = <0x20>;
-				snps,dis_rxdet_inp3_quirk;
-				snps,incr-burst-type-adjustment = <1>, <4>, <8>, <16>;
-				status = "disabled";
-			};
-
-			usb1: usb@3110000 {
-				compatible = "snps,dwc3";
-				reg = <0x0 0x3110000 0x0 0x10000>;
-				interrupts = <0 81 0x4>; /* Level high type */
-				dr_mode = "host";
-				snps,quirk-frame-length-adjustment = <0x20>;
-				snps,dis_rxdet_inp3_quirk;
-				snps,incr-burst-type-adjustment = <1>, <4>, <8>, <16>;
-				status = "disabled";
-			};
+		usb0: usb@3100000 {
+			status = "disabled";
+			compatible = "fsl,ls2088a-dwc3", "snps,dwc3";
+			reg = <0x0 0x3100000 0x0 0x10000>;
+			interrupts = <0 80 0x4>; /* Level high type */
+			dr_mode = "host";
+			snps,quirk-frame-length-adjustment = <0x20>;
+			snps,dis_rxdet_inp3_quirk;
+			snps,incr-burst-type-adjustment = <1>, <4>, <8>, <16>;
+			snps,host-vbus-glitches;
+		};
+
+		usb1: usb@3110000 {
+			status = "disabled";
+			compatible = "fsl,ls2088a-dwc3", "snps,dwc3";
+			reg = <0x0 0x3110000 0x0 0x10000>;
+			interrupts = <0 81 0x4>; /* Level high type */
+			dr_mode = "host";
+			snps,quirk-frame-length-adjustment = <0x20>;
+			snps,dis_rxdet_inp3_quirk;
+			snps,incr-burst-type-adjustment = <1>, <4>, <8>, <16>;
+			snps,host-vbus-glitches;
 		};
 
 		ccn@4000000 {
