--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/boot/dts/freescale/imx8dxl-ss-conn.dtsi	2026-04-08 18:20:05.014882590 +0200
+++ ./arch/arm64/boot/dts/freescale/imx8dxl-ss-conn.dtsi	2026-04-08 16:21:02.006616280 +0200
@@ -5,15 +5,20 @@
 
 /delete-node/ &enet1_lpcg;
 /delete-node/ &fec2;
+/delete-node/ &usbotg3;
+/delete-node/ &usb3_lpcg;
+/delete-node/ &usb3_phy;
 
-&conn_subsys {
+/ {
 	conn_enet0_root_clk: clock-conn-enet0-root {
 		compatible = "fixed-clock";
 		#clock-cells = <0>;
 		clock-frequency = <250000000>;
 		clock-output-names = "conn_enet0_root_clk";
 	};
+};
 
+&conn_subsys {
 	eqos: ethernet@5b050000 {
 		compatible = "nxp,imx8dxl-dwmac-eqos", "snps,dwmac-5.10a";
 		reg = <0x5b050000 0x10000>;
@@ -21,15 +26,16 @@
 		interrupts = <GIC_SPI 163 IRQ_TYPE_LEVEL_HIGH>,
 			     <GIC_SPI 162 IRQ_TYPE_LEVEL_HIGH>;
 		interrupt-names = "eth_wake_irq", "macirq";
-		clocks = <&eqos_lpcg IMX_LPCG_CLK_4>,
-			 <&eqos_lpcg IMX_LPCG_CLK_6>,
-			 <&eqos_lpcg IMX_LPCG_CLK_0>,
-			 <&eqos_lpcg IMX_LPCG_CLK_5>,
-			 <&eqos_lpcg IMX_LPCG_CLK_2>;
+		clocks = <&eqos_lpcg 2>,
+			 <&eqos_lpcg 4>,
+			 <&eqos_lpcg 0>,
+			 <&eqos_lpcg 3>,
+			 <&eqos_lpcg 1>;
 		clock-names = "stmmaceth", "pclk", "ptp_ref", "tx", "mem";
 		assigned-clocks = <&clk IMX_SC_R_ENET_1 IMX_SC_PM_CLK_PER>;
 		assigned-clock-rates = <125000000>;
 		power-domains = <&pd IMX_SC_R_ENET_1>;
+		snps,clk-csr = <0>;
 		status = "disabled";
 	};
 
@@ -62,14 +68,14 @@
 
 	usbmisc2: usbmisc@5b0e0200 {
 		#index-cells = <1>;
-		compatible = "fsl,imx7ulp-usbmisc", "fsl,imx7d-usbmisc", "fsl,imx6q-usbmisc";
+		compatible = "fsl,imx8dxl-usbmisc", "fsl,imx7ulp-usbmisc", "fsl,imx7d-usbmisc", "fsl,imx6q-usbmisc";
 		reg = <0x5b0e0200 0x200>;
 	};
 
 	usbphy2: usbphy@5b110000 {
 		compatible = "fsl,imx8dxl-usbphy", "fsl,imx7ulp-usbphy";
 		reg = <0x5b110000 0x1000>;
-		clocks = <&usb2_2_lpcg IMX_LPCG_CLK_7>;
+		clocks = <&usb2_2_lpcg 0>;
 		power-domains = <&pd IMX_SC_R_USB_1_PHY>;
 		status = "disabled";
 	};
@@ -83,9 +89,7 @@
 			 <&conn_axi_clk>,
 			 <&clk IMX_SC_R_ENET_1 IMX_SC_PM_CLK_PER>,
 			 <&conn_ipg_clk>;
-		clock-indices = <IMX_LPCG_CLK_0>, <IMX_LPCG_CLK_2>,
-				<IMX_LPCG_CLK_4>, <IMX_LPCG_CLK_5>,
-				<IMX_LPCG_CLK_6>;
+		bit-offset = <0 8 16 20 24>;
 		clock-output-names = "eqos_ptp",
 				     "eqos_mem_clk",
 				     "eqos_aclk",
@@ -98,7 +102,7 @@
 		compatible = "fsl,imx8qxp-lpcg";
 		reg = <0x5b280000 0x10000>;
 		#clock-cells = <1>;
-		clock-indices = <IMX_LPCG_CLK_7>;
+		bit-offset = <28>;
 		clocks = <&conn_ipg_clk>;
 		clock-output-names = "usboh3_2_phy_ipg_clk";
 		power-domains = <&pd IMX_SC_R_USB_1_PHY>;
@@ -106,6 +110,14 @@
 
 };
 
+&dma_apbh {
+	compatible = "fsl,imx8dxl-dma-apbh", "fsl,imx28-dma-apbh";
+	interrupts = <GIC_SPI 176 IRQ_TYPE_LEVEL_HIGH>,
+		     <GIC_SPI 176 IRQ_TYPE_LEVEL_HIGH>,
+		     <GIC_SPI 176 IRQ_TYPE_LEVEL_HIGH>,
+		     <GIC_SPI 176 IRQ_TYPE_LEVEL_HIGH>;
+};
+
 &enet0_lpcg {
 	clocks = <&conn_enet0_root_clk>,
 		 <&conn_enet0_root_clk>,
@@ -116,7 +128,7 @@
 };
 
 &fec1 {
-	compatible = "fsl,imx8qm-fec";
+	compatible = "fsl,imx8dxl-fec", "fsl,imx8qm-fec";
 	interrupts = <GIC_SPI 160 IRQ_TYPE_LEVEL_HIGH>,
 		     <GIC_SPI 158 IRQ_TYPE_LEVEL_HIGH>,
 		     <GIC_SPI 159 IRQ_TYPE_LEVEL_HIGH>,
@@ -125,6 +137,11 @@
 	assigned-clock-rates = <125000000>;
 };
 
+&gpmi {
+	compatible = "fsl,imx8dxl-gpmi-nand", "fsl,imx8qxp-gpmi-nand";
+	interrupts = <GIC_SPI 174 IRQ_TYPE_LEVEL_HIGH>;
+};
+
 &usdhc1 {
 	compatible = "fsl,imx8dxl-usdhc", "fsl,imx8qxp-usdhc";
 	interrupts = <GIC_SPI 138 IRQ_TYPE_LEVEL_HIGH>;
