--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/boot/dts/freescale/imx8mq.dtsi	2026-04-08 18:20:05.022882813 +0200
+++ ./arch/arm64/boot/dts/freescale/imx8mq.dtsi	2026-04-08 16:21:02.050273135 +0200
@@ -11,7 +11,6 @@
 #include "dt-bindings/input/input.h"
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/thermal/thermal.h>
-#include <dt-bindings/interconnect/imx8mq.h>
 #include "imx8mq-pinfunc.h"
 
 / {
@@ -21,6 +20,8 @@
 	#size-cells = <2>;
 
 	aliases {
+		csi0 = &mipi_csi_1;
+		csi1 = &mipi_csi_2;
 		ethernet0 = &fec1;
 		gpio0 = &gpio1;
 		gpio1 = &gpio2;
@@ -98,6 +99,15 @@
 		clock-output-names = "clk_ext4";
 	};
 
+	irq_sec_vio: caam_secvio {
+		compatible = "fsl,imx6q-caam-secvio";
+		interrupts = <GIC_SPI 20 IRQ_TYPE_LEVEL_HIGH>;
+		jtag-tamper = "disabled";
+		watchdog-tamper = "enabled";
+		internal-boot-tamper = "enabled";
+		external-pin-tamper = "disabled";
+	};
+
 	cpus {
 		#address-cells = <1>;
 		#size-cells = <0>;
@@ -184,6 +194,21 @@
 			cache-line-size = <64>;
 			cache-sets = <1024>;
 		};
+
+		idle-states {
+			entry-method = "psci";
+
+			CPU_SLEEP: cpu-sleep {
+				compatible = "arm,idle-state";
+				arm,psci-suspend-param = <0x0010033>;
+				local-timer-stop;
+				entry-latency-us = <1000>;
+				exit-latency-us = <700>;
+				min-residency-us = <2700>;
+				wakeup-latency-us = <1500>;
+			};
+		};
+
 	};
 
 	a53_opp_table: opp-table {
@@ -263,49 +288,8 @@
 						<&A53_0 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
 						<&A53_1 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
 						<&A53_2 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
-						<&A53_3 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
-				};
-			};
-		};
-
-		gpu-thermal {
-			polling-delay-passive = <250>;
-			polling-delay = <2000>;
-			thermal-sensors = <&tmu 1>;
-
-			trips {
-				gpu_alert: gpu-alert {
-					temperature = <80000>;
-					hysteresis = <2000>;
-					type = "passive";
-				};
-
-				gpu-crit {
-					temperature = <90000>;
-					hysteresis = <2000>;
-					type = "critical";
-				};
-			};
-
-			cooling-maps {
-				map0 {
-					trip = <&gpu_alert>;
-					cooling-device =
-						<&gpu THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
-				};
-			};
-		};
-
-		vpu-thermal {
-			polling-delay-passive = <250>;
-			polling-delay = <2000>;
-			thermal-sensors = <&tmu 2>;
-
-			trips {
-				vpu-crit {
-					temperature = <90000>;
-					hysteresis = <2000>;
-					type = "critical";
+						<&A53_3 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+						<&gpu3d 0 1>;
 				};
 			};
 		};
@@ -321,8 +305,27 @@
 		arm,no-tick-in-suspend;
 	};
 
+	busfreq { /* BUSFREQ */
+		compatible = "fsl,imx_busfreq";
+		clocks = <&clk IMX8MQ_DRAM_PLL_OUT>, <&clk IMX8MQ_CLK_DRAM_ALT>,
+			 <&clk IMX8MQ_CLK_DRAM_APB>, <&clk IMX8MQ_CLK_DRAM_APB>,
+			 <&clk IMX8MQ_CLK_DRAM_CORE>, <&clk IMX8MQ_CLK_DRAM_ALT_ROOT>,
+			 <&clk IMX8MQ_SYS1_PLL_40M>, <&clk IMX8MQ_SYS1_PLL_400M>,
+			 <&clk IMX8MQ_SYS1_PLL_100M>, <&clk IMX8MQ_SYS1_PLL_800M>,
+			 <&clk IMX8MQ_CLK_NOC>, <&clk IMX8MQ_CLK_MAIN_AXI>,
+			 <&clk IMX8MQ_CLK_AHB>, <&clk IMX8MQ_CLK_25M>,
+			 <&clk IMX8MQ_SYS2_PLL_333M>, <&clk IMX8MQ_SYS1_PLL_133M>;
+		clock-names = "dram_pll", "dram_alt_src", "dram_apb_src", "dram_apb_pre_div",
+			      "dram_core", "dram_alt_root", "sys1_pll_40m", "sys1_pll_400m",
+			      "sys1_pll_100m", "sys1_pll_800m", "noc_div", "main_axi_src",
+			      "ahb_div", "osc_25m", "sys2_pll_333m", "sys1_pll_133m";
+		interrupts = <GIC_SPI 102 IRQ_TYPE_LEVEL_HIGH>, <GIC_SPI 109 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI 110 IRQ_TYPE_LEVEL_HIGH>, <GIC_SPI 111 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-name = "irq_busfreq_0", "irq_busfreq_1", "irq_busfreq_2", "irq_busfreq_3";
+	};
+
 	soc: soc@0 {
-		compatible = "fsl,imx8mq-soc", "simple-bus";
+		compatible = "simple-bus";
 		#address-cells = <1>;
 		#size-cells = <1>;
 		ranges = <0x0 0x0 0x0 0x3e000000>;
@@ -330,196 +333,9 @@
 		nvmem-cells = <&imx8mq_uid>;
 		nvmem-cell-names = "soc_unique_id";
 
-		etm0: etm@28440000 {
-			compatible = "arm,coresight-etm4x", "arm,primecell";
-			reg = <0x28440000 0x1000>;
-			cpu = <&A53_0>;
-			clocks = <&clk IMX8MQ_CLK_MAIN_AXI>;
-			clock-names = "apb_pclk";
-
-			out-ports {
-				port {
-					etm0_out_port: endpoint {
-						remote-endpoint = <&ca_funnel_in_port0>;
-					};
-				};
-			};
-		};
-
-		etm1: etm@28540000 {
-			compatible = "arm,coresight-etm4x", "arm,primecell";
-			reg = <0x28540000 0x1000>;
-			cpu = <&A53_1>;
-			clocks = <&clk IMX8MQ_CLK_MAIN_AXI>;
-			clock-names = "apb_pclk";
-
-			out-ports {
-				port {
-					etm1_out_port: endpoint {
-						remote-endpoint = <&ca_funnel_in_port1>;
-					};
-				};
-			};
-		};
-
-		etm2: etm@28640000 {
-			compatible = "arm,coresight-etm4x", "arm,primecell";
-			reg = <0x28640000 0x1000>;
-			cpu = <&A53_2>;
-			clocks = <&clk IMX8MQ_CLK_MAIN_AXI>;
-			clock-names = "apb_pclk";
-
-			out-ports {
-				port {
-					etm2_out_port: endpoint {
-						remote-endpoint = <&ca_funnel_in_port2>;
-					};
-				};
-			};
-		};
-
-		etm3: etm@28740000 {
-			compatible = "arm,coresight-etm4x", "arm,primecell";
-			reg = <0x28740000 0x1000>;
-			cpu = <&A53_3>;
-			clocks = <&clk IMX8MQ_CLK_MAIN_AXI>;
-			clock-names = "apb_pclk";
-
-			out-ports {
-				port {
-					etm3_out_port: endpoint {
-						remote-endpoint = <&ca_funnel_in_port3>;
-					};
-				};
-			};
-		};
-
-		funnel {
-			/*
-			 * non-configurable funnel don't show up on the AMBA
-			 * bus.  As such no need to add "arm,primecell".
-			 */
-			compatible = "arm,coresight-static-funnel";
-
-			in-ports {
-				#address-cells = <1>;
-				#size-cells = <0>;
-
-				port@0 {
-					reg = <0>;
-
-					ca_funnel_in_port0: endpoint {
-						remote-endpoint = <&etm0_out_port>;
-					};
-				};
-
-				port@1 {
-					reg = <1>;
-
-					ca_funnel_in_port1: endpoint {
-						remote-endpoint = <&etm1_out_port>;
-					};
-				};
-
-				port@2 {
-					reg = <2>;
-
-					ca_funnel_in_port2: endpoint {
-						remote-endpoint = <&etm2_out_port>;
-					};
-				};
-
-				port@3 {
-					reg = <3>;
-
-					ca_funnel_in_port3: endpoint {
-						remote-endpoint = <&etm3_out_port>;
-					};
-				};
-			};
-
-			out-ports {
-				port {
-					ca_funnel_out_port0: endpoint {
-						remote-endpoint = <&hugo_funnel_in_port0>;
-					};
-				};
-			};
-		};
-
-		funnel@28c03000 {
-			compatible = "arm,coresight-dynamic-funnel", "arm,primecell";
-			reg = <0x28c03000 0x1000>;
-			clocks = <&clk IMX8MQ_CLK_MAIN_AXI>;
-			clock-names = "apb_pclk";
-
-			in-ports {
-				#address-cells = <1>;
-				#size-cells = <0>;
-
-				port@0 {
-					reg = <0>;
-
-					hugo_funnel_in_port0: endpoint {
-						remote-endpoint = <&ca_funnel_out_port0>;
-					};
-				};
-
-				port@1 {
-					reg = <1>;
-
-					hugo_funnel_in_port1: endpoint {
-					/* M4 input */
-					};
-				};
-				/* the other input ports are not connect to anything */
-			};
-
-			out-ports {
-				port {
-					hugo_funnel_out_port0: endpoint {
-						remote-endpoint = <&etf_in_port>;
-					};
-				};
-			};
-		};
-
-		etf@28c04000 {
-			compatible = "arm,coresight-tmc", "arm,primecell";
-			reg = <0x28c04000 0x1000>;
-			clocks = <&clk IMX8MQ_CLK_MAIN_AXI>;
-			clock-names = "apb_pclk";
-
-			in-ports {
-				port {
-					etf_in_port: endpoint {
-						remote-endpoint = <&hugo_funnel_out_port0>;
-					};
-				};
-			};
-
-			out-ports {
-				port {
-					etf_out_port: endpoint {
-						remote-endpoint = <&etr_in_port>;
-					};
-				};
-			};
-		};
-
-		etr@28c06000 {
-			compatible = "arm,coresight-tmc", "arm,primecell";
-			reg = <0x28c06000 0x1000>;
-			clocks = <&clk IMX8MQ_CLK_MAIN_AXI>;
-			clock-names = "apb_pclk";
-
-			in-ports {
-				port {
-					etr_in_port: endpoint {
-						remote-endpoint = <&etf_out_port>;
-					};
-				};
-			};
+		caam_sm: caam-sm@100000 {
+			compatible = "fsl,imx6q-caam-sm";
+			reg = <0x100000 0x8000>;
 		};
 
 		aips1: bus@30000000 { /* AIPS1 */
@@ -531,57 +347,70 @@
 
 			sai1: sai@30010000 {
 				#sound-dai-cells = <0>;
-				compatible = "fsl,imx8mq-sai";
+				compatible = "fsl,imx8mq-sai",
+					     "fsl,imx6sx-sai";
 				reg = <0x30010000 0x10000>;
 				interrupts = <GIC_SPI 95 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_SAI1_IPG>,
-				         <&clk IMX8MQ_CLK_SAI1_ROOT>,
-				         <&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
-				clock-names = "bus", "mclk1", "mclk2", "mclk3";
-				dmas = <&sdma2 8 24 0>, <&sdma1 9 24 0>;
+					<&clk IMX8MQ_CLK_DUMMY>,
+					<&clk IMX8MQ_CLK_SAI1_ROOT>,
+					<&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
+				clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
+				dmas = <&sdma2 8 1 0>, <&sdma2 9 1 0>;
 				dma-names = "rx", "tx";
+				fsl,dataline = <0 0xff 0xff>;
 				status = "disabled";
 			};
 
 			sai6: sai@30030000 {
 				#sound-dai-cells = <0>;
-				compatible = "fsl,imx8mq-sai";
+				compatible = "fsl,imx8mq-sai",
+					     "fsl,imx6sx-sai";
 				reg = <0x30030000 0x10000>;
 				interrupts = <GIC_SPI 90 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_SAI6_IPG>,
-				         <&clk IMX8MQ_CLK_SAI6_ROOT>,
-				         <&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
-				clock-names = "bus", "mclk1", "mclk2", "mclk3";
+					<&clk IMX8MQ_CLK_DUMMY>,
+					<&clk IMX8MQ_CLK_SAI6_ROOT>,
+					<&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
+				clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
 				dmas = <&sdma2 4 24 0>, <&sdma2 5 24 0>;
 				dma-names = "rx", "tx";
+				fsl,shared-interrupt;
 				status = "disabled";
 			};
 
 			sai5: sai@30040000 {
 				#sound-dai-cells = <0>;
-				compatible = "fsl,imx8mq-sai";
+				compatible = "fsl,imx8mq-sai",
+					     "fsl,imx6sx-sai";
 				reg = <0x30040000 0x10000>;
 				interrupts = <GIC_SPI 90 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_SAI5_IPG>,
-				         <&clk IMX8MQ_CLK_SAI5_ROOT>,
-				         <&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
-				clock-names = "bus", "mclk1", "mclk2", "mclk3";
+					<&clk IMX8MQ_CLK_DUMMY>,
+					<&clk IMX8MQ_CLK_SAI5_ROOT>,
+					<&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
+				clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
 				dmas = <&sdma2 2 24 0>, <&sdma2 3 24 0>;
 				dma-names = "rx", "tx";
+				fsl,shared-interrupt;
+				fsl,dataline = <0 0xf 0xf>;
 				status = "disabled";
 			};
 
 			sai4: sai@30050000 {
 				#sound-dai-cells = <0>;
-				compatible = "fsl,imx8mq-sai";
+				compatible = "fsl,imx8mq-sai",
+					     "fsl,imx6sx-sai";
 				reg = <0x30050000 0x10000>;
 				interrupts = <GIC_SPI 100 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_SAI4_IPG>,
-				         <&clk IMX8MQ_CLK_SAI4_ROOT>,
-				         <&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
-				clock-names = "bus", "mclk1", "mclk2", "mclk3";
+					<&clk IMX8MQ_CLK_DUMMY>,
+					<&clk IMX8MQ_CLK_SAI4_ROOT>,
+					<&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
+				clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
 				dmas = <&sdma2 0 24 0>, <&sdma2 1 24 0>;
 				dma-names = "rx", "tx";
+				fsl,dataline = <0 0x0 0xf>;
 				status = "disabled";
 			};
 
@@ -780,7 +609,7 @@
 			};
 
 			ocotp: efuse@30350000 {
-				compatible = "fsl,imx8mq-ocotp", "syscon";
+				compatible = "fsl,imx8mq-ocotp", "syscon", "simple-mfd";
 				reg = <0x30350000 0x10000>;
 				clocks = <&clk IMX8MQ_CLK_OCOTP_ROOT>;
 				#address-cells = <1>;
@@ -810,6 +639,12 @@
 				fec_mac_address: mac-address@90 { /* 0x640 */
 					reg = <0x90 6>;
 				};
+
+				imx8mq_soc: imx8mq-soc {
+					compatible = "fsl,imx8mq-soc";
+					nvmem-cells = <&imx8mq_uid>;
+					nvmem-cell-names = "soc_unique_id";
+				};
 			};
 
 			anatop: clock-controller@30360000 {
@@ -819,6 +654,13 @@
 				#clock-cells = <1>;
 			};
 
+			caam_snvs: caam-snvs@30370000 {
+				compatible = "fsl,imx6q-caam-snvs";
+				reg = <0x30370000 0x10000>;
+				clocks = <&clk IMX8MQ_CLK_SNVS_ROOT>;
+				clock-names = "ipg";
+			};
+
 			snvs: snvs@30370000 {
 				compatible = "fsl,sec-v4.0-mon", "syscon", "simple-mfd";
 				reg = <0x30370000 0x10000>;
@@ -864,20 +706,25 @@
 						  <&clk IMX8MQ_AUDIO_PLL1_BYPASS>,
 						  <&clk IMX8MQ_AUDIO_PLL2_BYPASS>,
 						  <&clk IMX8MQ_AUDIO_PLL1>,
-						  <&clk IMX8MQ_AUDIO_PLL2>;
+						  <&clk IMX8MQ_AUDIO_PLL2>,
+						  <&clk IMX8MQ_CLK_NAND_USDHC_BUS>;
 				assigned-clock-rates = <0>, <0>,
 						       <800000000>,
 						       <0>,
 						       <0>,
 						       <0>,
 						       <786432000>,
-						       <722534400>;
+						       <722534400>,
+						       <0>;
 				assigned-clock-parents = <&clk IMX8MQ_SYS1_PLL_800M>,
 							 <&clk IMX8MQ_ARM_PLL_OUT>,
-							 <0>,
+							 <&clk IMX8MQ_SYS1_PLL_800M>,
 							 <&clk IMX8MQ_SYS2_PLL_500M>,
 							 <&clk IMX8MQ_AUDIO_PLL1>,
-							 <&clk IMX8MQ_AUDIO_PLL2>;
+							 <&clk IMX8MQ_AUDIO_PLL2>,
+							 <0>,
+						         <0>,
+							 <&clk IMX8MQ_SYS1_PLL_266M>;
 			};
 
 			src: reset-controller@30390000 {
@@ -893,6 +740,7 @@
 				interrupts = <GIC_SPI 87 IRQ_TYPE_LEVEL_HIGH>;
 				interrupt-parent = <&gic>;
 				interrupt-controller;
+				broken-wake-request-signals;
 				#interrupt-cells = <3>;
 
 				pgc {
@@ -1061,7 +909,7 @@
 				 <0x08000000 0x08000000 0x10000000>;
 
 			spdif1: spdif@30810000 {
-				compatible = "fsl,imx35-spdif";
+				compatible = "fsl,imx8mm-spdif", "fsl,imx35-spdif";
 				reg = <0x30810000 0x10000>;
 				interrupts = <GIC_SPI 6 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_IPG_ROOT>, /* core */
@@ -1166,7 +1014,7 @@
 			};
 
 			spdif2: spdif@308a0000 {
-				compatible = "fsl,imx35-spdif";
+				compatible = "fsl,imx8mm-spdif", "fsl,imx35-spdif";
 				reg = <0x308a0000 0x10000>;
 				interrupts = <GIC_SPI 13 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_IPG_ROOT>, /* core */
@@ -1195,9 +1043,10 @@
 				reg = <0x308b0000 0x10000>;
 				interrupts = <GIC_SPI 96 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_SAI2_IPG>,
+					 <&clk IMX8MQ_CLK_DUMMY>,
 					 <&clk IMX8MQ_CLK_SAI2_ROOT>,
 					 <&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
-				clock-names = "bus", "mclk1", "mclk2", "mclk3";
+				clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
 				dmas = <&sdma1 10 24 0>, <&sdma1 11 24 0>;
 				dma-names = "rx", "tx";
 				status = "disabled";
@@ -1205,13 +1054,15 @@
 
 			sai3: sai@308c0000 {
 				#sound-dai-cells = <0>;
-				compatible = "fsl,imx8mq-sai";
+				compatible = "fsl,imx8mq-sai",
+					     "fsl,imx6sx-sai";
 				reg = <0x308c0000 0x10000>;
 				interrupts = <GIC_SPI 50 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_SAI3_IPG>,
-				         <&clk IMX8MQ_CLK_SAI3_ROOT>,
-				         <&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
-				clock-names = "bus", "mclk1", "mclk2", "mclk3";
+					<&clk IMX8MQ_CLK_DUMMY>,
+					<&clk IMX8MQ_CLK_SAI3_ROOT>,
+					<&clk IMX8MQ_CLK_DUMMY>, <&clk IMX8MQ_CLK_DUMMY>;
+				clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3";
 				dmas = <&sdma1 12 24 0>, <&sdma1 13 24 0>;
 				dma-names = "rx", "tx";
 				status = "disabled";
@@ -1251,20 +1102,24 @@
 			mipi_dsi: dsi@30a00000 {
 				compatible = "fsl,imx8mq-nwl-dsi";
 				reg = <0x30a00000 0x300>;
-				#address-cells = <1>;
-				#size-cells = <0>;
 				clocks = <&clk IMX8MQ_CLK_DSI_CORE>,
 					 <&clk IMX8MQ_CLK_DSI_AHB>,
 					 <&clk IMX8MQ_CLK_DSI_IPG_DIV>,
 					 <&clk IMX8MQ_CLK_DSI_PHY_REF>,
+					 <&clk IMX8MQ_VIDEO_PLL1>,
 					 <&clk IMX8MQ_CLK_LCDIF_PIXEL>;
-				clock-names = "core", "rx_esc", "tx_esc", "phy_ref", "lcdif";
-				assigned-clocks = <&clk IMX8MQ_CLK_DSI_AHB>,
+				clock-names = "core", "rx_esc", "tx_esc", "phy_ref", "video_pll", "lcdif";
+				assigned-clocks = <&clk IMX8MQ_CLK_DSI_PHY_REF>,
 						  <&clk IMX8MQ_CLK_DSI_CORE>,
+						  <&clk IMX8MQ_CLK_DSI_AHB>,
 						  <&clk IMX8MQ_CLK_DSI_IPG_DIV>;
-				assigned-clock-parents = <&clk IMX8MQ_SYS1_PLL_80M>,
-							 <&clk IMX8MQ_SYS1_PLL_266M>;
-				assigned-clock-rates = <80000000>, <266000000>, <20000000>;
+				assigned-clock-parents = <&clk IMX8MQ_VIDEO_PLL1_OUT>,
+							 <&clk IMX8MQ_SYS1_PLL_266M>,
+							 <&clk IMX8MQ_SYS1_PLL_80M>;
+				assigned-clock-rates = <27000000>,
+						       <266000000>,
+						       <80000000>,
+						       <20000000>;
 				interrupts = <GIC_SPI 34 IRQ_TYPE_LEVEL_HIGH>;
 				mux-controls = <&mux 0>;
 				power-domains = <&pgc_mipi>;
@@ -1289,6 +1144,16 @@
 							reg = <0>;
 							remote-endpoint = <&lcdif_mipi_dsi>;
 						};
+
+						endpoint@1 {
+							reg = <1>;
+						};
+					};
+
+					port@1 {
+						reg = <1>;
+						endpoint {
+						};
 					};
 				};
 			};
@@ -1364,108 +1229,33 @@
 				status = "disabled";
 			};
 
-			mipi_csi1: csi@30a70000 {
-				compatible = "fsl,imx8mq-mipi-csi2";
+			mipi_csi_1: mipi_csi1@30a70000 {
+				compatible = "fsl,mxc-mipi-csi2_yav";
 				reg = <0x30a70000 0x1000>;
+				interrupts = <GIC_SPI 44 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX8MQ_CLK_CSI1_CORE>,
-				   <&clk IMX8MQ_CLK_CSI1_ESC>,
-				   <&clk IMX8MQ_CLK_CSI1_PHY_REF>;
-				clock-names = "core", "esc", "ui";
+						<&clk IMX8MQ_CLK_CSI1_ESC>,
+						<&clk IMX8MQ_CLK_CSI1_PHY_REF>;
+				clock-names = "clk_core", "clk_esc", "clk_pxl";
 				assigned-clocks = <&clk IMX8MQ_CLK_CSI1_CORE>,
-				    <&clk IMX8MQ_CLK_CSI1_PHY_REF>,
-				    <&clk IMX8MQ_CLK_CSI1_ESC>;
-				assigned-clock-rates = <266000000>, <333000000>, <66000000>;
-				assigned-clock-parents = <&clk IMX8MQ_SYS1_PLL_266M>,
-					<&clk IMX8MQ_SYS2_PLL_1000M>,
-					<&clk IMX8MQ_SYS1_PLL_800M>;
+						  <&clk IMX8MQ_CLK_CSI1_PHY_REF>,
+						  <&clk IMX8MQ_CLK_CSI1_ESC>;
+				assigned-clock-rates = <133000000>, <100000000>, <66000000>;
 				power-domains = <&pgc_mipi_csi1>;
-				resets = <&src IMX8MQ_RESET_MIPI_CSI1_CORE_RESET>,
-					 <&src IMX8MQ_RESET_MIPI_CSI1_PHY_REF_RESET>,
-					 <&src IMX8MQ_RESET_MIPI_CSI1_ESC_RESET>;
-				fsl,mipi-phy-gpr = <&iomuxc_gpr 0x88>;
-				interconnects = <&noc IMX8MQ_ICM_CSI1 &noc IMX8MQ_ICS_DRAM>;
-				interconnect-names = "dram";
+				csis-phy-reset = <&src 0x4c 7>;
+				phy-gpr = <&iomuxc_gpr 0x88>;
 				status = "disabled";
-
-				ports {
-					#address-cells = <1>;
-					#size-cells = <0>;
-
-					port@1 {
-						reg = <1>;
-
-						csi1_mipi_ep: endpoint {
-							remote-endpoint = <&csi1_ep>;
-						};
-					};
-				};
 			};
 
-			csi1: csi@30a90000 {
-				compatible = "fsl,imx8mq-csi";
+			csi1_bridge: csi1_bridge@30a90000 {
+				compatible = "fsl,imx8mq-csi", "fsl,imx6s-csi";
 				reg = <0x30a90000 0x10000>;
 				interrupts = <GIC_SPI 42 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&clk IMX8MQ_CLK_CSI1_ROOT>;
-				clock-names = "mclk";
+				clocks = <&clk IMX8MQ_CLK_DUMMY>,
+					<&clk IMX8MQ_CLK_CSI1_ROOT>,
+					<&clk IMX8MQ_CLK_DUMMY>;
+				clock-names = "disp-axi", "csi_mclk", "disp_dcic";
 				status = "disabled";
-
-				port {
-					csi1_ep: endpoint {
-						remote-endpoint = <&csi1_mipi_ep>;
-					};
-				};
-			};
-
-			mipi_csi2: csi@30b60000 {
-				compatible = "fsl,imx8mq-mipi-csi2";
-				reg = <0x30b60000 0x1000>;
-				clocks = <&clk IMX8MQ_CLK_CSI2_CORE>,
-				   <&clk IMX8MQ_CLK_CSI2_ESC>,
-				   <&clk IMX8MQ_CLK_CSI2_PHY_REF>;
-				clock-names = "core", "esc", "ui";
-				assigned-clocks = <&clk IMX8MQ_CLK_CSI2_CORE>,
-				    <&clk IMX8MQ_CLK_CSI2_PHY_REF>,
-				    <&clk IMX8MQ_CLK_CSI2_ESC>;
-				assigned-clock-rates = <266000000>, <333000000>, <66000000>;
-				assigned-clock-parents = <&clk IMX8MQ_SYS1_PLL_266M>,
-					<&clk IMX8MQ_SYS2_PLL_1000M>,
-					<&clk IMX8MQ_SYS1_PLL_800M>;
-				power-domains = <&pgc_mipi_csi2>;
-				resets = <&src IMX8MQ_RESET_MIPI_CSI2_CORE_RESET>,
-					 <&src IMX8MQ_RESET_MIPI_CSI2_PHY_REF_RESET>,
-					 <&src IMX8MQ_RESET_MIPI_CSI2_ESC_RESET>;
-				fsl,mipi-phy-gpr = <&iomuxc_gpr 0xa4>;
-				interconnects = <&noc IMX8MQ_ICM_CSI2 &noc IMX8MQ_ICS_DRAM>;
-				interconnect-names = "dram";
-				status = "disabled";
-
-				ports {
-					#address-cells = <1>;
-					#size-cells = <0>;
-
-					port@1 {
-						reg = <1>;
-
-						csi2_mipi_ep: endpoint {
-							remote-endpoint = <&csi2_ep>;
-						};
-					};
-				};
-			};
-
-			csi2: csi@30b80000 {
-				compatible = "fsl,imx8mq-csi";
-				reg = <0x30b80000 0x10000>;
-				interrupts = <GIC_SPI 43 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&clk IMX8MQ_CLK_CSI2_ROOT>;
-				clock-names = "mclk";
-				status = "disabled";
-
-				port {
-					csi2_ep: endpoint {
-						remote-endpoint = <&csi2_mipi_ep>;
-					};
-				};
 			};
 
 			mu: mailbox@30aa0000 {
@@ -1506,6 +1296,35 @@
 				status = "disabled";
 			};
 
+			mipi_csi_2: mipi_csi2@30b60000 {
+				compatible = "fsl,mxc-mipi-csi2_yav";
+				reg = <0x30b60000 0x1000>;
+				interrupts = <GIC_SPI 45 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&clk IMX8MQ_CLK_CSI2_CORE>,
+						<&clk IMX8MQ_CLK_CSI2_ESC>,
+						<&clk IMX8MQ_CLK_CSI2_PHY_REF>;
+				clock-names = "clk_core", "clk_esc", "clk_pxl";
+				assigned-clocks = <&clk IMX8MQ_CLK_CSI2_CORE>,
+						  <&clk IMX8MQ_CLK_CSI2_PHY_REF>,
+						  <&clk IMX8MQ_CLK_CSI2_ESC>;
+				assigned-clock-rates = <133000000>, <100000000>, <66000000>;
+				power-domains = <&pgc_mipi_csi2>;
+				csis-phy-reset = <&src 0x50 7>;
+				phy-gpr = <&iomuxc_gpr 0xa4>;
+				status = "disabled";
+			};
+
+			csi2_bridge: csi2_bridge@30b80000 {
+				compatible = "fsl,imx8mq-csi", "fsl,imx6s-csi";
+				reg = <0x30b80000 0x10000>;
+				interrupts = <GIC_SPI 43 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&clk IMX8MQ_CLK_DUMMY>,
+					<&clk IMX8MQ_CLK_CSI2_ROOT>,
+					<&clk IMX8MQ_CLK_DUMMY>;
+				clock-names = "disp-axi", "csi_mclk", "disp_dcic";
+				status = "disabled";
+			};
+
 			qspi0: spi@30bb0000 {
 				#address-cells = <1>;
 				#size-cells = <0>;
@@ -1563,31 +1382,6 @@
 			};
 		};
 
-		noc: interconnect@32700000 {
-			compatible = "fsl,imx8mq-noc", "fsl,imx8m-noc";
-			reg = <0x32700000 0x100000>;
-			clocks = <&clk IMX8MQ_CLK_NOC>;
-			fsl,ddrc = <&ddrc>;
-			#interconnect-cells = <1>;
-			operating-points-v2 = <&noc_opp_table>;
-
-			noc_opp_table: opp-table {
-				compatible = "operating-points-v2";
-
-				opp-133000000 {
-					opp-hz = /bits/ 64 <133333333>;
-				};
-
-				opp-400000000 {
-					opp-hz = /bits/ 64 <400000000>;
-				};
-
-				opp-800000000 {
-					opp-hz = /bits/ 64 <800000000>;
-				};
-			};
-		};
-
 		aips4: bus@32c00000 { /* AIPS4 */
 			compatible = "fsl,aips-bus", "simple-bus";
 			reg = <0x32c00000 0x400000>;
@@ -1595,6 +1389,14 @@
 			#size-cells = <1>;
 			ranges = <0x32c00000 0x32c00000 0x400000>;
 
+			hdmi: hdmi@32c00000 {
+				reg = <0x32c00000 0x100000>,
+					<0x32e40000 0x40000>;
+				interrupts = <GIC_SPI 16 IRQ_TYPE_LEVEL_HIGH>,
+							 <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+				interrupt-names = "plug_in", "plug_out";
+			};
+
 			irqsteer: interrupt-controller@32e2d000 {
 				compatible = "fsl,imx8m-irqsteer", "fsl,imx-irqsteer";
 				reg = <0x32e2d000 0x1000>;
@@ -1606,31 +1408,33 @@
 				interrupt-controller;
 				#interrupt-cells = <1>;
 			};
-		};
 
-		gpu: gpu@38000000 {
-			compatible = "vivante,gc";
-			reg = <0x38000000 0x40000>;
-			interrupts = <GIC_SPI 3 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&clk IMX8MQ_CLK_GPU_ROOT>,
-			         <&clk IMX8MQ_CLK_GPU_SHADER_DIV>,
-			         <&clk IMX8MQ_CLK_GPU_AXI>,
-			         <&clk IMX8MQ_CLK_GPU_AHB>;
-			clock-names = "core", "shader", "bus", "reg";
-			#cooling-cells = <2>;
-			assigned-clocks = <&clk IMX8MQ_CLK_GPU_CORE_SRC>,
-			                  <&clk IMX8MQ_CLK_GPU_SHADER_SRC>,
-			                  <&clk IMX8MQ_CLK_GPU_AXI>,
-			                  <&clk IMX8MQ_CLK_GPU_AHB>,
-			                  <&clk IMX8MQ_GPU_PLL_BYPASS>;
-			assigned-clock-parents = <&clk IMX8MQ_GPU_PLL_OUT>,
-			                         <&clk IMX8MQ_GPU_PLL_OUT>,
-			                         <&clk IMX8MQ_GPU_PLL_OUT>,
-			                         <&clk IMX8MQ_GPU_PLL_OUT>,
-			                         <&clk IMX8MQ_GPU_PLL>;
-			assigned-clock-rates = <800000000>, <800000000>,
-			                       <800000000>, <800000000>, <0>;
-			power-domains = <&pgc_gpu>;
+			dcss: display-controller@32e00000 {
+				compatible = "nxp,imx8mq-dcss";
+				reg = <0x32e00000 0x2d000>, <0x32e2f000 0x1000>;
+				interrupts = <6>, <8>, <9>, <16>, <17>;
+				interrupt-names = "ctxld", "ctxld_kick", "vblank",
+								  "dtrc_ch1", "dtrc_ch2";
+				interrupt-parent = <&irqsteer>;
+				clocks = <&clk IMX8MQ_CLK_DISP_APB_ROOT>,
+					 <&clk IMX8MQ_CLK_DISP_AXI_ROOT>,
+					 <&clk IMX8MQ_CLK_DISP_RTRM_ROOT>,
+					 <&clk IMX8MQ_VIDEO2_PLL_OUT>,
+					 <&clk IMX8MQ_CLK_DISP_DTRC>,
+					 <&clk IMX8MQ_VIDEO2_PLL1_REF_SEL>,
+					 <&hdmi_phy_27m>;
+				clock-names = "apb", "axi", "rtrm", "pix", "dtrc", "pll_src",
+						      "pll_phy_ref";
+				assigned-clocks = <&clk IMX8MQ_CLK_DISP_AXI>,
+						  <&clk IMX8MQ_CLK_DISP_RTRM>,
+						  <&clk IMX8MQ_VIDEO2_PLL1_REF_SEL>;
+				assigned-clock-parents = <&clk IMX8MQ_SYS1_PLL_800M>,
+							 <&clk IMX8MQ_SYS1_PLL_800M>,
+							 <&clk IMX8MQ_CLK_27M>;
+				assigned-clock-rates = <800000000>,
+							   <400000000>;
+				status = "disabled";
+			};
 		};
 
 		usb_dwc3_0: usb@38100000 {
@@ -1649,7 +1453,9 @@
 			phys = <&usb3_phy0>, <&usb3_phy0>;
 			phy-names = "usb2-phy", "usb3-phy";
 			power-domains = <&pgc_otg1>;
+			snps,power-down-scale = <2>;
 			snps,parkmode-disable-ss-quirk;
+			snps,dis_rxdet_inp3_quirk;
 			status = "disabled";
 		};
 
@@ -1681,7 +1487,9 @@
 			phys = <&usb3_phy1>, <&usb3_phy1>;
 			phy-names = "usb2-phy", "usb3-phy";
 			power-domains = <&pgc_otg2>;
+			snps,power-down-scale = <2>;
 			snps,parkmode-disable-ss-quirk;
+			snps,dis_rxdet_inp3_quirk;
 			status = "disabled";
 		};
 
@@ -1697,31 +1505,33 @@
 			status = "disabled";
 		};
 
-		vpu_g1: video-codec@38300000 {
-			compatible = "nxp,imx8mq-vpu-g1";
-			reg = <0x38300000 0x10000>;
-			interrupts = <GIC_SPI 7 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&clk IMX8MQ_CLK_VPU_G1_ROOT>;
-			power-domains = <&vpu_blk_ctrl IMX8MQ_VPUBLK_PD_G1>;
-		};
-
-		vpu_g2: video-codec@38310000 {
-			compatible = "nxp,imx8mq-vpu-g2";
-			reg = <0x38310000 0x10000>;
-			interrupts = <GIC_SPI 8 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&clk IMX8MQ_CLK_VPU_G2_ROOT>;
-			power-domains = <&vpu_blk_ctrl IMX8MQ_VPUBLK_PD_G2>;
-		};
-
-		vpu_blk_ctrl: blk-ctrl@38320000 {
-			compatible = "fsl,imx8mq-vpu-blk-ctrl";
-			reg = <0x38320000 0x100>;
-			power-domains = <&pgc_vpu>, <&pgc_vpu>, <&pgc_vpu>;
-			power-domain-names = "bus", "g1", "g2";
-			clocks = <&clk IMX8MQ_CLK_VPU_G1_ROOT>,
-				 <&clk IMX8MQ_CLK_VPU_G2_ROOT>;
-			clock-names = "g1", "g2";
-			#power-domain-cells = <1>;
+		dma_apbh: dma-apbh@33000000 {
+			compatible = "fsl,imx7d-dma-apbh", "fsl,imx28-dma-apbh";
+			reg = <0x33000000 0x2000>;
+			interrupts = <GIC_SPI 12 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 12 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 12 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 12 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "gpmi0", "gpmi1", "gpmi2", "gpmi3";
+			#dma-cells = <1>;
+			dma-channels = <4>;
+			clocks = <&clk IMX8MQ_CLK_NAND_USDHC_BUS_RAWNAND_CLK>;
+		};
+
+		gpmi: gpmi-nand@33002000{
+			compatible = "fsl,imx7d-gpmi-nand";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			reg = <0x33002000 0x2000>, <0x33004000 0x4000>;
+			reg-names = "gpmi-nand", "bch";
+			interrupts = <GIC_SPI 14 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "bch";
+			clocks = <&clk IMX8MQ_CLK_RAWNAND_ROOT>,
+				<&clk IMX8MQ_CLK_NAND_USDHC_BUS_RAWNAND_CLK>;
+			clock-names = "gpmi_io", "gpmi_bch_apb";
+			dmas = <&dma_apbh 0>;
+			dma-names = "rx-tx";
+			status = "disabled";
 		};
 
 		pcie0: pcie@33800000 {
@@ -1812,9 +1622,11 @@
 
 		pcie1_ep: pcie-ep@33c00000 {
 			compatible = "fsl,imx8mq-pcie-ep";
-			reg = <0x33c00000 0x000400000>,
-			      <0x20000000 0x08000000>;
-			reg-names = "dbi", "addr_space";
+			reg = <0x33c00000 0x100000>,
+			      <0x33d00000 0x100000>,
+			      <0x33f00000 0x100000>,
+			      <0x20000000 0x8000000>;
+			reg-names = "dbi", "dbi2", "atu", "addr_space";
 			num-lanes = <1>;
 			interrupts = <GIC_SPI 80 IRQ_TYPE_LEVEL_HIGH>;
 			interrupt-names = "dma";
@@ -1855,22 +1667,58 @@
 			interrupt-parent = <&gic>;
 		};
 
-		ddrc: memory-controller@3d400000 {
-			compatible = "fsl,imx8mq-ddrc", "fsl,imx8m-ddrc";
-			reg = <0x3d400000 0x400000>;
-			clock-names = "core", "pll", "alt", "apb";
-			clocks = <&clk IMX8MQ_CLK_DRAM_CORE>,
-				 <&clk IMX8MQ_DRAM_PLL_OUT>,
-				 <&clk IMX8MQ_CLK_DRAM_ALT>,
-				 <&clk IMX8MQ_CLK_DRAM_APB>;
-			status = "disabled";
-		};
-
 		ddr-pmu@3d800000 {
 			compatible = "fsl,imx8mq-ddr-pmu", "fsl,imx8m-ddr-pmu";
 			reg = <0x3d800000 0x400000>;
 			interrupt-parent = <&gic>;
 			interrupts = <GIC_SPI 98 IRQ_TYPE_LEVEL_HIGH>;
 		};
+
+		vpu: vpu@38300000 {
+			compatible = "nxp,imx8mq-hantro";
+			reg = <0x38300000 0x200000>;
+			reg-names = "regs_hantro";
+			interrupts = <GIC_SPI 7 IRQ_TYPE_LEVEL_HIGH>, <GIC_SPI 8 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "irq_hantro_g1", "irq_hantro_g2";
+			clocks = <&clk IMX8MQ_CLK_VPU_G1_ROOT>, <&clk IMX8MQ_CLK_VPU_G2_ROOT>, <&clk IMX8MQ_CLK_VPU_DEC_ROOT>;
+			clock-names = "clk_hantro_g1", "clk_hantro_g2", "clk_hantro_bus";
+			assigned-clocks = <&clk IMX8MQ_CLK_VPU_G1>, <&clk IMX8MQ_CLK_VPU_G2>, <&clk IMX8MQ_CLK_VPU_BUS>;
+			assigned-clock-parents = <&clk IMX8MQ_VPU_PLL_OUT>, <&clk IMX8MQ_VPU_PLL_OUT>, <&clk IMX8MQ_SYS1_PLL_800M>;
+			assigned-clock-rates = <600000000>, <600000000>, <800000000>;
+			power-domains = <&pgc_vpu>;
+			status = "disabled";
+		};
+	};
+
+	gpu3d: gpu3d@38000000 {
+		compatible = "fsl,imx8mq-gpu", "fsl,imx6q-gpu";
+		reg = <0x0 0x38000000 0x0 0x40000>, <0x0 0x40000000 0x0 0xC0000000>, <0x0 0x0 0x0 0x10000000>;
+		reg-names = "iobase_3d", "phys_baseaddr", "contiguous_mem";
+		interrupts = <GIC_SPI 3 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-names = "irq_3d";
+		clocks = 	<&clk IMX8MQ_CLK_GPU_ROOT>,
+				<&clk IMX8MQ_CLK_GPU_SHADER_DIV>,
+				<&clk IMX8MQ_CLK_GPU_AXI>,
+				<&clk IMX8MQ_CLK_GPU_AHB>;
+		clock-names = "gpu3d_clk", "gpu3d_shader_clk", "gpu3d_axi_clk", "gpu3d_ahb_clk";
+		assigned-clocks = <&clk IMX8MQ_CLK_GPU_CORE_SRC>,
+				<&clk IMX8MQ_CLK_GPU_SHADER_SRC>,
+				<&clk IMX8MQ_CLK_GPU_AXI>,
+				<&clk IMX8MQ_CLK_GPU_AHB>;
+		assigned-clock-parents = <&clk IMX8MQ_GPU_PLL_OUT>,
+				<&clk IMX8MQ_GPU_PLL_OUT>,
+				<&clk IMX8MQ_GPU_PLL_OUT>,
+				<&clk IMX8MQ_GPU_PLL_OUT>;
+		assigned-clock-rates = <800000000>, <800000000>, <800000000>, <400000000>;
+		power-domains = <&pgc_gpu>;
+		status = "disabled";
+
+		throttle,max_state = <1>;
+		#cooling-cells = <2>;
+	};
+
+	vpu_v4l2: vpu_v4l2 {
+		compatible = "nxp,imx8m-vsiv4l2";
+		status = "disabled";
 	};
 };
