--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/boot/dts/freescale/imx8qp-lpddr4-val.dts	1970-01-01 01:00:00.000000000 +0100
+++ ./arch/arm64/boot/dts/freescale/imx8qp-lpddr4-val.dts	2026-04-08 16:21:02.066534792 +0200
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2019 NXP
+ */
+
+/dts-v1/;
+
+#include "imx8qp.dtsi"
+#include "imx8q-val.dtsi"
+
+/ {
+	model = "Freescale i.MX8QP Validation Board";
+	compatible = "fsl,imx8qp-val", "fsl,imx8qp", "fsl,imx8qm";
+};
