--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/arm64/kvm/hyp/pgtable.c	2026-04-08 18:20:05.106885154 +0200
+++ ./arch/arm64/kvm/hyp/pgtable.c	2026-04-08 16:21:02.453521707 +0200
@@ -50,6 +50,8 @@
 					 KVM_PTE_LEAF_ATTR_LO_S2_S2AP_W | \
 					 KVM_PTE_LEAF_ATTR_HI_S2_XN)
 
+#define KVM_PTE_LEAF_ATTR_S2_DEVICE     BIT(55)
+
 #define KVM_INVALID_PTE_OWNER_MASK	GENMASK(9, 2)
 #define KVM_MAX_OWNER_ID		1
 
@@ -163,6 +165,7 @@
 	pte |= attr & (KVM_PTE_LEAF_ATTR_LO | KVM_PTE_LEAF_ATTR_HI);
 	pte |= FIELD_PREP(KVM_PTE_TYPE, type);
 	pte |= KVM_PTE_VALID;
+	pte |= attr & KVM_PTE_LEAF_ATTR_S2_DEVICE;
 
 	return pte;
 }
@@ -698,7 +701,10 @@
 	bool device = prot & KVM_PGTABLE_PROT_DEVICE;
 	kvm_pte_t attr = device ? KVM_S2_MEMATTR(pgt, DEVICE_nGnRE) :
 			    KVM_S2_MEMATTR(pgt, NORMAL);
-	u32 sh = KVM_PTE_LEAF_ATTR_LO_S2_SH_IS;
+	u32 sh = 0;
+
+	if (!(prot & KVM_PGTABLE_PROT_DEVICE_NS))
+		sh = KVM_PTE_LEAF_ATTR_LO_S2_SH_IS;
 
 	if (!(prot & KVM_PGTABLE_PROT_X))
 		attr |= KVM_PTE_LEAF_ATTR_HI_S2_XN;
@@ -714,6 +720,8 @@
 	attr |= FIELD_PREP(KVM_PTE_LEAF_ATTR_LO_S2_SH, sh);
 	attr |= KVM_PTE_LEAF_ATTR_LO_S2_AF;
 	attr |= prot & KVM_PTE_LEAF_ATTR_HI_SW;
+	if ((prot & KVM_PGTABLE_PROT_DEVICE_SH) || (prot & KVM_PGTABLE_PROT_DEVICE_NS))
+		attr |= KVM_PTE_LEAF_ATTR_S2_DEVICE;
 	*ptep = attr;
 
 	return 0;
@@ -946,7 +954,8 @@
 
 	/* Perform CMOs before installation of the guest stage-2 PTE */
 	if (!kvm_pgtable_walk_skip_cmo(ctx) && mm_ops->dcache_clean_inval_poc &&
-	    stage2_pte_cacheable(pgt, new))
+	    stage2_pte_cacheable(pgt, new) &&
+	    !(new & KVM_PTE_LEAF_ATTR_S2_DEVICE))
 		mm_ops->dcache_clean_inval_poc(kvm_pte_follow(new, mm_ops),
 					       granule);
 
@@ -1332,7 +1341,8 @@
 	struct kvm_pgtable *pgt = ctx->arg;
 	struct kvm_pgtable_mm_ops *mm_ops = pgt->mm_ops;
 
-	if (!kvm_pte_valid(ctx->old) || !stage2_pte_cacheable(pgt, ctx->old))
+	if (!kvm_pte_valid(ctx->old) || !stage2_pte_cacheable(pgt, ctx->old) ||
+	   (ctx->old & KVM_PTE_LEAF_ATTR_S2_DEVICE))
 		return 0;
 
 	if (mm_ops->dcache_clean_inval_poc)
