--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/powerpc/boot/dts/fsl/bsc9131si-post.dtsi	2026-04-08 18:20:05.268640167 +0200
+++ ./arch/powerpc/boot/dts/fsl/bsc9131si-post.dtsi	2026-04-08 16:21:03.151974168 +0200
@@ -35,7 +35,7 @@
 &ifc {
 	#address-cells = <2>;
 	#size-cells = <1>;
-	compatible = "fsl,ifc", "simple-bus";
+	compatible = "fsl,ifc";
 	interrupts = <16 2 0 0 20 2 0 0>;
 };
 
