--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/powerpc/boot/dts/fsl/qoriq-fman3-0-10g-1.dtsi	2026-04-08 18:20:05.272889781 +0200
+++ ./arch/powerpc/boot/dts/fsl/qoriq-fman3-0-10g-1.dtsi	2026-04-08 16:21:03.188836359 +0200
@@ -35,14 +35,14 @@
 fman@400000 {
 	fman0_rx_0x11: port@91000 {
 		cell-index = <0x11>;
-		compatible = "fsl,fman-v3-port-rx";
+		compatible = "fsl,fman-v3-port-rx", "fsl,fman-port-10g-rx";
 		reg = <0x91000 0x1000>;
 		fsl,fman-10g-port;
 	};
 
 	fman0_tx_0x31: port@b1000 {
 		cell-index = <0x31>;
-		compatible = "fsl,fman-v3-port-tx";
+		compatible = "fsl,fman-v3-port-tx", "fsl,fman-port-10g-tx";
 		reg = <0xb1000 0x1000>;
 		fsl,fman-10g-port;
 	};
