--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/powerpc/boot/dts/fsl/t1023si-post.dtsi	2026-04-08 18:20:05.273889808 +0200
+++ ./arch/powerpc/boot/dts/fsl/t1023si-post.dtsi	2026-04-08 16:21:03.191980085 +0200
@@ -52,7 +52,7 @@
 &ifc {
 	#address-cells = <2>;
 	#size-cells = <1>;
-	compatible = "fsl,ifc", "simple-bus";
+	compatible = "fsl,ifc";
 	interrupts = <25 2 0 0>;
 };
 
