--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/arch/powerpc/platforms/86xx/Kconfig	2026-04-08 18:20:05.363892317 +0200
+++ ./arch/powerpc/platforms/86xx/Kconfig	2026-04-08 16:21:03.442562868 +0200
@@ -4,6 +4,7 @@
 	depends on PPC_BOOK3S_32
 	select FSL_SOC
 	select ALTIVEC
+	select FSL_PMC if SUSPEND
 	help
 	  The Freescale E600 SoCs have 74xx cores.
 
