--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/crypto/tcrypt.c	2026-04-08 18:20:05.629899731 +0200
+++ ./crypto/tcrypt.c	2026-04-08 16:21:04.198444120 +0200
@@ -251,7 +251,7 @@
 	const int *b_size;
 	const char *key;
 	const char *e;
-	void *assoc;
+	void *assoc, *assoc_out;
 	char *iv;
 	int ret;
 
@@ -377,6 +377,8 @@
 
 				assoc = cur->axbuf[0];
 				memset(assoc, 0xff, aad_size);
+				assoc_out = cur->axbuf[1];
+				memset(assoc_out, 0xff, aad_size);
 
 				sg_init_aead(cur->sg, cur->xbuf,
 					     bs + (enc ? 0 : authsize),
@@ -384,7 +386,7 @@
 
 				sg_init_aead(cur->sgout, cur->xoutbuf,
 					     bs + (enc ? authsize : 0),
-					     assoc, aad_size);
+					     assoc_out, aad_size);
 
 				aead_request_set_ad(cur->req, aad_size);
 
@@ -402,6 +404,9 @@
 						       ret);
 						break;
 					}
+
+					memset(assoc, 0xff, aad_size);
+					memset(assoc_out, 0xff, aad_size);
 				}
 
 				aead_request_set_crypt(cur->req, cur->sg,
@@ -528,7 +533,7 @@
 	struct scatterlist *sg;
 	struct scatterlist *sgout;
 	const char *e;
-	void *assoc;
+	void *assoc, *assoc_out;
 	char *iv;
 	char *xbuf[XBUFSIZE];
 	char *xoutbuf[XBUFSIZE];
@@ -599,6 +604,8 @@
 
 			assoc = axbuf[0];
 			memset(assoc, 0xff, aad_size);
+			assoc_out = axbuf[1];
+			memset(assoc_out, 0xff, aad_size);
 
 			if ((*keysize + bs) > TVMEMSIZE * PAGE_SIZE) {
 				pr_err("template (%u) too big for tvmem (%lu)\n",
@@ -636,7 +643,7 @@
 				     assoc, aad_size);
 
 			sg_init_aead(sgout, xoutbuf,
-				     bs + (enc ? authsize : 0), assoc,
+				     bs + (enc ? authsize : 0), assoc_out,
 				     aad_size);
 
 			aead_request_set_ad(req, aad_size);
@@ -658,6 +665,9 @@
 					       ret);
 					break;
 				}
+
+				memset(assoc, 0xff, aad_size);
+				memset(assoc_out, 0xff, aad_size);
 			}
 
 			aead_request_set_crypt(req, sg, sgout,
@@ -1768,6 +1778,10 @@
 		ret = min(ret, tcrypt_test("hmac(streebog512)"));
 		break;
 
+	case 117:
+		ret += tcrypt_test("rsa");
+		break;
+
 	case 150:
 		ret = min(ret, tcrypt_test("ansi_cprng"));
 		break;
@@ -1855,6 +1869,12 @@
 		ret = min(ret, tcrypt_test("cfb(aria)"));
 		ret = min(ret, tcrypt_test("ctr(aria)"));
 		break;
+	case 193:
+		ret += tcrypt_test("tls11(hmac(sha1),cbc(aes))");
+		break;
+	case 194:
+		ret += tcrypt_test("tls12(hmac(sha256),cbc(aes))");
+		break;
 	case 200:
 		test_cipher_speed("ecb(aes)", ENCRYPT, sec, NULL, 0,
 				speed_template_16_24_32);
