--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/ata/Kconfig	2026-04-08 18:20:05.716902155 +0200
+++ ./drivers/ata/Kconfig	2026-04-08 16:21:04.401645394 +0200
@@ -206,6 +206,18 @@
 
 	  If unsure, say N.
 
+if AHCI_IMX
+
+config AHCI_IMX_PMP
+	bool "SATA Port Multiplier support on I.MX"
+	depends on SATA_PMP
+	default N
+	help
+	  This option enables support for the PMP on Freescale i.MX SoC's.
+
+	  If unsure, say N.
+endif # AHCI_IMX
+
 config AHCI_CEVA
 	tristate "CEVA AHCI SATA support"
 	depends on OF
