--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/ata/libahci.c	2026-04-08 18:20:05.720902267 +0200
+++ ./drivers/ata/libahci.c	2026-04-08 16:21:04.408029388 +0200
@@ -1522,6 +1522,14 @@
 	if (fbs_disabled)
 		ahci_enable_fbs(ap);
 
+#ifdef CONFIG_AHCI_IMX_PMP
+	if (ap->flags & (1 << 31)) {
+		if (ap->flags & (1 << 29))
+			ap->flags |= (1 << 30);
+		ata_msleep(ap, 40);
+	}
+#endif
+
 	return 0;
 
  fail:
@@ -1903,6 +1911,11 @@
 	if (unlikely(ap->pflags & ATA_PFLAG_RESETTING))
 		status &= ~PORT_IRQ_BAD_PMP;
 
+#ifdef CONFIG_AHCI_IMX_PMP
+	status &= ~PORT_IRQ_BAD_PMP;
+	status &= ~PORT_IRQ_IF_ERR;
+#endif
+
 	if (sata_lpm_ignore_phy_events(&ap->link)) {
 		status &= ~PORT_IRQ_PHYRDY;
 		ahci_scr_write(&ap->link, SCR_ERROR, SERR_PHYRDY_CHG);
@@ -1960,6 +1973,12 @@
 	void __iomem *port_mmio = ahci_port_base(ap);
 	u32 status;
 
+#ifdef CONFIG_AHCI_IMX_PMP
+	status = readl(port_mmio + PORT_SCR_NTF);
+	if (status)
+		ap->flags |= (1 << 31);
+#endif
+
 	status = readl(port_mmio + PORT_IRQ_STAT);
 	writel(status, port_mmio + PORT_IRQ_STAT);
 
@@ -2081,10 +2100,22 @@
 {
 	struct ahci_port_priv *pp = qc->ap->private_data;
 	u8 *rx_fis = pp->rx_fis;
+#ifdef CONFIG_AHCI_IMX_PMP
+	u8 *fis;
+#endif
 
 	if (pp->fbs_enabled)
 		rx_fis += qc->dev->link->pmp * AHCI_RX_FIS_SZ;
 
+#ifdef CONFIG_AHCI_IMX_PMP
+	if (qc->ap->flags & (1 << 31)) {
+		if (!(qc->ap->flags & (1 << 30))) {
+			fis = pp->rx_fis + RX_FIS_D2H_REG;
+			memcpy(rx_fis + RX_FIS_D2H_REG, fis, 0x14);
+		}
+	}
+#endif
+
 	/*
 	 * After a successful execution of an ATA PIO data-in command,
 	 * the device doesn't send D2H Reg FIS to update the TF and
