--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/base/dd.c	2026-04-08 18:20:05.748903047 +0200
+++ ./drivers/base/dd.c	2026-04-08 16:21:04.441023271 +0200
@@ -111,14 +111,6 @@
 		 */
 		mutex_unlock(&deferred_probe_mutex);
 
-		/*
-		 * Force the device to the end of the dpm_list since
-		 * the PM code assumes that the order we add things to
-		 * the list is a good order for suspend but deferred
-		 * probe makes that very unsafe.
-		 */
-		device_pm_move_to_tail(dev);
-
 		dev_dbg(dev, "Retrying from deferred list\n");
 		bus_probe_device(dev);
 		mutex_lock(&deferred_probe_mutex);
@@ -411,6 +403,14 @@
 	device_pm_check_callbacks(dev);
 
 	/*
+	 * Force the device to the end of the dpm_list since
+	 * the PM code assumes that the order we add things to
+	 * the list is a good order for suspend but deferred
+	 * probe makes that very unsafe.
+	 */
+	device_pm_move_to_tail(dev);
+
+	/*
 	 * Make sure the device is no longer in one of the deferred lists and
 	 * kick off retrying all pending devices
 	 */
