--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/clk/imx/clk-imx6ul.c	2026-04-08 18:20:05.827905249 +0200
+++ ./drivers/clk/imx/clk-imx6ul.c	2026-04-08 16:21:04.570623456 +0200
@@ -526,6 +526,12 @@
 	clk_set_rate(hws[IMX6UL_CLK_ENET2_REF]->clk, 50000000);
 	clk_set_rate(hws[IMX6UL_CLK_CSI]->clk, 24000000);
 
+        /* Set the UART parent if needed */
+        if (uart_from_osc)
+		clk_set_parent(hws[IMX6UL_CLK_UART_SEL]->clk, hws[IMX6UL_CLK_OSC]->clk);
+        else
+		clk_set_parent(hws[IMX6UL_CLK_UART_SEL]->clk, hws[IMX6UL_CLK_PLL3_80M]->clk);
+
 	if (clk_on_imx6ull())
 		clk_prepare_enable(hws[IMX6UL_CLK_AIPSTZ3]->clk);
 
