--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/crypto/caam/caamrng.c	2026-04-08 18:20:05.951908705 +0200
+++ ./drivers/crypto/caam/caamrng.c	2026-04-08 16:21:04.744877563 +0200
@@ -15,6 +15,7 @@
 #include <linux/dma-mapping.h>
 #include <linux/kernel.h>
 #include <linux/kfifo.h>
+#include <linux/busfreq-imx.h>
 
 #include "compat.h"
 
@@ -101,6 +102,7 @@
 		return -ENOMEM;
 	}
 
+	request_bus_freq(BUS_FREQ_HIGH);
 	init_completion(done);
 	err = caam_jr_enqueue(jrdev,
 			      caam_init_desc(desc, dst_dma),
@@ -110,6 +112,7 @@
 		err = 0;
 	}
 
+	release_bus_freq(BUS_FREQ_HIGH);
 	dma_unmap_single(jrdev, dst_dma, len, DMA_FROM_DEVICE);
 
 	return err ?: (ret ?: len);
