--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/crypto/caam/debugfs.c	2026-04-08 18:20:05.951908705 +0200
+++ ./drivers/crypto/caam/debugfs.c	2026-04-08 16:21:04.745162482 +0200
@@ -77,7 +77,7 @@
 	debugfs_create_file("fault_status", 0444, ctrlpriv->ctl,
 			    &perfmon->status, &caam_fops_u32_ro);
 
-	if (ctrlpriv->optee_en)
+	if (ctrlpriv->scu_en || ctrlpriv->optee_en)
 		return;
 
 	/* Internal covering keys (useful in non-secure mode only) */
