--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/crypto/mxs-dcp.c	2026-04-08 18:20:05.977369896 +0200
+++ ./drivers/crypto/mxs-dcp.c	2026-04-08 16:21:04.799906626 +0200
@@ -3,6 +3,7 @@
  * Freescale i.MX23/i.MX28 Data Co-Processor driver
  *
  * Copyright (C) 2013 Marek Vasut <marex@denx.de>
+ * Copyright 2022 NXP
  */
 
 #include <linux/dma-mapping.h>
@@ -15,6 +16,13 @@
 #include <linux/platform_device.h>
 #include <linux/stmp_device.h>
 #include <linux/clk.h>
+#include <linux/libfdt.h>
+#include <linux/of_fdt.h>
+#include <soc/fsl/dcp-blob.h>
+
+#ifdef CONFIG_PM_SLEEP
+#include <linux/freezer.h>
+#endif
 
 #include <crypto/aes.h>
 #include <crypto/sha1.h>
@@ -124,7 +132,10 @@
  * design of Linux Crypto API.
  */
 static struct dcp *global_sdcp;
-
+#ifdef CONFIG_PM_SLEEP
+static uint32_t ctrl_bak;
+static int dcp_vmi_irq_bak, dcp_irq_bak;
+#endif
 /* DCP register layout. */
 #define MXS_DCP_CTRL				0x00
 #define MXS_DCP_CTRL_GATHER_RESIDUAL_WRITES	(1 << 23)
@@ -155,6 +166,7 @@
 #define MXS_DCP_CONTROL0_HASH_TERM		(1 << 13)
 #define MXS_DCP_CONTROL0_HASH_INIT		(1 << 12)
 #define MXS_DCP_CONTROL0_PAYLOAD_KEY		(1 << 11)
+#define MXS_DCP_CONTROL0_OTP_KEY		(1 << 10)
 #define MXS_DCP_CONTROL0_CIPHER_ENCRYPT		(1 << 8)
 #define MXS_DCP_CONTROL0_CIPHER_INIT		(1 << 9)
 #define MXS_DCP_CONTROL0_ENABLE_HASH		(1 << 6)
@@ -164,6 +176,8 @@
 
 #define MXS_DCP_CONTROL1_HASH_SELECT_SHA256	(2 << 16)
 #define MXS_DCP_CONTROL1_HASH_SELECT_SHA1	(0 << 16)
+#define MXS_DCP_CONTROL1_AES_OTP_CRYPTO_KEY	(0xff << 8)
+#define MXS_DCP_CONTROL1_AES_OTP_UNIQUE_KEY	(0xfe << 8)
 #define MXS_DCP_CONTROL1_CIPHER_MODE_CBC	(1 << 4)
 #define MXS_DCP_CONTROL1_CIPHER_MODE_ECB	(0 << 4)
 #define MXS_DCP_CONTROL1_CIPHER_SELECT_AES128	(0 << 0)
@@ -224,7 +238,14 @@
 	struct dcp *sdcp = global_sdcp;
 	struct dcp_dma_desc *desc = &sdcp->coh->desc[actx->chan];
 	struct dcp_aes_req_ctx *rctx = skcipher_request_ctx(req);
-	int ret;
+	int ret, len_crypto_hdl, len_unique_hdl;
+	const struct fdt_property *prop_crypto, *prop_unique;
+	int nodeoff = fdt_node_offset_by_compatible(initial_boot_params, -1, "fsl,imx28-dcp");
+
+	if (nodeoff < 0) {
+		pr_info("node to update the SoC serial number is not found.\n");
+		return nodeoff;
+	}
 
 	key_phys = dma_map_single(sdcp->dev, sdcp->coh->aes_key,
 				  2 * AES_KEYSIZE_128, DMA_TO_DEVICE);
@@ -249,36 +270,51 @@
 		ret = -EINVAL;
 		goto aes_done_run;
 	}
-
 	/* Fill in the DMA descriptor. */
 	desc->control0 = MXS_DCP_CONTROL0_DECR_SEMAPHORE |
-		    MXS_DCP_CONTROL0_INTERRUPT |
-		    MXS_DCP_CONTROL0_ENABLE_CIPHER;
-
-	/* Payload contains the key. */
-	desc->control0 |= MXS_DCP_CONTROL0_PAYLOAD_KEY;
-
+			MXS_DCP_CONTROL0_INTERRUPT |
+			MXS_DCP_CONTROL0_ENABLE_CIPHER;
 	if (rctx->enc)
 		desc->control0 |= MXS_DCP_CONTROL0_CIPHER_ENCRYPT;
-	if (init)
-		desc->control0 |= MXS_DCP_CONTROL0_CIPHER_INIT;
 
 	desc->control1 = MXS_DCP_CONTROL1_CIPHER_SELECT_AES128;
-
 	if (rctx->ecb)
 		desc->control1 |= MXS_DCP_CONTROL1_CIPHER_MODE_ECB;
 	else
 		desc->control1 |= MXS_DCP_CONTROL1_CIPHER_MODE_CBC;
+	if (init)
+		desc->control0 |= MXS_DCP_CONTROL0_CIPHER_INIT;
 
+	/*Read device property*/
+	prop_crypto = fdt_get_property(initial_boot_params, nodeoff,
+					"otp_crypto_key", &len_crypto_hdl);
+	prop_unique = fdt_get_property(initial_boot_params, nodeoff,
+					"otp_unique_key", &len_unique_hdl);
+
+	if (len_crypto_hdl > 0 && !memcmp(sdcp->coh->aes_key,
+				prop_crypto->data, AES_KEYSIZE_128)) {
+		/* Use AES_OTP CRYPTO_KEY. */
+		desc->control0 |= MXS_DCP_CONTROL0_OTP_KEY;
+		/* Use AES_OTP crypto key. */
+		desc->control1 |= MXS_DCP_CONTROL1_AES_OTP_CRYPTO_KEY;
+		desc->payload = 0;
+	} else if (len_unique_hdl > 0 && (!memcmp(sdcp->coh->aes_key,
+			prop_unique->data, AES_KEYSIZE_128))) {
+		/* Use AES_OTP unique key. */
+		desc->control1 |= MXS_DCP_CONTROL1_AES_OTP_UNIQUE_KEY;
+		desc->payload = 0;
+	} else {
+		/* Payload contains the key. */
+		desc->control0 |= MXS_DCP_CONTROL0_PAYLOAD_KEY;
+		desc->payload = key_phys;
+	}
 	desc->next_cmd_addr = 0;
 	desc->source = src_phys;
 	desc->destination = dst_phys;
 	desc->size = actx->fill;
-	desc->payload = key_phys;
 	desc->status = 0;
 
 	ret = mxs_dcp_start_dma(actx);
-
 aes_done_run:
 	dma_unmap_single(sdcp->dev, dst_phys, DCP_BUF_SZ, DMA_FROM_DEVICE);
 err_dst:
@@ -317,6 +353,9 @@
 	int init = 0;
 	bool limit_hit = false;
 
+	if (!req->cryptlen)
+		return 0;
+
 	actx->fill = 0;
 
 	/* Copy the key from the temporary location. */
@@ -397,9 +436,15 @@
 
 	int ret;
 
+#ifdef CONFIG_PM_SLEEP
+	set_freezable();
+#endif
 	while (!kthread_should_stop()) {
 		set_current_state(TASK_INTERRUPTIBLE);
 
+#ifdef CONFIG_PM_SLEEP
+		try_to_freeze();
+#endif
 		spin_lock(&sdcp->lock[chan]);
 		backlog = crypto_get_backlog(&sdcp->queue[chan]);
 		arq = crypto_dequeue_request(&sdcp->queue[chan]);
@@ -437,6 +482,10 @@
 	skcipher_request_set_crypt(&rctx->fallback_req, req->src, req->dst,
 				   req->cryptlen, req->iv);
 
+#ifdef CONFIG_PM_SLEEP
+set_freezable();
+try_to_freeze();
+#endif
 	if (enc)
 		ret = crypto_skcipher_encrypt(&rctx->fallback_req);
 	else
@@ -693,9 +742,15 @@
 	struct crypto_async_request *arq;
 	int ret;
 
+#ifdef CONFIG_PM_SLEEP
+	set_freezable();
+#endif
 	while (!kthread_should_stop()) {
 		set_current_state(TASK_INTERRUPTIBLE);
 
+#ifdef CONFIG_PM_SLEEP
+		try_to_freeze();
+#endif
 		spin_lock(&sdcp->lock[chan]);
 		backlog = crypto_get_backlog(&sdcp->queue[chan]);
 		arq = crypto_dequeue_request(&sdcp->queue[chan]);
@@ -968,6 +1023,49 @@
 	return IRQ_HANDLED;
 }
 
+#ifdef CONFIG_PM_SLEEP
+static int mxs_dcp_resume(struct device *dev)
+{
+	struct dcp *sdcp = global_sdcp;
+	int ret;
+
+	/* Restart the DCP block */
+	ret = stmp_reset_block(sdcp->base);
+	if (ret) {
+		dev_err(dev, "Failed reset\n");
+		clk_disable_unprepare(sdcp->dcp_clk);
+		return ret;
+	}
+
+	/* Restore control register */
+	writel(ctrl_bak, sdcp->base + MXS_DCP_CTRL);
+	/* Enable all DCP DMA channels */
+	writel(MXS_DCP_CHANNELCTRL_ENABLE_CHANNEL_MASK,
+	       sdcp->base + MXS_DCP_CHANNELCTRL);
+
+	/* Re-enable DCP interrupts */
+	enable_irq(dcp_irq_bak);
+	enable_irq(dcp_vmi_irq_bak);
+
+	return 0;
+}
+
+static int mxs_dcp_suspend(struct device *dev)
+{
+	struct dcp *sdcp = global_sdcp;
+
+	/* Backup control register */
+	ctrl_bak = readl(sdcp->base + MXS_DCP_CTRL);
+	/* Temporarily disable DCP interrupts */
+	disable_irq(dcp_irq_bak);
+	disable_irq(dcp_vmi_irq_bak);
+
+	return 0;
+}
+
+SIMPLE_DEV_PM_OPS(mxs_dcp_pm_ops, mxs_dcp_suspend, mxs_dcp_resume);
+#endif
+
 static int mxs_dcp_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
@@ -987,7 +1085,10 @@
 	dcp_irq = platform_get_irq(pdev, 1);
 	if (dcp_irq < 0)
 		return dcp_irq;
-
+#ifdef CONFIG_PM_SLEEP
+	dcp_vmi_irq_bak = dcp_vmi_irq;
+	dcp_irq_bak = dcp_irq;
+#endif
 	sdcp = devm_kzalloc(dev, sizeof(*sdcp), GFP_KERNEL);
 	if (!sdcp)
 		return -ENOMEM;
@@ -1154,6 +1255,36 @@
 	return 0;
 }
 
+/*
+ * mxs_dcp_blob_to_key transfers content of hardware blob to key.
+ * Returns -EINVAL, if user wants to add hardware key other than
+ * otp_crypto_key & otp_unique_key.
+ * Returns 0, in case of success.
+ */
+int mxs_dcp_blob_to_key(struct dcp_key_payload *p)
+{
+	int len_crypto_hdl, len_unique_hdl;
+	const struct fdt_property *prop_crypto, *prop_unique;
+	int nodeoff = fdt_node_offset_by_compatible(initial_boot_params, -1, "fsl,imx28-dcp");
+
+	if (nodeoff < 0) {
+		pr_info("node to update the SoC serial number is not found.\n");
+		return nodeoff;
+	}
+
+	prop_crypto = fdt_get_property(initial_boot_params, nodeoff,
+					"otp_crypto_key", &len_crypto_hdl);
+	prop_unique = fdt_get_property(initial_boot_params, nodeoff,
+					"otp_unique_key", &len_unique_hdl);
+	memcpy(p->key, p->blob, p->blob_len);
+	p->key_len = p->blob_len;
+	if (memcmp(prop_crypto->data, p->blob, AES_KEYSIZE_128) &&
+		memcmp(prop_unique->data, p->blob, AES_KEYSIZE_128))
+		return -EINVAL;
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(mxs_dcp_blob_to_key);
 static const struct of_device_id mxs_dcp_dt_ids[] = {
 	{ .compatible = "fsl,imx23-dcp", .data = NULL, },
 	{ .compatible = "fsl,imx28-dcp", .data = NULL, },
@@ -1168,6 +1299,9 @@
 	.driver	= {
 		.name		= "mxs-dcp",
 		.of_match_table	= mxs_dcp_dt_ids,
+#ifdef CONFIG_PM_SLEEP
+		.pm = &mxs_dcp_pm_ops
+#endif
 	},
 };
 
