--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/dma/fsl-dpaa2-qdma/dpdmai.h	2026-04-08 18:20:06.000064117 +0200
+++ ./drivers/dma/fsl-dpaa2-qdma/dpdmai.h	2026-04-08 16:21:04.865264125 +0200
@@ -1,43 +1,45 @@
 /* SPDX-License-Identifier: GPL-2.0 */
-/* Copyright 2019 NXP */
+/* Copyright 2019-2021 NXP */
 
 #ifndef __FSL_DPDMAI_H
 #define __FSL_DPDMAI_H
 
 /* DPDMAI Version */
-#define DPDMAI_VER_MAJOR	2
-#define DPDMAI_VER_MINOR	2
+#define DPDMAI_VER_MAJOR	3
+#define DPDMAI_VER_MINOR	3
 
-#define DPDMAI_CMD_BASE_VERSION	0
+#define DPDMAI_CMD_BASE_VERSION	1
+#define DPDMAI_CMD_VERSION_2	2
 #define DPDMAI_CMD_ID_OFFSET	4
 
-#define DPDMAI_CMDID_FORMAT(x)	(((x) << DPDMAI_CMD_ID_OFFSET) | \
-				DPDMAI_CMD_BASE_VERSION)
+#define DPDMAI_CMD(id)		((id << DPDMAI_CMD_ID_OFFSET) | DPDMAI_CMD_BASE_VERSION)
+#define DPDMAI_CMD_V2(id)	((id << DPDMAI_CMD_ID_OFFSET) | DPDMAI_CMD_VERSION_2)
 
 /* Command IDs */
-#define DPDMAI_CMDID_CLOSE		DPDMAI_CMDID_FORMAT(0x800)
-#define DPDMAI_CMDID_OPEN               DPDMAI_CMDID_FORMAT(0x80E)
-#define DPDMAI_CMDID_CREATE             DPDMAI_CMDID_FORMAT(0x90E)
-#define DPDMAI_CMDID_DESTROY            DPDMAI_CMDID_FORMAT(0x900)
-
-#define DPDMAI_CMDID_ENABLE             DPDMAI_CMDID_FORMAT(0x002)
-#define DPDMAI_CMDID_DISABLE            DPDMAI_CMDID_FORMAT(0x003)
-#define DPDMAI_CMDID_GET_ATTR           DPDMAI_CMDID_FORMAT(0x004)
-#define DPDMAI_CMDID_RESET              DPDMAI_CMDID_FORMAT(0x005)
-#define DPDMAI_CMDID_IS_ENABLED         DPDMAI_CMDID_FORMAT(0x006)
-
-#define DPDMAI_CMDID_SET_IRQ            DPDMAI_CMDID_FORMAT(0x010)
-#define DPDMAI_CMDID_GET_IRQ            DPDMAI_CMDID_FORMAT(0x011)
-#define DPDMAI_CMDID_SET_IRQ_ENABLE     DPDMAI_CMDID_FORMAT(0x012)
-#define DPDMAI_CMDID_GET_IRQ_ENABLE     DPDMAI_CMDID_FORMAT(0x013)
-#define DPDMAI_CMDID_SET_IRQ_MASK       DPDMAI_CMDID_FORMAT(0x014)
-#define DPDMAI_CMDID_GET_IRQ_MASK       DPDMAI_CMDID_FORMAT(0x015)
-#define DPDMAI_CMDID_GET_IRQ_STATUS     DPDMAI_CMDID_FORMAT(0x016)
-#define DPDMAI_CMDID_CLEAR_IRQ_STATUS	DPDMAI_CMDID_FORMAT(0x017)
-
-#define DPDMAI_CMDID_SET_RX_QUEUE	DPDMAI_CMDID_FORMAT(0x1A0)
-#define DPDMAI_CMDID_GET_RX_QUEUE       DPDMAI_CMDID_FORMAT(0x1A1)
-#define DPDMAI_CMDID_GET_TX_QUEUE       DPDMAI_CMDID_FORMAT(0x1A2)
+#define DPDMAI_CMDID_CLOSE		DPDMAI_CMD(0x800)
+#define DPDMAI_CMDID_OPEN               DPDMAI_CMD(0x80E)
+#define DPDMAI_CMDID_CREATE             DPDMAI_CMD_V2(0x90E)
+#define DPDMAI_CMDID_DESTROY            DPDMAI_CMD(0x98E)
+#define DPDMAI_CMDID_GET_API_VERSION	DPDMAI_CMD(0xa0E)
+
+#define DPDMAI_CMDID_ENABLE             DPDMAI_CMD(0x002)
+#define DPDMAI_CMDID_DISABLE            DPDMAI_CMD(0x003)
+#define DPDMAI_CMDID_GET_ATTR           DPDMAI_CMD_V2(0x004)
+#define DPDMAI_CMDID_RESET              DPDMAI_CMD(0x005)
+#define DPDMAI_CMDID_IS_ENABLED		DPDMAI_CMD(0x006)
+
+#define DPDMAI_CMDID_SET_RX_QUEUE	DPDMAI_CMD_V2(0x1A0)
+#define DPDMAI_CMDID_GET_RX_QUEUE       DPDMAI_CMD_V2(0x1A1)
+#define DPDMAI_CMDID_GET_TX_QUEUE       DPDMAI_CMD_V2(0x1A2)
+
+/* Macros for accessing command fields smaller than 1byte */
+#define DPDMAI_MASK(field)	\
+	GENMASK(DPDMAI_##field##_SHIFT + DPDMAI_##field##_SIZE - 1,	\
+		DPDMAI_##field##_SHIFT)
+#define dpdmai_set_field(var, field, val)	\
+	((var) |= (((val) << DPDMAI_##field##_SHIFT) & DPDMAI_MASK(field)))
+#define dpdmai_get_field(var, field)	\
+	(((var) & DPDMAI_MASK(field)) >> DPDMAI_##field##_SHIFT)
 
 #define MC_CMD_HDR_TOKEN_O 32  /* Token field offset */
 #define MC_CMD_HDR_TOKEN_S 16  /* Token field size */
@@ -49,11 +51,21 @@
  * Contains initialization APIs and runtime control APIs for DPDMAI
  */
 
+/*
+ * Maximum number of Tx/Rx queues per DPDMAI object
+ */
+#define DPDMAI_MAX_QUEUE_NUM	8
+
 /**
  * Maximum number of Tx/Rx priorities per DPDMAI object
  */
 #define DPDMAI_PRIO_NUM		2
 
+/**
+ * All queues considered; see dpdmai_set_rx_queue()
+ */
+#define DPDMAI_ALL_QUEUES	((uint8_t)(-1))
+
 /* DPDMAI queue modification options */
 
 /**
@@ -66,6 +78,69 @@
  */
 #define DPDMAI_QUEUE_OPT_DEST		0x2
 
+struct dpdmai_cmd_open {
+	u32	dpdmai_id;
+};
+
+struct dpdmai_cmd_create {
+	u8	num_queues;
+	u8	priorities[2];
+};
+
+struct dpdmai_cmd_destroy {
+	u32	dpdmai_id;
+};
+
+#define DPDMAI_ENABLE_SHIFT	0
+#define DPDMAI_ENABLE_SIZE	1
+
+struct dpdmai_rsp_is_enabled {
+	/* only the LSB bit */
+	u8 en;
+};
+
+struct dpdmai_rsp_get_attributes {
+	u32 id;
+	u8 num_of_priorities;
+	u8 num_of_queues;
+};
+
+#define DPDMAI_DEST_TYPE_SHIFT	0
+#define DPDMAI_DEST_TYPE_SIZE	4
+
+struct dpdmai_cmd_set_rx_queue {
+	u32 dest_id;
+	u8 dest_priority;
+	u8 priority;
+	/* from LSB: dest_type:4 */
+	u8 dest_type;
+	u8 queue_idx;
+	u64 user_ctx;
+	u32 options;
+};
+
+struct dpdmai_cmd_get_queue {
+	u8 pad[5];
+	u8 priority;
+	u8 queue_idx;
+};
+
+struct dpdmai_rsp_get_rx_queue {
+	u32 dest_id;
+	u8 dest_priority;
+	u8 pad1;
+	/* from LSB: dest_type:4 */
+	u8 dest_type;
+	u8 pad2;
+	u64 user_ctx;
+	u32 fqid;
+};
+
+struct dpdmai_rsp_get_tx_queue {
+	u64 pad;
+	u32 fqid;
+};
+
 /**
  * struct dpdmai_cfg - Structure representing DPDMAI configuration
  * @priorities: Priorities for the DMA hardware processing; valid priorities are
@@ -73,6 +148,7 @@
  *	should be configured with 0
  */
 struct dpdmai_cfg {
+	u8 num_queues;
 	u8 priorities[DPDMAI_PRIO_NUM];
 };
 
@@ -83,17 +159,14 @@
  * @num_of_priorities: number of priorities
  */
 struct dpdmai_attr {
-	int	id;
+	int id;
 	/**
 	 * struct version - DPDMAI version
 	 * @major: DPDMAI major version
 	 * @minor: DPDMAI minor version
 	 */
-	struct {
-		u16 major;
-		u16 minor;
-	} version;
 	u8 num_of_priorities;
+	u8 num_of_queues;
 };
 
 /**
@@ -158,22 +231,33 @@
 	u32 fqid;
 };
 
+struct dpdmai_tx_queue_attr {
+	u32 fqid;
+};
+
 int dpdmai_open(struct fsl_mc_io *mc_io, u32 cmd_flags,
 		int dpdmai_id, u16 *token);
 int dpdmai_close(struct fsl_mc_io *mc_io, u32 cmd_flags, u16 token);
-int dpdmai_destroy(struct fsl_mc_io *mc_io, u32 cmd_flags, u16 token);
 int dpdmai_create(struct fsl_mc_io *mc_io, u32 cmd_flags,
-		  const struct dpdmai_cfg *cfg, u16 *token);
+		  const struct dpdmai_cfg *cfg, u16 token,
+		  u32 *dpdmai_id);
+int dpdmai_destroy(struct fsl_mc_io *mc_io, u32 cmd_flags,
+		   u32 dpdmai_id, u16 token);
 int dpdmai_enable(struct fsl_mc_io *mc_io, u32 cmd_flags, u16 token);
 int dpdmai_disable(struct fsl_mc_io *mc_io, u32 cmd_flags, u16 token);
+int dpdmai_is_enabled(struct fsl_mc_io *mc_io, u32 cmd_flags,
+		      u16 token, int *en);
 int dpdmai_reset(struct fsl_mc_io *mc_io, u32 cmd_flags, u16 token);
 int dpdmai_get_attributes(struct fsl_mc_io *mc_io, u32 cmd_flags,
 			  u16 token, struct dpdmai_attr	*attr);
-int dpdmai_set_rx_queue(struct fsl_mc_io *mc_io, u32 cmd_flags, u16 token,
-			u8 priority, const struct dpdmai_rx_queue_cfg *cfg);
-int dpdmai_get_rx_queue(struct fsl_mc_io *mc_io, u32 cmd_flags, u16 token,
-			u8 priority, struct dpdmai_rx_queue_attr *attr);
+int dpdmai_set_rx_queue(struct fsl_mc_io *mc_io, u32 cmd_flags,
+			u16 token, u8 queue_idx, u8 priority,
+			const struct dpdmai_rx_queue_cfg *cfg);
+int dpdmai_get_rx_queue(struct fsl_mc_io *mc_io, u32 cmd_flags,
+			u16 token, u8 queue_idx, u8 priority,
+			struct dpdmai_rx_queue_attr *attr);
 int dpdmai_get_tx_queue(struct fsl_mc_io *mc_io, u32 cmd_flags,
-			u16 token, u8 priority, u32 *fqid);
+			u16 token, u8 queue_idx, u8 priority,
+			struct dpdmai_tx_queue_attr *attr);
 
 #endif /* __FSL_DPDMAI_H */
