--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/firmware/arm_scmi/scmi_power_control.c	2026-04-08 18:20:06.037911102 +0200
+++ ./drivers/firmware/arm_scmi/scmi_power_control.c	2026-04-08 16:21:04.945489175 +0200
@@ -50,6 +50,7 @@
 #include <linux/reboot.h>
 #include <linux/scmi_protocol.h>
 #include <linux/slab.h>
+#include <linux/suspend.h>
 #include <linux/time64.h>
 #include <linux/timer.h>
 #include <linux/types.h>
@@ -90,6 +91,7 @@
 	struct notifier_block reboot_nb;
 
 	struct delayed_work forceful_work;
+	struct work_struct suspend_work;
 };
 
 #define userspace_nb_to_sconf(x)	\
@@ -236,6 +238,7 @@
 		sc->required_transition, timeout_ms, adj_timeout_ms);
 
 	switch (sc->required_transition) {
+	case SCMI_IMX_SYSTEM_FULL_SHUTDOWN:
 	case SCMI_SYSTEM_SHUTDOWN:
 		/*
 		 * When triggered early at boot-time the 'orderly' call will
@@ -245,10 +248,15 @@
 		 */
 		orderly_poweroff(true);
 		break;
+	case SCMI_IMX_SYSTEM_FULL_RESET:
 	case SCMI_SYSTEM_COLDRESET:
 	case SCMI_SYSTEM_WARMRESET:
 		orderly_reboot();
 		break;
+	case SCMI_IMX_SYSTEM_FULL_SUSPEND:
+	case SCMI_SYSTEM_SUSPEND:
+		schedule_work(&sc->suspend_work);
+		break;
 	default:
 		break;
 	}
@@ -277,7 +285,10 @@
 	struct scmi_system_power_state_notifier_report *er = data;
 	struct scmi_syspower_conf *sc = userspace_nb_to_sconf(nb);
 
-	if (er->system_state >= SCMI_SYSTEM_POWERUP) {
+	if ((er->system_state >= SCMI_SYSTEM_MAX &&
+	    er->system_state <= SCMI_IMX_SYSTEM_WAKE) ||
+	    er->system_state == SCMI_SYSTEM_POWERUP ||
+	    er->system_state >= SCMI_IMX_SYSTEM_FULL_WAKE) {
 		dev_err(sc->dev, "Ignoring unsupported system_state: 0x%X\n",
 			er->system_state);
 		return NOTIFY_DONE;
@@ -307,7 +318,7 @@
 	sc->required_transition = er->system_state;
 
 	/* Leaving a trace in logs of who triggered the shutdown/reboot. */
-	dev_info(sc->dev, "Serving shutdown/reboot request: %d\n",
+	dev_info(sc->dev, "Serving shutdown/reboot request: 0x%x\n",
 		 sc->required_transition);
 
 	scmi_request_graceful_transition(sc, er->timeout);
@@ -315,6 +326,16 @@
 	return NOTIFY_OK;
 }
 
+static void scmi_suspend_work_func(struct work_struct *work)
+{
+	struct scmi_syspower_conf *sc =
+		container_of(work, struct scmi_syspower_conf, suspend_work);
+
+	pm_suspend(PM_SUSPEND_MEM);
+
+	sc->state = SCMI_SYSPOWER_IDLE;
+}
+
 static int scmi_syspower_probe(struct scmi_device *sdev)
 {
 	int ret;
@@ -338,6 +359,8 @@
 	sc->userspace_nb.notifier_call = &scmi_userspace_notifier;
 	sc->dev = &sdev->dev;
 
+	INIT_WORK(&sc->suspend_work, scmi_suspend_work_func);
+
 	return handle->notify_ops->devm_event_notifier_register(sdev,
 							   SCMI_PROTOCOL_SYSTEM,
 					 SCMI_EVENT_SYSTEM_POWER_STATE_NOTIFIER,
