--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/firmware/imx/Kconfig	2026-04-08 18:20:06.045518410 +0200
+++ ./drivers/firmware/imx/Kconfig	2026-04-08 16:21:04.961211489 +0200
@@ -28,3 +28,46 @@
 	depends on IMX_SCU
 	help
 	  The System Controller Firmware (SCFW) based power domain driver.
+
+config IMX_SECO_MU
+	tristate "i.MX Security Controller (SECO) support"
+	depends on IMX_MBOX
+	default y if IMX_SCU
+
+	help
+	  It is possible to use APIs exposed by the SECO like HSM and SHE using the
+	  SAB protocol via the shared Messaging Unit. This driver exposes these
+	  interfaces via a set of file descriptors allowing to configure shared
+	  memory, send and receive messages.
+
+config IMX_SEC_ENCLAVE
+	tristate "i.MX Embedded Secure Enclave - EdgeLock Enclave Firmware driver."
+	depends on IMX_MBOX && ARCH_MXC && ARM64
+	default m if ARCH_MXC
+
+	help
+	  It is possible to use APIs exposed by the iMX Secure Enclave HW IP called:
+          - EdgeLock Enclave Firmware (for i.MX8ULP, i.MX93),
+            like base, HSM, V2X & SHE using the SAB protocol via the shared Messaging
+            Unit. This driver exposes these interfaces via a set of file descriptors
+            allowing to configure shared memory, send and receive messages.
+
+config IMX_ELE_TRNG
+	tristate "i.MX ELE True Random Number Generator"
+	default y
+	select CRYPTO_RNG
+	select HW_RANDOM
+	help
+	  This driver provides kernel-side support for the Random Number generation,
+          through NXP hardware IP for secure-enclave called EdgeLock Enclave.
+
+config IMX_ISPMU
+	tristate "IMX ISP MU Protocol driver"
+	depends on IMX_MBOX
+	help
+	  The ISP Firmware is a low-level code that manages the ISP core
+	  and runs on a dedicated Cortex-M core to provide queue and resource
+	  management of the ISP core.
+
+	  This driver manages the IPC interface between host CPU and the
+	  ISP firmware running on M0+ core.
