--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/firmware/imx/rm.c	2026-04-08 18:20:06.046098244 +0200
+++ ./drivers/firmware/imx/rm.c	2026-04-08 16:21:04.962052543 +0200
@@ -13,6 +13,11 @@
 	u16 resource;
 } __packed __aligned(4);
 
+struct imx_sc_msg_rm_pt {
+	struct imx_sc_rpc_msg hdr;
+	u8 val;
+} __packed __aligned(4);
+
 /*
  * This function check @resource is owned by current partition or not
  *
@@ -88,3 +93,124 @@
 	return 0;
 }
 EXPORT_SYMBOL(imx_sc_rm_get_resource_owner);
+
+/*
+ * This function returns the current partition number
+ *
+ * @param[in]     ipc         IPC handle
+ * @param[out]    pt          holding the partition number
+ *
+ * @return Returns 0 for success and < 0 for errors.
+ */
+int imx_sc_rm_get_partition(struct imx_sc_ipc *ipc, u8 *pt)
+{
+	struct imx_sc_msg_rm_pt msg;
+	struct imx_sc_rpc_msg *hdr = &msg.hdr;
+	int ret;
+
+	hdr->ver = IMX_SC_RPC_VERSION;
+	hdr->svc = IMX_SC_RPC_SVC_RM;
+	hdr->func = IMX_SC_RM_FUNC_GET_PARTITION;
+	hdr->size = 1;
+
+	ret = imx_scu_call_rpc(ipc, &msg, true);
+	if (ret)
+		return ret;
+
+	if (pt != NULL)
+		*pt = msg.val;
+
+	return 0;
+}
+EXPORT_SYMBOL(imx_sc_rm_get_partition);
+
+struct imx_sc_msg_rm_find_memreg {
+	struct imx_sc_rpc_msg hdr;
+	union {
+		struct {
+			u32 add_start_hi;
+			u32 add_start_lo;
+			u32 add_end_hi;
+			u32 add_end_lo;
+		} req;
+		struct {
+			u8 val;
+		} resp;
+	} data;
+}  __packed __aligned(4);
+
+int imx_sc_rm_find_memreg(struct imx_sc_ipc *ipc, u8 *mr, u64 addr_start,
+			  u64 addr_end)
+{
+	struct imx_sc_msg_rm_find_memreg msg;
+	struct imx_sc_rpc_msg *hdr = &msg.hdr;
+	int ret;
+
+	hdr->ver = IMX_SC_RPC_VERSION;
+	hdr->svc = IMX_SC_RPC_SVC_RM;
+	hdr->func = IMX_SC_RM_FUNC_FIND_MEMREG;
+	hdr->size = 5;
+
+	msg.data.req.add_start_hi = addr_start >> 32;
+	msg.data.req.add_start_lo = addr_start;
+	msg.data.req.add_end_hi = addr_end >> 32;
+	msg.data.req.add_end_lo = addr_end;
+
+	ret = imx_scu_call_rpc(ipc, &msg, true);
+	if (ret)
+		return ret;
+
+	if (mr)
+		*mr = msg.data.resp.val;
+
+	return 0;
+}
+EXPORT_SYMBOL(imx_sc_rm_find_memreg);
+
+struct imx_sc_msg_set_memreg_permissions {
+	struct imx_sc_rpc_msg hdr;
+	u8 mr;
+	u8 pt;
+	u8 perm;
+} __packed __aligned(4);
+
+int imx_sc_rm_set_memreg_permissions(struct imx_sc_ipc *ipc, u8 mr,
+				     u8 pt, u8 perm)
+{
+	struct imx_sc_msg_set_memreg_permissions msg;
+	struct imx_sc_rpc_msg *hdr = &msg.hdr;
+
+	hdr->ver = IMX_SC_RPC_VERSION;
+	hdr->svc = IMX_SC_RPC_SVC_RM;
+	hdr->func = IMX_SC_RM_FUNC_SET_MEMREG_PERMISSIONS;
+	hdr->size = 2;
+
+	msg.mr = mr;
+	msg.pt = pt;
+	msg.perm = perm;
+
+	return imx_scu_call_rpc(ipc, &msg, true);
+}
+EXPORT_SYMBOL(imx_sc_rm_set_memreg_permissions);
+
+int imx_sc_rm_get_did(struct imx_sc_ipc *ipc, u8 *did)
+{
+	struct imx_sc_rpc_msg msg;
+	struct imx_sc_rpc_msg *hdr = &msg;
+	int ret;
+
+	hdr->ver = IMX_SC_RPC_VERSION;
+	hdr->svc = IMX_SC_RPC_SVC_RM;
+	hdr->func = IMX_SC_RM_FUNC_GET_DID;
+	hdr->size = 1;
+
+	ret = imx_scu_call_rpc(ipc, &msg, true);
+	if (ret < 0)
+		return ret;
+
+	if (did)
+		*did = msg.func;
+
+	return 0;
+}
+EXPORT_SYMBOL(imx_sc_rm_get_did);
