--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/gpio/Kconfig	2026-04-08 18:20:06.054318906 +0200
+++ ./drivers/gpio/Kconfig	2026-04-08 16:21:04.984173416 +0200
@@ -354,6 +354,12 @@
 
 	  If unsure, say N.
 
+config GPIO_IMX_RPMSG
+	tristate "NXP i.MX7ULP RPMSG GPIO support"
+	depends on ARCH_MXC && RPMSG && GPIOLIB
+	help
+	  This driver support i.MX7ULP RPMSG virtual GPIOs.
+
 config GPIO_IMX_SCU
        def_bool y
        depends on IMX_SCU
@@ -691,8 +697,8 @@
 	  Say yes here to support UniPhier GPIOs.
 
 config GPIO_VF610
-	bool "VF610 GPIO support"
-	default y if SOC_VF610
+	tristate "Vybrid vf610 GPIO support"
+	default y
 	depends on ARCH_MXC
 	select GPIOLIB_IRQCHIP
 	help
@@ -1174,6 +1180,13 @@
 	  This option enables support for on-chip GPIO found
 	  on Analog Devices ADP5520 PMICs.
 
+config GPIO_ADP5585
+	tristate "GPIO Support for ADP5585"
+	depends on MFD_ADP5585
+	help
+	  This option enables support for on-chip GPIO found
+	  on Analog Devices ADP5585.
+
 config GPIO_ALTERA_A10SR
 	tristate "Altera Arria10 System Resource GPIO"
 	depends on MFD_ALTERA_A10SR
@@ -1790,6 +1803,16 @@
 	  Say yes here to enable a driver for GPIO multiplexers based on latches
 	  connected to other GPIOs.
 
+config GPIO_QIXIS_CPLD
+	tristate "NXP QIXIS CPLD GPIO support"
+	depends on MFD_SIMPLE_MFD_I2C || COMPILE_TEST
+	select GPIO_REGMAP
+	help
+	  This enables support for the GPIOs found on the QIXIS CPLD which is
+	  integrated on some NXP Layerscape boards such as LX2160ARDB. The
+	  driver can be build as a module as well which will be names
+	  gpio-qixis-cpld.
+
 config GPIO_MOCKUP
 	tristate "GPIO Testing Driver"
 	select IRQ_SIM
