--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/gpu/arm/midgard/mmu/Kbuild	1970-01-01 01:00:00.000000000 +0100
+++ ./drivers/gpu/arm/midgard/mmu/Kbuild	2026-04-08 16:21:05.068144372 +0200
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note
+#
+# (C) COPYRIGHT 2021-2024 ARM Limited. All rights reserved.
+#
+# This program is free software and is provided to you under the terms of the
+# GNU General Public License version 2 as published by the Free Software
+# Foundation, and any use by you of this program is subject to the terms
+# of such GNU license.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, you can access it online at
+# http://www.gnu.org/licenses/gpl-2.0.html.
+#
+#
+
+mali_kbase-y += \
+    mmu/mali_kbase_mmu.o \
+    mmu/mali_kbase_mmu_hw_direct.o \
+    mmu/mali_kbase_mmu_faults_decoder_luts.o \
+    mmu/mali_kbase_mmu_faults_decoder.o \
+    mmu/mali_kbase_mmu_mode_aarch64.o
+
+ifeq ($(CONFIG_MALI_CSF_SUPPORT),y)
+    mali_kbase-y += mmu/backend/mali_kbase_mmu_csf.o \
+    mmu/backend/mali_kbase_mmu_faults_decoder_luts_csf.o
+else
+    mali_kbase-y += mmu/backend/mali_kbase_mmu_jm.o \
+	mmu/backend/mali_kbase_mmu_faults_decoder_luts_jm.o
+
+endif
