--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/gpu/drm/bridge/lontium-lt9611uxc.c	2026-04-08 18:20:07.005359245 +0200
+++ ./drivers/gpu/drm/bridge/lontium-lt9611uxc.c	2026-04-08 16:21:05.801450204 +0200
@@ -96,9 +96,13 @@
  * Enumerate them here to check whether the mode is supported.
  */
 static struct lt9611uxc_mode lt9611uxc_modes[] = {
+	{ 4096, 2160, 30 },
+	{ 3840, 2160, 30 },
+	{ 3840, 2160, 25 },
 	{ 1920, 1080, 60 },
 	{ 1920, 1080, 30 },
 	{ 1920, 1080, 25 },
+	{ 1920, 1080, 24 },
 	{ 1366, 768, 60 },
 	{ 1360, 768, 60 },
 	{ 1280, 1024, 60 },
@@ -377,6 +381,23 @@
 			    const struct drm_display_mode *mode)
 {
 	struct lt9611uxc_mode *lt9611uxc_mode;
+	struct drm_bridge *pre_bridge;
+
+	 /*
+	  * i.MX95 MIPI DSI cannot support both 1920x1080p60 (requiring approximately
+	  * 148.444MHz pixel clock) and 4Kp30 (requiring exactly 297MHz pixel clock)
+	  * using a single video PLL configuration.
+	  *
+	  * To ensure correct pixel clock operation for all modes, remove the
+	  * video modes with its 148.5MHz pixel clock.
+	  */
+	if (mode->clock == 148500) {
+		pre_bridge = drm_bridge_get_prev_bridge(bridge);
+		if (pre_bridge)
+			if (of_device_is_compatible(pre_bridge->of_node,
+						    "nxp,imx95-mipi-dsi"))
+				return MODE_BAD;
+	}
 
 	lt9611uxc_mode = lt9611uxc_find_mode(mode);
 
