--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/gpu/drm/bridge/nwl-dsi.h	2026-04-08 18:20:07.005938078 +0200
+++ ./drivers/gpu/drm/bridge/nwl-dsi.h	2026-04-08 16:21:05.802269556 +0200
@@ -69,9 +69,7 @@
 #define NWL_DSI_RX_FIFO_UDFLW			BIT(6)
 #define NWL_DSI_RX_PKT_HDR_RCVD			BIT(7)
 #define NWL_DSI_RX_PKT_PAYLOAD_DATA_RCVD	BIT(8)
-#define NWL_DSI_BTA_TIMEOUT			BIT(29)
 #define NWL_DSI_LP_RX_TIMEOUT			BIT(30)
-#define NWL_DSI_HS_TX_TIMEOUT			BIT(31)
 
 #define NWL_DSI_IRQ_STATUS2			0x2a4
 #define NWL_DSI_SINGLE_BIT_ECC_ERR		BIT(0)
@@ -88,9 +86,7 @@
 #define NWL_DSI_RX_FIFO_UDFLW_MASK		BIT(6)
 #define NWL_DSI_RX_PKT_HDR_RCVD_MASK		BIT(7)
 #define NWL_DSI_RX_PKT_PAYLOAD_DATA_RCVD_MASK	BIT(8)
-#define NWL_DSI_BTA_TIMEOUT_MASK		BIT(29)
 #define NWL_DSI_LP_RX_TIMEOUT_MASK		BIT(30)
-#define NWL_DSI_HS_TX_TIMEOUT_MASK		BIT(31)
 
 #define NWL_DSI_IRQ_MASK2			0x2ac
 #define NWL_DSI_SINGLE_BIT_ECC_ERR_MASK		BIT(0)
