--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/gpu/drm/fsl-dcu/fsl_dcu_drm_drv.c	2026-04-08 18:20:07.034550264 +0200
+++ ./drivers/gpu/drm/fsl-dcu/fsl_dcu_drm_drv.c	2026-04-08 16:21:05.863225694 +0200
@@ -50,6 +50,23 @@
 	.volatile_reg = fsl_dcu_drm_is_volatile_reg,
 };
 
+static int fsl_dcu_scfg_config_ls1021a(void)
+{
+	struct regmap *scfg;
+
+	scfg = syscon_regmap_lookup_by_compatible("fsl,ls1021a-scfg");
+	if (IS_ERR(scfg))
+		return PTR_ERR(scfg);
+
+	/*
+	 * For simplicity, enable the PIXCLK unconditionally. It might
+	 * be possible to disable the clock in PM or on unload as a future
+	 * improvement.
+	 */
+	return regmap_update_bits(scfg, SCFG_PIXCLKCR, SCFG_PIXCLKCR_PXCEN,
+				  SCFG_PIXCLKCR_PXCEN);
+}
+
 static void fsl_dcu_irq_reset(struct drm_device *dev)
 {
 	struct fsl_dcu_drm_device *fsl_dev = dev->dev_private;
@@ -109,18 +126,12 @@
 		return ret;
 	}
 
-	scfg = syscon_regmap_lookup_by_compatible("fsl,ls1021a-scfg");
-	if (PTR_ERR(scfg) != -ENODEV) {
-		/*
-		 * For simplicity, enable the PIXCLK unconditionally,
-		 * resulting in increased power consumption. Disabling
-		 * the clock in PM or on unload could be implemented as
-		 * a future improvement.
-		 */
-		ret = regmap_update_bits(scfg, SCFG_PIXCLKCR, SCFG_PIXCLKCR_PXCEN,
-					 SCFG_PIXCLKCR_PXCEN);
-		if (ret < 0)
-			return dev_err_probe(dev->dev, ret, "failed to enable pixclk\n");
+	if (of_device_is_compatible(fsl_dev->np, "fsl,ls1021a-dcu")) {
+		ret = fsl_dcu_scfg_config_ls1021a();
+		if (ret < 0) {
+			dev_err(dev->dev, "failed to enable pixclk\n");
+			goto done_vblank;
+		}
 	}
 
 	ret = drm_vblank_init(dev, dev->mode_config.num_crtc);
