--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/gpu/drm/mxsfb/mxsfb_kms.c	2026-04-08 18:20:07.145702117 +0200
+++ ./drivers/gpu/drm/mxsfb/mxsfb_kms.c	2026-04-08 16:21:06.017861027 +0200
@@ -8,6 +8,7 @@
  * Copyright (C) 2008 Embedded Alley Solutions, Inc All Rights Reserved.
  */
 
+#include <linux/busfreq-imx.h>
 #include <linux/clk.h>
 #include <linux/io.h>
 #include <linux/iopoll.h>
@@ -54,9 +55,19 @@
 	struct drm_device *drm = mxsfb->drm;
 	const u32 format = mxsfb->crtc.primary->state->fb->format->format;
 	u32 ctrl, ctrl1;
+	bool bgr_format = true;
 
-	DRM_DEV_DEBUG_DRIVER(drm->dev, "Using bus_format: 0x%08X\n",
-			     bus_format);
+	/* Do some clean-up that we might have from a previous mode */
+	ctrl = CTRL_SHIFT_DIR(1);
+	ctrl |= CTRL_SHIFT_NUM(0xff);
+	ctrl |= CTRL_SET_WORD_LENGTH(0xff);
+	ctrl |= CTRL_DF16;
+	ctrl |= CTRL_SET_BUS_WIDTH(0xff);
+	writel(ctrl, mxsfb->base + LCDC_CTRL + REG_CLR);
+	if (mxsfb->devdata->has_ctrl2)
+		writel(CTRL2_ODD_LINE_PATTERN(CTRL2_LINE_PATTERN_CLR) |
+		       CTRL2_EVEN_LINE_PATTERN(CTRL2_LINE_PATTERN_CLR),
+		       mxsfb->base + LCDC_V4_CTRL2 + REG_CLR);
 
 	ctrl = CTRL_BYPASS_COUNT | CTRL_MASTER;
 
@@ -64,20 +75,70 @@
 	ctrl1 = readl(mxsfb->base + LCDC_CTRL1);
 	ctrl1 &= CTRL1_CUR_FRAME_DONE_IRQ_EN | CTRL1_CUR_FRAME_DONE_IRQ;
 
+	DRM_DEV_DEBUG_DRIVER(drm->dev, "Setting up %p4cc mode\n", &format);
+
 	switch (format) {
-	case DRM_FORMAT_RGB565:
-		dev_dbg(drm->dev, "Setting up RGB565 mode\n");
-		ctrl |= CTRL_WORD_LENGTH_16;
+	case DRM_FORMAT_BGR565: /* BG16 */
+		if (!mxsfb->devdata->has_ctrl2)
+			goto err;
+		writel(CTRL2_ODD_LINE_PATTERN(CTRL2_LINE_PATTERN_BGR) |
+			CTRL2_EVEN_LINE_PATTERN(CTRL2_LINE_PATTERN_BGR),
+			mxsfb->base + LCDC_V4_CTRL2 + REG_SET);
+		fallthrough;
+	case DRM_FORMAT_RGB565: /* RG16 */
+		ctrl |= CTRL_SET_WORD_LENGTH(0);
+		ctrl &= ~CTRL_DF16;
 		ctrl1 |= CTRL1_SET_BYTE_PACKAGING(0xf);
 		break;
-	case DRM_FORMAT_XRGB8888:
-		dev_dbg(drm->dev, "Setting up XRGB8888 mode\n");
-		ctrl |= CTRL_WORD_LENGTH_24;
+	case DRM_FORMAT_XBGR1555: /* XB15 */
+		fallthrough;
+	case DRM_FORMAT_ABGR1555: /* AB15 */
+		if (!mxsfb->devdata->has_ctrl2)
+			goto err;
+		writel(CTRL2_ODD_LINE_PATTERN(CTRL2_LINE_PATTERN_BGR) |
+		       CTRL2_EVEN_LINE_PATTERN(CTRL2_LINE_PATTERN_BGR),
+		       mxsfb->base + LCDC_V4_CTRL2 + REG_SET);
+		fallthrough;
+	case DRM_FORMAT_XRGB1555: /* XR15 */
+		fallthrough;
+	case DRM_FORMAT_ARGB1555: /* AR15 */
+		ctrl |= CTRL_SET_WORD_LENGTH(0);
+		ctrl |= CTRL_DF16;
+		ctrl1 |= CTRL1_SET_BYTE_PACKAGING(0xf);
+		break;
+	case DRM_FORMAT_RGBX8888: /* RX24 */
+		fallthrough;
+	case DRM_FORMAT_RGBA8888: /* RA24 */
+		/* RGBX - > 0RGB */
+		ctrl |= CTRL_SHIFT_DIR(1);
+		ctrl |= CTRL_SHIFT_NUM(8);
+		bgr_format = false;
+		fallthrough;
+	case DRM_FORMAT_XBGR8888: /* XB24 */
+		fallthrough;
+	case DRM_FORMAT_ABGR8888: /* AB24 */
+		if (bgr_format) {
+			if (!mxsfb->devdata->has_ctrl2)
+				goto err;
+			writel(CTRL2_ODD_LINE_PATTERN(CTRL2_LINE_PATTERN_BGR) |
+			       CTRL2_EVEN_LINE_PATTERN(CTRL2_LINE_PATTERN_BGR),
+			       mxsfb->base + LCDC_V4_CTRL2 + REG_SET);
+		}
+		fallthrough;
+	case DRM_FORMAT_XRGB8888: /* XR24 */
+		fallthrough;
+	case DRM_FORMAT_ARGB8888: /* AR24 */
+		ctrl |= CTRL_SET_WORD_LENGTH(3);
 		/* Do not use packed pixels = one pixel per word instead. */
 		ctrl1 |= CTRL1_SET_BYTE_PACKAGING(0x7);
 		break;
+	default:
+		goto err;
 	}
 
+	DRM_DEV_DEBUG_DRIVER(drm->dev, "Using bus_format: 0x%08X\n",
+			     bus_format);
+
 	switch (bus_format) {
 	case MEDIA_BUS_FMT_RGB565_1X16:
 		ctrl |= CTRL_BUS_WIDTH_16;
@@ -94,7 +155,11 @@
 	}
 
 	writel(ctrl1, mxsfb->base + LCDC_CTRL1);
-	writel(ctrl, mxsfb->base + LCDC_CTRL);
+	writel(ctrl, mxsfb->base + LCDC_CTRL + REG_SET);
+
+	return;
+err:
+	dev_err(drm->dev, "Unhandled pixel format: %p4cc\n", &format);
 }
 
 static void mxsfb_set_mode(struct mxsfb_drm_private *mxsfb, u32 bus_flags)
@@ -164,6 +229,9 @@
 		writel(reg, mxsfb->base + LCDC_V4_CTRL2);
 	}
 
+	/* De-assert LCD Reset bit */
+	writel(CTRL_LCD_RESET, mxsfb->base + LCDC_CTRL1 + REG_SET);
+
 	/* If it was disabled, re-enable the mode again */
 	writel(CTRL_DOTCLK_MODE, mxsfb->base + LCDC_CTRL + REG_SET);
 
@@ -208,6 +276,11 @@
 {
 	u32 reg;
 
+	writel(CTRL_RUN, mxsfb->base + LCDC_CTRL + REG_CLR);
+
+	/* Assert LCD Reset bit */
+	writel(CTRL_LCD_RESET, mxsfb->base + LCDC_CTRL1 + REG_CLR);
+
 	/*
 	 * Even if we disable the controller here, it will still continue
 	 * until its FIFOs are running out of data
@@ -308,6 +381,28 @@
 	mxsfb_set_mode(mxsfb, bus_flags);
 }
 
+static enum drm_mode_status mxsfb_crtc_mode_valid(struct drm_crtc *crtc,
+					   const struct drm_display_mode *mode)
+{
+	struct mxsfb_drm_private *mxsfb = to_mxsfb_drm_private(crtc->dev);
+	u32 bpp;
+	u64 bw;
+
+	if (!crtc->primary->state->fb)
+		bpp = 32;
+	else
+		bpp = crtc->primary->state->fb->format->cpp[0] * 8;
+
+	bw = (u64)mode->clock * 1000;
+	bw = bw * mode->hdisplay * mode->vdisplay * (bpp / 8);
+	bw = div_u64(bw, mode->htotal * mode->vtotal);
+
+	if (mxsfb->max_bw && bw > mxsfb->max_bw)
+		return MODE_BAD;
+
+	return MODE_OK;
+}
+
 static int mxsfb_crtc_atomic_check(struct drm_crtc *crtc,
 				   struct drm_atomic_state *state)
 {
@@ -354,6 +449,7 @@
 	u32 bus_format = 0;
 	dma_addr_t dma_addr;
 
+	request_bus_freq(BUS_FREQ_HIGH);
 	pm_runtime_get_sync(drm->dev);
 	mxsfb_enable_axi_clk(mxsfb);
 
@@ -418,6 +514,7 @@
 
 	mxsfb_disable_axi_clk(mxsfb);
 	pm_runtime_put_sync(drm->dev);
+	release_bus_freq(BUS_FREQ_HIGH);
 }
 
 static int mxsfb_crtc_enable_vblank(struct drm_crtc *crtc)
@@ -476,6 +573,7 @@
 }
 
 static const struct drm_crtc_helper_funcs mxsfb_crtc_helper_funcs = {
+	.mode_valid = mxsfb_crtc_mode_valid,
 	.atomic_check = mxsfb_crtc_atomic_check,
 	.atomic_flush = mxsfb_crtc_atomic_flush,
 	.atomic_enable = mxsfb_crtc_atomic_enable,
@@ -525,27 +623,120 @@
 									     plane);
 	struct mxsfb_drm_private *mxsfb = to_mxsfb_drm_private(plane->dev);
 	struct drm_crtc_state *crtc_state;
+	struct drm_framebuffer *fb = plane_state->fb;
+	struct drm_plane_state *old_state = plane->state;
+	struct drm_framebuffer *old_fb = old_state->fb;
 
 	crtc_state = drm_atomic_get_new_crtc_state(state,
 						   &mxsfb->crtc);
 
+	if (plane->type == DRM_PLANE_TYPE_PRIMARY && old_fb && fb &&
+	    old_fb->pitches[0] != fb->pitches[0])
+		crtc_state->mode_changed = true;
+
 	return drm_atomic_helper_check_plane_state(plane_state, crtc_state,
 						   DRM_PLANE_NO_SCALING,
 						   DRM_PLANE_NO_SCALING,
 						   false, true);
 }
 
+static void mxsfb_set_fb_hcrop(struct mxsfb_drm_private *mxsfb, u32 src_w, u32 fb_w)
+{
+	u32 mask_cnt;
+	u32 vdctrl3, vdctrl4, transfer_count;
+
+	if (src_w == fb_w) {
+		writel(0x0, mxsfb->base + HW_EPDC_PIGEON_12_0);
+		writel(0x0, mxsfb->base + HW_EPDC_PIGEON_12_1);
+
+		return;
+	}
+
+	transfer_count = readl(mxsfb->base + LCDC_V4_TRANSFER_COUNT);
+	transfer_count &= ~TRANSFER_COUNT_SET_HCOUNT(0xffff);
+	transfer_count |= TRANSFER_COUNT_SET_HCOUNT(fb_w);
+	writel(transfer_count, mxsfb->base + LCDC_V4_TRANSFER_COUNT);
+
+	vdctrl4 = readl(mxsfb->base + LCDC_VDCTRL4);
+	vdctrl4 &= ~SET_DOTCLK_H_VALID_DATA_CNT(0x3ffff);
+	vdctrl4 |= SET_DOTCLK_H_VALID_DATA_CNT(fb_w);
+	writel(vdctrl4, mxsfb->base + LCDC_VDCTRL4);
+
+	/* configure related pigeon registers */
+	vdctrl3  = readl(mxsfb->base + LCDC_VDCTRL3);
+	mask_cnt = GET_HOR_WAIT_CNT(vdctrl3) - 5;
+
+	writel(PIGEON_12_0_SET_STATE_MASK(0x24) |
+	       PIGEON_12_0_SET_MASK_CNT(mask_cnt) |
+	       PIGEON_12_0_SET_MASK_CNT_SEL(0x6) |
+	       PIGEON_12_0_POL_ACTIVE_LOW |
+	       PIGEON_12_0_EN,
+	       mxsfb->base + HW_EPDC_PIGEON_12_0);
+
+	writel(PIGEON_12_1_SET_CLR_CNT(src_w) |
+	       PIGEON_12_1_SET_SET_CNT(0x0),
+	       mxsfb->base + HW_EPDC_PIGEON_12_1);
+
+	writel(0, mxsfb->base + HW_EPDC_PIGEON_12_2);
+}
+
 static void mxsfb_plane_primary_atomic_update(struct drm_plane *plane,
 					      struct drm_atomic_state *state)
 {
 	struct mxsfb_drm_private *mxsfb = to_mxsfb_drm_private(plane->dev);
 	struct drm_plane_state *new_pstate = drm_atomic_get_new_plane_state(state,
 									    plane);
+	struct drm_plane_state *old_pstate = drm_atomic_get_old_plane_state(state,
+									    plane);
+	struct drm_framebuffer *fb = new_pstate->fb;
+	struct drm_framebuffer *old_fb = old_pstate->fb;
 	dma_addr_t dma_addr;
+	u32 src_off, src_w, stride, cpp = 0;
 
 	dma_addr = drm_fb_dma_get_gem_addr(new_pstate->fb, new_pstate, 0);
-	if (dma_addr)
-		writel(dma_addr, mxsfb->base + mxsfb->devdata->next_buf);
+	if (!dma_addr)
+		return;
+
+	/* We will use the 'has_ctrl2' to assume IP version larger than 4 */
+	if (mxsfb->devdata->has_ctrl2) {
+		cpp = fb->format->cpp[0];
+		src_off = (new_pstate->src_y >> 16) * fb->pitches[0] +
+			  (new_pstate->src_x >> 16) * cpp;
+		dma_addr += fb->offsets[0] + src_off;
+	}
+
+	writel(dma_addr, mxsfb->base + mxsfb->devdata->next_buf);
+
+	if (mxsfb->devdata->has_ctrl2 &&
+	    unlikely(drm_atomic_crtc_needs_modeset(new_pstate->crtc->state))) {
+		stride = DIV_ROUND_UP(fb->pitches[0], cpp);
+		src_w = new_pstate->src_w >> 16;
+		mxsfb_set_fb_hcrop(mxsfb, src_w, stride);
+	}
+
+	/* Update format if changed */
+	if (old_fb && old_fb->format->format != fb->format->format) {
+		u32 bus_format = 0;
+		struct drm_bridge_state *bridge_state;
+
+		/* If there is a bridge attached to the LCDIF, use its bus format */
+		if (mxsfb->bridge) {
+			bridge_state =
+				drm_atomic_get_new_bridge_state(state,
+								mxsfb->bridge);
+			bus_format = bridge_state->input_bus_cfg.format;
+		}
+
+		/* If there is no bridge, use bus format from connector */
+		if (!bus_format && mxsfb->connector->display_info.num_bus_formats)
+			bus_format = mxsfb->connector->display_info.bus_formats[0];
+
+		/* If all else fails, default to RGB888_1X24 */
+		if (!bus_format)
+			bus_format = MEDIA_BUS_FMT_RGB888_1X24;
+
+		mxsfb_set_formats(mxsfb, bus_format);
+	}
 }
 
 static void mxsfb_plane_overlay_atomic_update(struct drm_plane *plane,
@@ -647,11 +838,26 @@
 	.atomic_destroy_state	= drm_atomic_helper_plane_destroy_state,
 };
 
-static const uint32_t mxsfb_primary_plane_formats[] = {
+static const uint32_t mxsfb_primary_plane_formats_v3[] = {
 	DRM_FORMAT_RGB565,
 	DRM_FORMAT_XRGB8888,
 };
 
+static const uint32_t mxsfb_primary_plane_formats_v4[] = {
+	DRM_FORMAT_RGB565,
+	DRM_FORMAT_BGR565,
+	DRM_FORMAT_XRGB8888,
+	DRM_FORMAT_ARGB8888,
+	DRM_FORMAT_XBGR8888,
+	DRM_FORMAT_ABGR8888,
+	DRM_FORMAT_RGBX8888,
+	DRM_FORMAT_RGBA8888,
+	DRM_FORMAT_ARGB1555,
+	DRM_FORMAT_XRGB1555,
+	DRM_FORMAT_ABGR1555,
+	DRM_FORMAT_XBGR1555,
+};
+
 static const uint32_t mxsfb_overlay_plane_formats[] = {
 	DRM_FORMAT_XRGB4444,
 	DRM_FORMAT_ARGB4444,
@@ -675,14 +881,23 @@
 {
 	struct drm_encoder *encoder = &mxsfb->encoder;
 	struct drm_crtc *crtc = &mxsfb->crtc;
+	uint32_t const *plane_formats;
+	unsigned int format_count;
 	int ret;
 
+	if (!mxsfb->devdata->has_ctrl2) {
+		plane_formats = &mxsfb_primary_plane_formats_v3[0];
+		format_count = ARRAY_SIZE(mxsfb_primary_plane_formats_v3);
+	} else {
+		plane_formats = &mxsfb_primary_plane_formats_v4[0];
+		format_count = ARRAY_SIZE(mxsfb_primary_plane_formats_v4);
+	}
+
 	drm_plane_helper_add(&mxsfb->planes.primary,
 			     &mxsfb_plane_primary_helper_funcs);
 	ret = drm_universal_plane_init(mxsfb->drm, &mxsfb->planes.primary, 1,
 				       &mxsfb_plane_funcs,
-				       mxsfb_primary_plane_formats,
-				       ARRAY_SIZE(mxsfb_primary_plane_formats),
+				       plane_formats, format_count,
 				       mxsfb_modifiers, DRM_PLANE_TYPE_PRIMARY,
 				       NULL);
 	if (ret)
