--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/hwmon/Kconfig	2026-04-08 18:20:07.323946939 +0200
+++ ./drivers/hwmon/Kconfig	2026-04-08 16:21:06.379885161 +0200
@@ -1015,6 +1015,16 @@
 	  If you say yes here you get support for the temperature
 	  sensor inside your CPU.
 
+config SENSORS_FP9931
+	tristate "Fitipower FP9931 panel temperature thermistor"
+	depends on MFD_FP9931
+	help
+	  If you say yes here you get support for FP9931 PMIC temperature
+	  sensor.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called fp9931_hwmon.
+
 config SENSORS_MAX1111
 	tristate "Maxim MAX1111 Serial 8-bit ADC chip and compatibles"
 	depends on SPI_MASTER
@@ -1069,6 +1079,15 @@
 	  This driver can also be built as a module. If so, the module
 	  will be called max1668.
 
+config SENSORS_MAX17135
+	tristate "Maxim MAX17135 EPD temperature sensor"
+	depends on I2C
+	help
+	  If you say yes here you get support for MAX17135 PMIC sensor.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called max17135_sensor.
+
 config SENSORS_MAX197
 	tristate "Maxim MAX197 and compatibles"
 	help
@@ -2371,6 +2390,13 @@
 	  sensors monitor various telemetry data of different components on the
 	  card, e.g. board temperature, FPGA core temperature/voltage/current.
 
+config SERIES_DIODE_RESISTANCE_PARAM
+	int "Adjust Series Diode Resistance"
+	default 3
+	help
+	  This is to adjust Series Diode resistance while reading
+	  remote temperature
+
 if ACPI
 
 comment "ACPI drivers"
