--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/iio/accel/fxls8962af-spi.c	2026-04-08 18:20:07.387948722 +0200
+++ ./drivers/iio/accel/fxls8962af-spi.c	2026-04-08 16:21:06.445095991 +0200
@@ -41,12 +41,18 @@
 };
 MODULE_DEVICE_TABLE(spi, fxls8962af_spi_id_table);
 
+static void fxls8962af_shutdown(struct spi_device *spi)
+{
+	fxls8962af_core_shutdown(&spi->dev);
+};
+
 static struct spi_driver fxls8962af_driver = {
 	.driver = {
 		   .name = "fxls8962af_spi",
 		   .pm = pm_ptr(&fxls8962af_pm_ops),
 		   .of_match_table = fxls8962af_spi_of_match,
 		   },
+	.shutdown = fxls8962af_shutdown,
 	.probe = fxls8962af_probe,
 	.id_table = fxls8962af_spi_id_table,
 };
