--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/iio/temperature/p3t/Kconfig	1970-01-01 01:00:00.000000000 +0100
+++ ./drivers/iio/temperature/p3t/Kconfig	2026-04-08 16:21:06.508079761 +0200
@@ -0,0 +1,27 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+config IIO_P3T1085
+	tristate
+	depends on (I2C || I3C)
+
+config IIO_P3T1085_I2C
+	tristate "NXP P3T1085 temprature sensor I2C driver"
+	select IIO_P3T1085
+	select REGMAP_I2C
+	help
+	  Say yes here to build support for NXP P3T1085 I2C temperature
+	  sensor.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called p3t1085_i2c
+
+config IIO_P3T1085_I3C
+	tristate "NXP P3T1085 temprature sensor I3C driver"
+	select IIO_P3T1085
+	select REGMAP_I3C
+	help
+	  Say yes here to build support for NXP P3T1085 I3C temperature
+	  sensor.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called p3t1085_i3c
