--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/input/keyboard/snvs_pwrkey.c	2026-04-08 18:20:07.527952623 +0200
+++ ./drivers/input/keyboard/snvs_pwrkey.c	2026-04-08 16:21:06.613637690 +0200
@@ -20,7 +20,6 @@
 #include <linux/mfd/syscon.h>
 #include <linux/regmap.h>
 
-#define SNVS_HPVIDR1_REG	0xBF8
 #define SNVS_LPSR_REG		0x4C	/* LP Status Register */
 #define SNVS_LPCR_REG		0x38	/* LP Control Register */
 #define SNVS_HPSR_REG		0x14
@@ -37,9 +36,11 @@
 	int keycode;
 	int keystate;  /* 1:pressed */
 	int wakeup;
+	bool suspended;
+	struct clk *clk;
 	struct timer_list check_timer;
 	struct input_dev *input;
-	u8 minor_rev;
+	bool  emulate_press;
 };
 
 static void imx_imx_snvs_check_for_events(struct timer_list *t)
@@ -48,7 +49,22 @@
 	struct input_dev *input = pdata->input;
 	u32 state;
 
+	if (pdata->clk) {
+		if (pdata->suspended)
+			clk_prepare_enable(pdata->clk);
+		else
+			clk_enable(pdata->clk);
+	}
+
 	regmap_read(pdata->snvs, SNVS_HPSR_REG, &state);
+
+	if (pdata->clk) {
+		if (pdata->suspended)
+			clk_disable_unprepare(pdata->clk);
+		else
+			clk_disable(pdata->clk);
+	}
+
 	state = state & SNVS_HPSR_BTN ? 1 : 0;
 
 	/* only report new event if status changed */
@@ -75,9 +91,22 @@
 
 	pm_wakeup_event(input->dev.parent, 0);
 
+	if (pdata->clk)
+		clk_enable(pdata->clk);
+
+	/*
+	 * Directly report press event in interrupt handler after suspend
+	 * to ensure no press event miss.
+	 */
+	if (pdata->suspended) {
+		pdata->keystate = 1;
+		input_event(input, EV_KEY, pdata->keycode, 1);
+		input_sync(input);
+	}
+
 	regmap_read(pdata->snvs, SNVS_LPSR_REG, &lp_status);
 	if (lp_status & SNVS_LPSR_SPO) {
-		if (pdata->minor_rev == 0) {
+		if (pdata->emulate_press) {
 			/*
 			 * The first generation i.MX6 SoCs only sends an
 			 * interrupt on button release. To mimic power-key
@@ -97,6 +126,9 @@
 	/* clear SPO status */
 	regmap_write(pdata->snvs, SNVS_LPSR_REG, SNVS_LPSR_SPO);
 
+	if (pdata->clk)
+		clk_disable(pdata->clk);
+
 	return IRQ_HANDLED;
 }
 
@@ -119,7 +151,6 @@
 	struct device_node *np;
 	struct clk *clk;
 	int error;
-	u32 vid;
 
 	/* Get SNVS register Page */
 	np = pdev->dev.of_node;
@@ -169,8 +200,19 @@
 	if (pdata->irq < 0)
 		return -EINVAL;
 
-	regmap_read(pdata->snvs, SNVS_HPVIDR1_REG, &vid);
-	pdata->minor_rev = vid & 0xff;
+	pdata->emulate_press = of_property_read_bool(np, "emulate-press");
+
+	pdata->clk = devm_clk_get(&pdev->dev, "snvs");
+	if (IS_ERR(pdata->clk)) {
+		pdata->clk = NULL;
+	} else {
+		error = clk_prepare_enable(pdata->clk);
+		if (error) {
+			dev_err(&pdev->dev,
+				"Could not enable the snvs clock\n");
+			return error;
+		}
+	}
 
 	regmap_update_bits(pdata->snvs, SNVS_LPCR_REG, SNVS_LPCR_DEP_EN, SNVS_LPCR_DEP_EN);
 
@@ -182,7 +224,8 @@
 	input = devm_input_allocate_device(&pdev->dev);
 	if (!input) {
 		dev_err(&pdev->dev, "failed to allocate the input device\n");
-		return -ENOMEM;
+		error = -ENOMEM;
+		goto error_probe;
 	}
 
 	input->name = pdev->name;
@@ -195,7 +238,7 @@
 	error = devm_add_action(&pdev->dev, imx_snvs_pwrkey_act, pdata);
 	if (error) {
 		dev_err(&pdev->dev, "failed to register remove action\n");
-		return error;
+		goto error_probe;
 	}
 
 	pdata->input = input;
@@ -207,13 +250,13 @@
 
 	if (error) {
 		dev_err(&pdev->dev, "interrupt not available.\n");
-		return error;
+		goto error_probe;
 	}
 
 	error = input_register_device(input);
 	if (error < 0) {
 		dev_err(&pdev->dev, "failed to register input device\n");
-		return error;
+		goto error_probe;
 	}
 
 	device_init_wakeup(&pdev->dev, pdata->wakeup);
@@ -222,8 +265,43 @@
 		dev_err(&pdev->dev, "irq wake enable failed.\n");
 
 	return 0;
+
+error_probe:
+	if (pdata->clk)
+		clk_disable_unprepare(pdata->clk);
+
+	return error;
+}
+
+static int __maybe_unused imx_snvs_pwrkey_suspend(struct device *dev)
+{
+	struct platform_device *pdev = to_platform_device(dev);
+	struct pwrkey_drv_data *pdata = platform_get_drvdata(pdev);
+
+	if (pdata->clk)
+		clk_disable_unprepare(pdata->clk);
+
+	pdata->suspended = true;
+
+	return 0;
 }
 
+static int __maybe_unused imx_snvs_pwrkey_resume(struct device *dev)
+{
+	struct platform_device *pdev = to_platform_device(dev);
+	struct pwrkey_drv_data *pdata = platform_get_drvdata(pdev);
+
+	if (pdata->clk)
+		clk_prepare_enable(pdata->clk);
+
+	pdata->suspended = false;
+
+	return 0;
+}
+
+static SIMPLE_DEV_PM_OPS(imx_snvs_pwrkey_pm_ops, imx_snvs_pwrkey_suspend,
+				imx_snvs_pwrkey_resume);
+
 static const struct of_device_id imx_snvs_pwrkey_ids[] = {
 	{ .compatible = "fsl,sec-v4.0-pwrkey" },
 	{ /* sentinel */ }
@@ -233,6 +311,7 @@
 static struct platform_driver imx_snvs_pwrkey_driver = {
 	.driver = {
 		.name = "snvs_pwrkey",
+		.pm     = &imx_snvs_pwrkey_pm_ops,
 		.of_match_table = imx_snvs_pwrkey_ids,
 	},
 	.probe = imx_snvs_pwrkey_probe,
