--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/irqchip/irq-gic-v3.c	2026-04-08 18:20:07.574953933 +0200
+++ ./drivers/irqchip/irq-gic-v3.c	2026-04-08 16:21:06.724762900 +0200
@@ -149,6 +149,10 @@
 	__INVALID_RANGE__
 };
 
+#ifdef CONFIG_GIC_GENTLE_CONFIG
+static u64 gic_cpu_to_affinity(int cpu);
+#endif
+
 static enum gic_intid_range __get_intid_range(irq_hw_number_t hwirq)
 {
 	switch (hwirq) {
@@ -658,6 +662,9 @@
 	void __iomem *base;
 	u32 offset, index;
 	int ret;
+#ifdef CONFIG_GIC_GENTLE_CONFIG
+	u64 affinity;
+#endif
 
 	range = get_intid_range(d);
 
@@ -675,6 +682,16 @@
 	else
 		base = gic_dist_base_alias(d);
 
+#ifdef CONFIG_GIC_GENTLE_CONFIG
+	/* In a SoC running multiple OSes on ARM clusters sharing the same GIC,
+	 * set the affinity of the SPI here.
+	 * This allows to set the affinity to only the interrupts
+	 * registered by the cluster.
+	 */
+	affinity = gic_cpu_to_affinity(smp_processor_id());
+	gic_write_irouter(affinity, base + GICD_IROUTER + irq * 8);
+#endif
+
 	offset = convert_offset_index(d, GICD_ICFGR, &index);
 
 	ret = gic_configure_irq(index, type, base + offset, NULL);
@@ -914,10 +931,26 @@
 static void __init gic_dist_init(void)
 {
 	unsigned int i;
+#ifndef CONFIG_GIC_GENTLE_CONFIG
 	u64 affinity;
+#endif
 	void __iomem *base = gic_data.dist_base;
 	u32 val;
 
+#ifdef CONFIG_GIC_GENTLE_CONFIG
+       /* In a SoC running multiple OSes on ARM clusters sharing the same GIC,
+        * we take care of not re-configuring the distributor
+        * when another OS already did it, else this could interfere
+        * with the on-going interrupts directed to the other OS.
+        */
+       u32 gicd_ctlr = readl_relaxed(base + GICD_CTLR);
+
+       if (gicd_ctlr & (GICD_CTLR_ENABLE_G1A | GICD_CTLR_ENABLE_G1)) {
+               printk(KERN_INFO "GIC Distributor already configured: skip gic_dist_init\n");
+               return;
+       }
+#endif
+
 	/* Disable the distributor */
 	writel_relaxed(0, base + GICD_CTLR);
 	gic_dist_wait_for_rwp();
@@ -959,6 +992,14 @@
 	writel_relaxed(val, base + GICD_CTLR);
 	gic_dist_wait_for_rwp();
 
+#ifndef CONFIG_GIC_GENTLE_CONFIG
+	/* In a SoC running multiple OSes on ARM clusters sharing the same GIC,
+	 * do not set the affinity to all interrupts as this
+	 * would conflict with the other cluster's GIC configuration.
+	 * This is now done in function gic_set_type() (called by request_irq)
+	 * which allows to limit this to the interrupts registered by the
+	 * cluster.
+	 */
 	/*
 	 * Set all global interrupts to the boot CPU only. ARE must be
 	 * enabled.
@@ -969,6 +1010,7 @@
 
 	for (i = 0; i < GIC_ESPI_NR; i++)
 		gic_write_irouter(affinity, base + GICD_IROUTERnE + i * 8);
+#endif
 }
 
 static int gic_iterate_rdists(int (*fn)(struct redist_region *, void __iomem *))
