--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/media/i2c/Kconfig	2026-04-08 18:20:07.678600353 +0200
+++ ./drivers/media/i2c/Kconfig	2026-04-08 16:21:06.894606614 +0200
@@ -558,6 +558,60 @@
 	  To compile this driver as a module, choose M here: the
 	  module's name is ov9734.
 
+config VIDEO_OX03C10_LIB
+	tristate "Omnivision OX03C10 camera sensor library."
+	depends on I2C
+	help
+	  This library supports the Omnivision OX03C10 camera sensor. The sensor is usually paired
+	  with a serializer chip in a single module and needs a deserializer on the remote side in
+	  order to interface with the host platform using MIPI-CSI2.
+
+	  To compile this library as a module, choose M here: the module will be called ox03c10.
+
+config VIDEO_AP1302
+	tristate "ON Semiconductor's Advanced Image Coporcessor AP1302 support"
+	depends on OF
+	depends on GPIOLIB && VIDEO_DEV && I2C
+	select MEDIA_CONTROLLER
+	select VIDEO_V4L2_SUBDEV_API
+	select V4L2_FWNODE
+	help
+	  This is a Video4Linux2 sensor driver for the ONSEMI AP1302 camera
+	  sensor with a MIPI CSI-2 interface
+
+config VIDEO_AP130X
+       tristate "ON Semiconductor AP130X ISP"
+       depends on I2C && VIDEO_DEV
+       select VIDEO_V4L2_SUBDEV_API
+       select V4L2_FWNODE
+
+       help
+         This is a Video4Linux2 sensor-level driver for the ON Semiconductor
+         AP1302 ISP.
+
+         To compile this driver as a module, choose M here: the
+         module will be called ap1302.
+
+config VIDEO_MT9M114
+	tristate "mt9m114 support"
+	depends on I2C && VIDEO_DEV
+	select V4L2_FWNODE
+	help
+	  This driver supports MT9M114 cameras from ON Semiconductor
+
+config VIDEO_OX03C10_DRV
+	tristate "OmniVision OX03C10 sensor support"
+	select VIDEO_OX03C10_LIB
+	select V4L2_FWNODE
+	select VIDEO_V4L2_SUBDEV_API
+	select MEDIA_CONTROLLER
+	depends on I2C
+	depends on VIDEO_DEV
+	depends on OF_GPIO
+	help
+	  This is a Video4Linux2 sensor driver for the OmniVision
+	  OX03C10 camera.
+
 config VIDEO_RDACM20
 	tristate "IMI RDACM20 camera support"
 	select VIDEO_MAX9271_LIB
@@ -1055,6 +1109,8 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called ks0127.
 
+source "drivers/media/i2c/ox05b1s/Kconfig"
+
 config VIDEO_MAX9286
 	tristate "Maxim MAX9286 GMSL deserializer support"
 	depends on I2C && I2C_MUX
@@ -1069,6 +1125,30 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called max9286.
 
+config VIDEO_MAX96717_LIB
+	tristate "Maxim MAX96717 GMSL2 serializer support library."
+	depends on I2C
+	help
+	  This library supports the Maxim MAX96717 GMSL2 serializer which receives video on a
+	  MIPI CSI-2 interface and outputs it on a GMSL2 serial link transceiver.
+
+	  To compile this library as a module, choose M here: the module will be called max96717.
+
+config VIDEO_MAX96724
+	tristate "Maxim MAX96724 GMSL deserializer support"
+	depends on I2C && I2C_MUX
+	depends on VIDEO_DEV
+	depends on OF_GPIO
+	select V4L2_FWNODE
+	select VIDEO_V4L2_SUBDEV_API
+	select MEDIA_CONTROLLER
+	help
+	  This driver supports the Maxim MAX96724 GMSL deserializer which converts four GMSL2/1
+	  inputs to 1, 2, or 4 MIPI D-PHY or C-PHY outputs.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called max96724.
+
 config VIDEO_ML86V7667
 	tristate "OKI ML86V7667 video decoder"
 	depends on VIDEO_DEV && I2C
@@ -1078,6 +1158,21 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called ml86v7667.
 
+config VIDEO_MX95MBCAM
+	tristate "NXP OX03C10 + MAX961717 combo GMSL2 camera support"
+	select VIDEO_MAX96717_LIB
+	select VIDEO_OX03C10_LIB
+	depends on I2C
+	depends on VIDEO_DEV
+	depends on OF_GPIO
+	select V4L2_FWNODE
+	select VIDEO_V4L2_SUBDEV_API
+	select MEDIA_CONTROLLER
+	help
+	  This driver adds support for the OX03C10 camera module over a GMSL2 Maxim link. The
+	  module contains a MAX96717 GMSL2 serializer and needs to be used in conjuction with a
+	  compatible deserializer.
+
 config VIDEO_SAA7110
 	tristate "Philips SAA7110 video decoder"
 	depends on VIDEO_DEV && I2C
