--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/media/platform/nxp/imx8-isi/imx8-isi-core.c	2026-04-08 18:20:07.762838159 +0200
+++ ./drivers/media/platform/nxp/imx8-isi/imx8-isi-core.c	2026-04-08 16:21:07.032432513 +0200
@@ -279,6 +279,10 @@
 	{ .id = "apb" },
 };
 
+static const struct clk_bulk_data mxc_imx95_clks[] = {
+	{ .id = "axi" },
+};
+
 static const struct mxc_isi_plat_data mxc_imx8mn_data = {
 	.model			= MXC_ISI_IMX8MN,
 	.num_ports		= 1,
@@ -291,6 +295,8 @@
 	.buf_active_reverse	= false,
 	.gasket_ops		= &mxc_imx8_gasket_ops,
 	.has_36bit_dma		= false,
+	.raw32_chan_cfg		= false,
+	.buf_max_size		= false,
 };
 
 static const struct mxc_isi_plat_data mxc_imx8mp_data = {
@@ -305,6 +311,8 @@
 	.buf_active_reverse	= true,
 	.gasket_ops		= &mxc_imx8_gasket_ops,
 	.has_36bit_dma		= true,
+	.raw32_chan_cfg		= false,
+	.buf_max_size		= false,
 };
 
 static const struct mxc_isi_plat_data mxc_imx93_data = {
@@ -319,6 +327,24 @@
 	.buf_active_reverse	= true,
 	.gasket_ops		= &mxc_imx93_gasket_ops,
 	.has_36bit_dma		= false,
+	.raw32_chan_cfg		= false,
+	.buf_max_size		= false,
+};
+
+static const struct mxc_isi_plat_data mxc_imx95_data = {
+	.model			= MXC_ISI_IMX95,
+	.num_ports		= 4,
+	.num_channels		= 8,
+	.reg_offset		= 0x10000,
+	.ier_reg		= &mxc_imx8_isi_ier_v2,
+	.set_thd		= &mxc_imx8_isi_thd_v1,
+	.clks			= mxc_imx95_clks,
+	.num_clks		= ARRAY_SIZE(mxc_imx95_clks),
+	.buf_active_reverse	= true,
+	.gasket_ops             = &mxc_imx95_gasket_ops,
+	.has_36bit_dma		= true,
+	.raw32_chan_cfg		= true,
+	.buf_max_size		= true,
 };
 
 /* -----------------------------------------------------------------------------
@@ -334,6 +360,7 @@
 		struct mxc_isi_pipe *pipe = &isi->pipes[i];
 
 		mxc_isi_video_suspend(pipe);
+		mxc_isi_m2m_suspend(pipe);
 	}
 
 	return pm_runtime_force_suspend(dev);
@@ -363,6 +390,13 @@
 			 */
 			err = ret;
 		}
+
+		ret = mxc_isi_m2m_resume(pipe);
+		if (ret) {
+			dev_err(dev, "Failed to resume ISI%u (%d) for m2m\n", i,
+				ret);
+			err = ret;
+		}
 	}
 
 	return err;
@@ -476,6 +510,13 @@
 
 	pm_runtime_enable(dev);
 
+	ret = pm_runtime_resume_and_get(dev);
+	if (ret) {
+		dev_err_probe(dev, ret, "Failed to enable ISI\n");
+		pm_runtime_disable(dev);
+		return ret;
+	}
+
 	ret = mxc_isi_crossbar_init(isi);
 	if (ret) {
 		dev_err(dev, "Failed to initialize crossbar: %d\n", ret);
@@ -499,11 +540,13 @@
 
 	mxc_isi_debug_init(isi);
 
+	pm_runtime_put(dev);
 	return 0;
 
 err_xbar:
 	mxc_isi_crossbar_cleanup(&isi->crossbar);
 err_pm:
+	pm_runtime_put(dev);
 	pm_runtime_disable(isi->dev);
 	return ret;
 }
@@ -530,9 +573,10 @@
 }
 
 static const struct of_device_id mxc_isi_of_match[] = {
-	{ .compatible = "fsl,imx8mn-isi", .data = &mxc_imx8mn_data },
-	{ .compatible = "fsl,imx8mp-isi", .data = &mxc_imx8mp_data },
-	{ .compatible = "fsl,imx93-isi", .data = &mxc_imx93_data },
+	{ .compatible = "nxp,imx8mn-isi", .data = &mxc_imx8mn_data },
+	{ .compatible = "nxp,imx8mp-isi", .data = &mxc_imx8mp_data },
+	{ .compatible = "nxp,imx93-isi", .data = &mxc_imx93_data },
+	{ .compatible = "nxp,imx95-isi", .data = &mxc_imx95_data },
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, mxc_isi_of_match);
