--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/mfd/si476x-i2c.c	2026-04-08 18:20:07.913963380 +0200
+++ ./drivers/mfd/si476x-i2c.c	2026-04-08 16:21:07.211678967 +0200
@@ -97,7 +97,7 @@
 
 static inline void si476x_core_schedule_polling_work(struct si476x_core *core)
 {
-	schedule_delayed_work(&core->status_monitor,
+	queue_delayed_work(system_freezable_wq, &core->status_monitor,
 			      usecs_to_jiffies(SI476X_STATUS_POLL_US));
 }
 
@@ -294,7 +294,7 @@
 			 */
 			udelay(100);
 
-			err = si476x_core_start(core, false);
+			err = si476x_core_start(core, true);
 			if (err < 0)
 				goto disable_regulators;
 
@@ -303,7 +303,7 @@
 
 		case SI476X_POWER_DOWN:
 			core->power_state = next_state;
-			err = si476x_core_stop(core, false);
+			err = si476x_core_stop(core, true);
 			if (err < 0)
 				core->power_state = SI476X_POWER_INCONSISTENT;
 disable_regulators:
@@ -734,8 +734,15 @@
 		memcpy(&core->pinmux, &pdata->pinmux,
 		       sizeof(struct si476x_pinmux));
 	} else {
-		dev_err(&client->dev, "No platform data provided\n");
-		return -EINVAL;
+		dev_warn(&client->dev, "Using default platform data.\n");
+		core->power_up_parameters.xcload = 0x28;
+		core->power_up_parameters.func = SI476X_FUNC_FM_RECEIVER;
+		core->power_up_parameters.freq = SI476X_FREQ_37P209375_MHZ;
+		core->diversity_mode = SI476X_PHDIV_DISABLED;
+		core->pinmux.dclk = SI476X_DCLK_DAUDIO;
+		core->pinmux.dfs  = SI476X_DFS_DAUDIO;
+		core->pinmux.dout = SI476X_DOUT_I2S_OUTPUT;
+		core->pinmux.xout = SI476X_XOUT_TRISTATE;
 	}
 
 	core->supplies[0].supply = "vd";
@@ -794,12 +801,18 @@
 
 	core->chip_id = id->driver_data;
 
+	/* Power down si476x first */
+	si476x_core_stop(core, true);
+
 	rval = si476x_core_get_revision_info(core);
 	if (rval < 0) {
 		rval = -ENODEV;
 		goto free_kfifo;
 	}
 
+	if (of_property_read_bool(client->dev.of_node, "revision-a10"))
+		core->revision = SI476X_REVISION_A10;
+
 	cell_num = 0;
 
 	cell = &core->cells[SI476X_RADIO_CELL];
@@ -815,6 +828,7 @@
 	    core->pinmux.xout == SI476X_XOUT_TRISTATE) {
 		cell = &core->cells[SI476X_CODEC_CELL];
 		cell->name          = "si476x-codec";
+		cell->of_compatible = "si476x-codec";
 		cell_num++;
 	}
 #endif
