--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/mmc/core/sdio.c	2026-04-08 18:20:07.947964327 +0200
+++ ./drivers/mmc/core/sdio.c	2026-04-08 16:21:07.236131220 +0200
@@ -1045,7 +1045,7 @@
 
 	/* Prevent processing of SDIO IRQs in suspended state. */
 	mmc_card_set_suspended(host->card);
-	cancel_work_sync(&host->sdio_irq_work);
+	cancel_delayed_work_sync(&host->sdio_irq_work);
 
 	mmc_claim_host(host);
 
@@ -1111,7 +1111,7 @@
 		if (!(host->caps2 & MMC_CAP2_SDIO_IRQ_NOTHREAD))
 			wake_up_process(host->sdio_irq_thread);
 		else if (host->caps & MMC_CAP_SDIO_IRQ)
-			schedule_work(&host->sdio_irq_work);
+			queue_delayed_work(system_wq, &host->sdio_irq_work, 1);
 	}
 
 out:
