--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/mmc/core/sdio_irq.c	2026-04-08 18:20:07.948964355 +0200
+++ ./drivers/mmc/core/sdio_irq.c	2026-04-08 16:21:07.236394308 +0200
@@ -124,7 +124,7 @@
 void sdio_irq_work(struct work_struct *work)
 {
 	struct mmc_host *host =
-		container_of(work, struct mmc_host, sdio_irq_work);
+		container_of(work, struct mmc_host, sdio_irq_work.work);
 
 	sdio_run_irqs(host);
 }
@@ -132,7 +132,7 @@
 void sdio_signal_irq(struct mmc_host *host)
 {
 	host->sdio_irq_pending = true;
-	schedule_work(&host->sdio_irq_work);
+	queue_delayed_work(system_wq, &host->sdio_irq_work, 0);
 }
 EXPORT_SYMBOL_GPL(sdio_signal_irq);
 
