--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/mxc/vpu/wave6/Kconfig	1970-01-01 01:00:00.000000000 +0100
+++ ./drivers/mxc/vpu/wave6/Kconfig	2026-04-08 16:21:07.330297641 +0200
@@ -0,0 +1,25 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+comment "Wave6 drivers"
+
+config MXC_VIDEO_WAVE6_CTRL
+	tristate "WAVE6 VPU Ctrl"
+	help
+	  Wave6 VPU Ctrl is to handle the global region.
+	  It help load the firmware and boot the vcpu,
+	  It also help manage the vpu power domain and clock,
+	  To compile this driver as a module choose m here.
+
+config MXC_VIDEO_WAVE6
+	tristate "WAVE6 VPU (Video Processing Unit) Codec IP"
+	depends on V4L_MEM2MEM_DRIVERS
+	depends on ARCH_MXC || COMPILE_TEST
+	depends on MEDIA_SUPPORT
+	depends on VIDEO_DEV
+	select MEDIA_CONTROLLER
+	select V4L2_MEM2MEM_DEV
+	select VIDEOBUF2_DMA_CONTIG
+	select MXC_VIDEO_WAVE6_CTRL
+	help
+	  Wave6 VPU Codec IP contains decoders and encoders.
+	  To compile this driver as a module choose m here.
