--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/dsa/ocelot/felix_vsc9959.c	2026-04-08 18:20:08.034005997 +0200
+++ ./drivers/net/dsa/ocelot/felix_vsc9959.c	2026-04-08 16:21:07.363840702 +0200
@@ -13,12 +13,14 @@
 #include <soc/mscc/ocelot.h>
 #include <linux/dsa/ocelot.h>
 #include <linux/pcs-lynx.h>
+#include <linux/phy/phy.h>
 #include <net/pkt_sched.h>
 #include <linux/iopoll.h>
 #include <linux/mdio.h>
 #include <linux/of.h>
 #include <linux/pci.h>
 #include <linux/time.h>
+#include "felix_tsn.h"
 #include "felix.h"
 
 #define VSC9959_NUM_PORTS		6
@@ -34,7 +36,8 @@
 					 OCELOT_PORT_MODE_QSGMII | \
 					 OCELOT_PORT_MODE_1000BASEX | \
 					 OCELOT_PORT_MODE_2500BASEX | \
-					 OCELOT_PORT_MODE_USXGMII)
+					 OCELOT_PORT_MODE_USXGMII | \
+					 OCELOT_PORT_MODE_10G_QXGMII)
 
 static const u32 vsc9959_port_modes[VSC9959_NUM_PORTS] = {
 	VSC9959_PORT_MODE_SERDES,
@@ -958,13 +961,14 @@
 	struct pci_dev *pdev = to_pci_dev(ocelot->dev);
 	struct felix *felix = ocelot_to_felix(ocelot);
 	struct enetc_mdio_priv *mdio_priv;
+	struct dsa_switch *ds = felix->ds;
 	struct device *dev = ocelot->dev;
 	resource_size_t imdio_base;
 	void __iomem *imdio_regs;
 	struct resource res;
 	struct enetc_hw *hw;
 	struct mii_bus *bus;
-	int port;
+	struct dsa_port *dp;
 	int rc;
 
 	felix->pcs = devm_kcalloc(dev, felix->info->num_ports,
@@ -1019,23 +1023,25 @@
 
 	felix->imdio = bus;
 
-	for (port = 0; port < felix->info->num_ports; port++) {
-		struct ocelot_port *ocelot_port = ocelot->ports[port];
+	dsa_switch_for_each_available_port(dp, ds) {
+		struct ocelot_port *ocelot_port = ocelot->ports[dp->index];
+		size_t num_phys = ocelot_port->serdes ? 1 : 0;
 		struct phylink_pcs *phylink_pcs;
 
-		if (dsa_is_unused_port(felix->ds, port))
+		/* Skip on internal ports */
+		if (dp->index == 4 || dp->index == 5)
 			continue;
 
-		if (ocelot_port->phy_mode == PHY_INTERFACE_MODE_INTERNAL)
-			continue;
-
-		phylink_pcs = lynx_pcs_create_mdiodev(felix->imdio, port);
+		phylink_pcs = lynx_pcs_create_mdiodev(felix->imdio, dp->index,
+						      &ocelot_port->serdes,
+						      num_phys);
 		if (IS_ERR(phylink_pcs))
 			continue;
 
-		felix->pcs[port] = phylink_pcs;
+		felix->pcs[dp->index] = phylink_pcs;
 
-		dev_info(dev, "Found PCS at internal MDIO address %d\n", port);
+		dev_info(dev, "Found PCS at internal MDIO address %d\n",
+			 dp->index);
 	}
 
 	return 0;
@@ -1380,11 +1386,12 @@
 		vsc9959_tas_guard_bands_update(ocelot, port);
 
 	mutex_unlock(&ocelot->fwd_domain_lock);
+
+	felix_cbs_reset(ocelot, port, speed);
 }
 
-static void vsc9959_new_base_time(struct ocelot *ocelot, ktime_t base_time,
-				  u64 cycle_time,
-				  struct timespec64 *new_base_ts)
+void vsc9959_new_base_time(struct ocelot *ocelot, ktime_t base_time,
+			   u64 cycle_time, struct timespec64 *new_base_ts)
 {
 	struct timespec64 ts;
 	ktime_t new_base_time;
@@ -2592,7 +2599,7 @@
 		 * reason, if sent as cut-through.
 		 */
 		if (ocelot_port->speed == min_speed) {
-			val = GENMASK(7, 0) & ~mm->active_preemptible_tcs;
+			val = ocelot_port->cut_thru & ~mm->active_preemptible_tcs;
 
 			for (tc = 0; tc < OCELOT_NUM_TC; tc++)
 				if (vsc9959_port_qmaxsdu_get(ocelot, port, tc))
@@ -2735,6 +2742,8 @@
 		goto err_register_ds;
 	}
 
+	felix_tsn_enable(ds);
+
 	return 0;
 
 err_register_ds:
