--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/dsa/ocelot/seville_vsc9953.c	2026-04-08 18:20:08.034005997 +0200
+++ ./drivers/net/dsa/ocelot/seville_vsc9953.c	2026-04-08 16:21:07.363957743 +0200
@@ -11,6 +11,7 @@
 #include <linux/mod_devicetable.h>
 #include <linux/of_mdio.h>
 #include <linux/pcs-lynx.h>
+#include <linux/phy/phy.h>
 #include <linux/dsa/ocelot.h>
 #include <linux/iopoll.h>
 #include "felix.h"
@@ -879,9 +880,10 @@
 static int vsc9953_mdio_bus_alloc(struct ocelot *ocelot)
 {
 	struct felix *felix = ocelot_to_felix(ocelot);
+	struct dsa_switch *ds = felix->ds;
 	struct device *dev = ocelot->dev;
 	struct mii_bus *bus;
-	int port;
+	struct dsa_port *dp;
 	int rc;
 
 	felix->pcs = devm_kcalloc(dev, felix->info->num_ports,
@@ -910,22 +912,22 @@
 
 	felix->imdio = bus;
 
-	for (port = 0; port < felix->info->num_ports; port++) {
-		struct ocelot_port *ocelot_port = ocelot->ports[port];
+	dsa_switch_for_each_available_port(dp, ds) {
+		struct ocelot_port *ocelot_port = ocelot->ports[dp->index];
+		size_t num_phys = ocelot_port->serdes ? 1 : 0;
 		struct phylink_pcs *phylink_pcs;
-		int addr = port + 4;
+		int addr = dp->index + 4;
 
-		if (dsa_is_unused_port(felix->ds, port))
+		if (dp->index == 8 || dp->index == 9)
 			continue;
 
-		if (ocelot_port->phy_mode == PHY_INTERFACE_MODE_INTERNAL)
-			continue;
-
-		phylink_pcs = lynx_pcs_create_mdiodev(felix->imdio, addr);
+		phylink_pcs = lynx_pcs_create_mdiodev(felix->imdio, addr,
+						      &ocelot_port->serdes,
+						      num_phys);
 		if (IS_ERR(phylink_pcs))
 			continue;
 
-		felix->pcs[port] = phylink_pcs;
+		felix->pcs[dp->index] = phylink_pcs;
 
 		dev_info(dev, "Found PCS at internal MDIO address %d\n", addr);
 	}
