--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/ethernet/altera/altera_tse_main.c	2026-04-08 18:20:08.047528519 +0200
+++ ./drivers/net/ethernet/altera/altera_tse_main.c	2026-04-08 16:21:07.379031844 +0200
@@ -1419,7 +1419,7 @@
 		goto err_init_pcs;
 	}
 
-	priv->pcs = lynx_pcs_create_mdiodev(pcs_bus, 0);
+	priv->pcs = lynx_pcs_create_mdiodev(pcs_bus, 0, NULL, 0);
 	if (IS_ERR(priv->pcs)) {
 		ret = PTR_ERR(priv->pcs);
 		goto err_init_pcs;
