--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c	2026-04-08 18:20:08.150292422 +0200
+++ ./drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c	2026-04-08 16:21:07.515545652 +0200
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: (GPL-2.0+ OR BSD-3-Clause)
 /* Copyright 2014-2016 Freescale Semiconductor Inc.
- * Copyright 2016-2022 NXP
+ * Copyright 2016-2022, 2024 NXP
  */
 #include <linux/init.h>
 #include <linux/module.h>
@@ -15,12 +15,14 @@
 #include <linux/bpf_trace.h>
 #include <linux/fsl/ptp_qoriq.h>
 #include <linux/ptp_classify.h>
+#include <linux/device/driver.h>
 #include <net/pkt_cls.h>
 #include <net/sock.h>
 #include <net/tso.h>
 #include <net/xdp_sock_drv.h>
 
 #include "dpaa2-eth.h"
+#include "dpaa2-eth-ceetm.h"
 
 /* CREATE_TRACE_POINTS only needs to be defined once. Other dpa files
  * using trace events only need to #include <trace/events/sched.h>
@@ -42,6 +44,14 @@
 	if (dpaa2_eth_cmp_dpni_ver(priv, DPNI_PTP_ONESTEP_VER_MAJOR,
 				   DPNI_PTP_ONESTEP_VER_MINOR) >= 0)
 		priv->features |= DPAA2_ETH_FEATURE_ONESTEP_CFG_DIRECT;
+
+	if (dpaa2_eth_cmp_dpni_ver(priv, DPNI_NUM_TX_TCS_VER_MAJOR,
+				   DPNI_NUM_TX_TCS_VER_MINOR) >= 0)
+		priv->features |= DPAA2_ETH_FEATURE_GET_NUM_TX_TCS;
+
+	if (dpaa2_eth_cmp_dpni_ver(priv, DPNI_MACSEC_VER_MAJOR,
+				   DPNI_MACSEC_VER_MINOR) >= 0)
+		priv->features |= DPAA2_ETH_FEATURE_MACSEC;
 }
 
 static void dpaa2_update_ptp_onestep_indirect(struct dpaa2_eth_priv *priv,
@@ -532,6 +542,16 @@
 	return dpaa2_eth_alloc_skb(priv, ch, fd, fd_length, fd_vaddr);
 }
 
+static bool frame_is_tcp(const struct dpaa2_fd *fd, struct dpaa2_fas *fas)
+{
+	struct dpaa2_fapr *fapr = dpaa2_get_fapr(fas, false);
+
+	if (!(dpaa2_fd_get_frc(fd) & DPAA2_FD_FRC_FAPRV))
+		return false;
+
+	return !!(fapr->faf_hi & DPAA2_FAF_HI_TCP_PRESENT);
+}
+
 void dpaa2_eth_receive_skb(struct dpaa2_eth_priv *priv,
 			   struct dpaa2_eth_channel *ch,
 			   const struct dpaa2_fd *fd, void *vaddr,
@@ -562,6 +582,11 @@
 	if (likely(dpaa2_fd_get_frc(fd) & DPAA2_FD_FRC_FASV)) {
 		status = le32_to_cpu(fas->status);
 		dpaa2_eth_validate_rx_csum(priv, status, skb);
+
+		if (priv->secy_id && (status & DPAA2_FAS_MS)) {
+			dst_hold(&priv->md_dst->dst);
+			skb_dst_set(skb, &priv->md_dst->dst);
+		}
 	}
 
 	skb->protocol = eth_type_trans(skb, priv->net_dev);
@@ -571,7 +596,10 @@
 	percpu_stats->rx_bytes += dpaa2_fd_get_len(fd);
 	ch->stats.bytes_per_cdan += dpaa2_fd_get_len(fd);
 
-	list_add_tail(&skb->list, ch->rx_list);
+	if (frame_is_tcp(fd, fas))
+		napi_gro_receive(&ch->napi, skb);
+	else
+		list_add_tail(&skb->list, ch->rx_list);
 }
 
 /* Main Rx frame processing routine */
@@ -796,7 +824,8 @@
 static void dpaa2_eth_enable_tx_tstamp(struct dpaa2_eth_priv *priv,
 				       struct dpaa2_fd *fd,
 				       void *buf_start,
-				       struct sk_buff *skb)
+				       struct sk_buff *skb,
+				       bool memset_faead)
 {
 	struct ptp_tstamp origin_timestamp;
 	u8 msgtype, twostep, udp;
@@ -821,6 +850,10 @@
 	 */
 	ctrl = DPAA2_FAEAD_A2V | DPAA2_FAEAD_UPDV | DPAA2_FAEAD_UPD;
 	faead = dpaa2_get_faead(buf_start, true);
+	if (memset_faead) {
+		faead->conf_fqid = 0;
+		faead->ctrl = 0;
+	}
 	faead->ctrl = cpu_to_le32(ctrl);
 
 	if (skb->cb[0] == TX_TSTAMP_ONESTEP_SYNC) {
@@ -1223,6 +1256,12 @@
 		return;
 	}
 
+	/* If there is no skb (in case of a wrong SWA type)
+	 * just exit the function
+	 */
+	if (!skb)
+		return;
+
 	/* Get the timestamp value */
 	if (swa->type != DPAA2_ETH_SWA_SW_TSO) {
 		if (skb->cb[0] == TX_TSTAMP) {
@@ -1392,6 +1431,39 @@
 	return err;
 }
 
+static void dpaa2_eth_mark_macsec(struct dpaa2_eth_priv *priv, struct sk_buff *skb,
+				  struct dpaa2_fd *fd, int num_fds,
+				  bool memset_faead)
+{
+	struct dpaa2_faead *faead;
+	struct dpaa2_fd *crr_fd;
+	dma_addr_t fd_addr;
+	void *buf_start;
+	u32 ctrl, frc;
+	int i;
+
+	for (i = 0; i < num_fds; i++) {
+		crr_fd = &fd[i];
+		fd_addr = dpaa2_fd_get_addr(crr_fd);
+		buf_start = dpaa2_iova_to_virt(priv->iommu_domain, fd_addr);
+
+		frc = dpaa2_fd_get_frc(crr_fd);
+		dpaa2_fd_set_frc(crr_fd, frc | DPAA2_FD_FRC_FAEADV);
+
+		ctrl = dpaa2_fd_get_ctrl(fd);
+		dpaa2_fd_set_ctrl(fd, ctrl | DPAA2_FD_CTRL_ASAL);
+
+		faead = dpaa2_get_faead(buf_start, true);
+		if (memset_faead) {
+			faead->conf_fqid = 0;
+			faead->ctrl = 0;
+		}
+		ctrl = le32_to_cpu(faead->ctrl);
+		ctrl |= DPAA2_FAEAD_MCVV | DPAA2_FAEAD_A2V | DPAA2_FAEAD_MCV;
+		faead->ctrl = cpu_to_le32(ctrl);
+	}
+}
+
 static netdev_tx_t __dpaa2_eth_tx(struct sk_buff *skb,
 				  struct net_device *net_dev)
 {
@@ -1401,21 +1473,20 @@
 	struct rtnl_link_stats64 *percpu_stats;
 	unsigned int needed_headroom;
 	int num_fds = 1, max_retries;
+	bool memset_faead = true;
 	struct dpaa2_eth_fq *fq;
 	struct netdev_queue *nq;
+	int err, i, ch_id = 0;
 	struct dpaa2_fd *fd;
 	u16 queue_mapping;
 	void *swa = NULL;
 	u8 prio = 0;
-	int err, i;
 	u32 fd_len;
 
 	percpu_stats = this_cpu_ptr(priv->percpu_stats);
 	percpu_extras = this_cpu_ptr(priv->percpu_extras);
 	fd = (this_cpu_ptr(priv->fd))->array;
 
-	needed_headroom = dpaa2_eth_needed_headroom(skb);
-
 	/* We'll be holding a back-reference to the skb until Tx Confirmation;
 	 * we don't want that overwritten by a concurrent Tx with a cloned skb.
 	 */
@@ -1426,6 +1497,8 @@
 		return NETDEV_TX_OK;
 	}
 
+	needed_headroom = dpaa2_eth_needed_headroom(skb);
+
 	/* Setup the FD fields */
 
 	if (skb_is_gso(skb)) {
@@ -1456,8 +1529,13 @@
 		goto err_build_fd;
 	}
 
-	if (swa && skb->cb[0])
-		dpaa2_eth_enable_tx_tstamp(priv, fd, swa, skb);
+	if (swa && skb->cb[0]) {
+		dpaa2_eth_enable_tx_tstamp(priv, fd, swa, skb, memset_faead);
+		memset_faead = false;
+	}
+
+	if (dpaa2_macsec_skb_is_offload(skb))
+		dpaa2_eth_mark_macsec(priv, skb, fd, num_fds, memset_faead);
 
 	/* Tracing point */
 	for (i = 0; i < num_fds; i++)
@@ -1481,6 +1559,16 @@
 		queue_mapping %= dpaa2_eth_queue_count(priv);
 	}
 	fq = &priv->fq[queue_mapping];
+
+	if (dpaa2_eth_ceetm_is_enabled(priv)) {
+		err = dpaa2_ceetm_classify(skb, net_dev->qdisc, &ch_id, &prio);
+		if (err) {
+			dpaa2_eth_free_tx_fd(priv, NULL, fq, fd, false);
+			percpu_stats->tx_dropped++;
+			return NETDEV_TX_OK;
+		}
+	}
+
 	nq = netdev_get_tx_queue(net_dev, queue_mapping);
 	netdev_tx_sent_queue(nq, fd_len);
 
@@ -2116,7 +2204,7 @@
 
 	td.threshold = DPAA2_ETH_CG_TAILDROP_THRESH(priv);
 	td.units = DPNI_CONGESTION_UNIT_FRAMES;
-	for (i = 0; i < dpaa2_eth_tc_count(priv); i++) {
+	for (i = 0; i < dpaa2_eth_rx_tc_count(priv); i++) {
 		err = dpni_set_taildrop(priv->mc_io, 0, priv->mc_token,
 					DPNI_CP_GROUP, DPNI_QUEUE_RX,
 					i, 0, &td);
@@ -2270,7 +2358,7 @@
 		goto out;
 
 	do {
-		err = dpni_get_statistics(priv->mc_io, 0, priv->mc_token, 6,
+		err = dpni_get_statistics(priv->mc_io, 0, priv->mc_token, 6, 0,
 					  &stats);
 		if (err)
 			goto out;
@@ -2390,6 +2478,8 @@
 	int err;
 
 	netdev_for_each_uc_addr(ha, net_dev) {
+		if (ether_addr_equal(ha->addr, net_dev->dev_addr))
+			continue;
 		err = dpni_add_mac_addr(priv->mc_io, 0, priv->mc_token,
 					ha->addr);
 		if (err)
@@ -2941,9 +3031,9 @@
 	if (num_tc == net_dev->num_tc)
 		return 0;
 
-	if (num_tc  > dpaa2_eth_tc_count(priv)) {
+	if (num_tc  > dpaa2_eth_tx_tc_count(priv)) {
 		netdev_err(net_dev, "Max %d traffic classes supported\n",
-			   dpaa2_eth_tc_count(priv));
+			   dpaa2_eth_tx_tc_count(priv));
 		return -EOPNOTSUPP;
 	}
 
@@ -3014,6 +3104,8 @@
 		return dpaa2_eth_setup_mqprio(net_dev, type_data);
 	case TC_SETUP_QDISC_TBF:
 		return dpaa2_eth_setup_tbf(net_dev, type_data);
+	case TC_SETUP_BLOCK:
+		return 0;
 	default:
 		return -EOPNOTSUPP;
 	}
@@ -3346,7 +3438,7 @@
 		priv->fq[priv->num_fqs++].flowid = (u16)i;
 	}
 
-	for (j = 0; j < dpaa2_eth_tc_count(priv); j++) {
+	for (j = 0; j < dpaa2_eth_rx_tc_count(priv); j++) {
 		for (i = 0; i < dpaa2_eth_queue_count(priv); i++) {
 			priv->fq[priv->num_fqs].type = DPAA2_RX_FQ;
 			priv->fq[priv->num_fqs].consume = dpaa2_eth_rx;
@@ -3653,7 +3745,7 @@
 		fq = &priv->fq[i];
 		if (fq->type != DPAA2_TX_CONF_FQ)
 			continue;
-		for (j = 0; j < dpaa2_eth_tc_count(priv); j++) {
+		for (j = 0; j < dpaa2_eth_tx_tc_count(priv); j++) {
 			err = dpni_get_queue(priv->mc_io, 0, priv->mc_token,
 					     DPNI_QUEUE_TX, j, fq->flowid,
 					     &queue, &qid);
@@ -3692,7 +3784,7 @@
 	 * Also, we need to extract just the 3-bit PCP field from the VLAN
 	 * header and we can only do that by using a mask
 	 */
-	if (dpaa2_eth_tc_count(priv) == 1 || !dpaa2_eth_fs_mask_enabled(priv)) {
+	if (dpaa2_eth_rx_tc_count(priv) == 1 || !dpaa2_eth_fs_mask_enabled(priv)) {
 		dev_dbg(dev, "VLAN-based QoS classification not supported\n");
 		return -EOPNOTSUPP;
 	}
@@ -3756,7 +3848,7 @@
 	 * classes to accommodate all priority levels, the lowest ones end up
 	 * on TC 0 which was configured as default
 	 */
-	for (i = dpaa2_eth_tc_count(priv) - 1, pcp = 7; i >= 0; i--, pcp--) {
+	for (i = dpaa2_eth_rx_tc_count(priv) - 1, pcp = 7; i >= 0; i--, pcp--) {
 		*(__be16 *)key = cpu_to_be16(pcp << VLAN_PRIO_SHIFT);
 		dma_sync_single_for_device(dev, key_params.key_iova,
 					   key_size * 2, DMA_TO_DEVICE);
@@ -3821,6 +3913,8 @@
 		goto close;
 	}
 
+	dpaa2_eth_detect_features(priv);
+
 	ls_dev->mc_io = priv->mc_io;
 	ls_dev->mc_handle = priv->mc_token;
 
@@ -3830,8 +3924,14 @@
 		goto close;
 	}
 
-	err = dpni_get_attributes(priv->mc_io, 0, priv->mc_token,
-				  &priv->dpni_attrs);
+	if (priv->features & DPAA2_ETH_FEATURE_GET_NUM_TX_TCS) {
+		err = dpni_get_attributes(priv->mc_io, 0, priv->mc_token,
+					  &priv->dpni_attrs, DPNI_CMDID_GET_ATTR_V2);
+	} else {
+		err = dpni_get_attributes(priv->mc_io, 0, priv->mc_token,
+					  &priv->dpni_attrs, DPNI_CMDID_GET_ATTR);
+		priv->dpni_attrs.num_tx_tcs = priv->dpni_attrs.num_rx_tcs;
+	}
 	if (err) {
 		dev_err(dev, "dpni_get_attributes() failed (err=%d)\n", err);
 		goto close;
@@ -3942,7 +4042,7 @@
 	struct dpni_queue_id qid;
 	int i, err;
 
-	for (i = 0; i < dpaa2_eth_tc_count(priv); i++) {
+	for (i = 0; i < dpaa2_eth_tx_tc_count(priv); i++) {
 		err = dpni_get_queue(priv->mc_io, 0, priv->mc_token,
 				     DPNI_QUEUE_TX, i, fq->flowid,
 				     &queue, &qid);
@@ -4094,7 +4194,7 @@
 	dist_cfg.dist_size = dpaa2_eth_queue_count(priv);
 	dist_cfg.dist_mode = DPNI_DIST_MODE_HASH;
 
-	for (i = 0; i < dpaa2_eth_tc_count(priv); i++) {
+	for (i = 0; i < dpaa2_eth_rx_tc_count(priv); i++) {
 		err = dpni_set_rx_tc_dist(priv->mc_io, 0, priv->mc_token,
 					  i, &dist_cfg);
 		if (err) {
@@ -4119,7 +4219,7 @@
 	dist_cfg.dist_size = dpaa2_eth_queue_count(priv);
 	dist_cfg.enable = 1;
 
-	for (i = 0; i < dpaa2_eth_tc_count(priv); i++) {
+	for (i = 0; i < dpaa2_eth_rx_tc_count(priv); i++) {
 		dist_cfg.tc = i;
 		err = dpni_set_rx_hash_dist(priv->mc_io, 0, priv->mc_token,
 					    &dist_cfg);
@@ -4151,7 +4251,7 @@
 	dist_cfg.dist_size = dpaa2_eth_queue_count(priv);
 	dist_cfg.enable = 1;
 
-	for (i = 0; i < dpaa2_eth_tc_count(priv); i++) {
+	for (i = 0; i < dpaa2_eth_rx_tc_count(priv); i++) {
 		dist_cfg.tc = i;
 		err = dpni_set_rx_fs_dist(priv->mc_io, 0, priv->mc_token,
 					  &dist_cfg);
@@ -4582,8 +4682,6 @@
 		return err;
 	}
 
-	dpaa2_eth_detect_features(priv);
-
 	/* Capabilities listing */
 	supported |= IFF_LIVE_ADDR_CHANGE;
 
@@ -4648,6 +4746,12 @@
 	if (IS_ERR(dpmac_dev) || dpmac_dev->dev.type != &fsl_mc_bus_dpmac_type)
 		return 0;
 
+#if IS_ENABLED(CONFIG_MACSEC)
+	dpaa2_eth_macsec_init(priv);
+#endif
+
+	dpaa2_mac_driver_detach(dpmac_dev);
+
 	mac = kzalloc(sizeof(struct dpaa2_mac), GFP_KERNEL);
 	if (!mac)
 		return -ENOMEM;
@@ -4703,7 +4807,12 @@
 		dpaa2_mac_disconnect(mac);
 
 	dpaa2_mac_close(mac);
+	dpaa2_mac_driver_attach(mac->mc_dev);
 	kfree(mac);
+
+#if IS_ENABLED(CONFIG_MACSEC)
+	dpaa2_eth_macsec_deinit(priv);
+#endif
 }
 
 static irqreturn_t dpni_irq0_handler_thread(int irq_num, void *arg)
@@ -4823,7 +4932,9 @@
 	dev = &dpni_dev->dev;
 
 	/* Net device */
-	net_dev = alloc_etherdev_mq(sizeof(*priv), DPAA2_ETH_MAX_NETDEV_QUEUES);
+	net_dev = alloc_etherdev_mqs(sizeof(*priv),
+				     DPAA2_ETH_MAX_NETDEV_TX_QUEUES,
+				     DPAA2_ETH_MAX_NETDEV_RX_QUEUES);
 	if (!net_dev) {
 		dev_err(dev, "alloc_etherdev_mq() failed\n");
 		return -ENOMEM;
@@ -4887,6 +4998,30 @@
 	if (err)
 		goto err_bind;
 
+	/* Setting (creating a CGID and assigning it on a frame queue that does
+	 * not have a congestion group associated  or modifying
+	 * an existing CGID on a frame queue to a different CGID) a CGID on a
+	 * frame queue that is in service(has frames in it),
+	 * will cause the byte/frame counter of that CGR to become corrupted.
+	 *
+	 * Below is the detailed solution and the impact if it is not applied
+	 * precisely.
+	 *
+	 * Enable tail drop at probe time to prevent instantaneous counter
+	 * of a congestion group to be 0 while traffic might be in flight on
+	 * the ingress queues. The instantaneous counter will be
+	 * decremented by the number of frames that are dequeued from the ingress
+	 * queue, once the interface is up. Because its value is 0,
+	 * this subtraction yields  an invalid overflow  value.(7FFFFFFFFF)
+	 * The outcome will be greater than
+	 * any configured value for the tail drop threshold, for any queue in the
+	 * group;
+	 * as a consequence, the group will always discard frames and this will
+	 * trigger a situation in which all the frames received on a queue that
+	 * belongs to the congestion group, are discarded continuously.
+	 */
+	dpaa2_eth_set_rx_taildrop(priv, false, priv->pfc_enabled);
+
 	/* Add a NAPI context for each channel */
 	dpaa2_eth_add_ch_napi(priv);
 
@@ -5107,18 +5242,27 @@
 
 	dpaa2_eth_dbg_init();
 	err = fsl_mc_driver_register(&dpaa2_eth_driver);
-	if (err) {
-		dpaa2_eth_dbg_exit();
-		return err;
-	}
+	if (err)
+		goto out_debugfs_err;
+
+	err = dpaa2_ceetm_register();
+	if (err)
+		goto out_ceetm_err;
 
 	return 0;
+
+out_ceetm_err:
+	fsl_mc_driver_unregister(&dpaa2_eth_driver);
+out_debugfs_err:
+	dpaa2_eth_dbg_exit();
+	return err;
 }
 
 static void __exit dpaa2_eth_driver_exit(void)
 {
-	dpaa2_eth_dbg_exit();
+	dpaa2_ceetm_unregister();
 	fsl_mc_driver_unregister(&dpaa2_eth_driver);
+	dpaa2_eth_dbg_exit();
 }
 
 module_init(dpaa2_eth_driver_init);
