--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/ethernet/freescale/dpaa2/dpaa2-mac.c	2026-04-08 18:20:08.150292422 +0200
+++ ./drivers/net/ethernet/freescale/dpaa2/dpaa2-mac.c	2026-04-08 16:21:07.516379878 +0200
@@ -1,11 +1,12 @@
 // SPDX-License-Identifier: (GPL-2.0+ OR BSD-3-Clause)
-/* Copyright 2019 NXP */
+/* Copyright 2019, 2024 NXP */
 
 #include <linux/acpi.h>
 #include <linux/pcs-lynx.h>
 #include <linux/phy/phy.h>
 #include <linux/property.h>
-
+#include <linux/fsl/mc.h>
+#include <linux/msi.h>
 #include "dpaa2-eth.h"
 #include "dpaa2-mac.h"
 
@@ -15,7 +16,14 @@
 #define DPMAC_PROTOCOL_CHANGE_VER_MAJOR		4
 #define DPMAC_PROTOCOL_CHANGE_VER_MINOR		8
 
+#define DPMAC_STATS_BUNDLE_VER_MAJOR		4
+#define DPMAC_STATS_BUNDLE_VER_MINOR		10
+
 #define DPAA2_MAC_FEATURE_PROTOCOL_CHANGE	BIT(0)
+#define DPAA2_MAC_FEATURE_STATS_BUNDLE		BIT(1)
+
+static void dpaa2_mac_setup_stats(struct dpaa2_mac *mac);
+static void dpaa2_mac_clear_stats(struct dpaa2_mac *mac);
 
 static int dpaa2_mac_cmp_ver(struct dpaa2_mac *mac,
 			     u16 ver_major, u16 ver_minor)
@@ -32,6 +40,10 @@
 	if (dpaa2_mac_cmp_ver(mac, DPMAC_PROTOCOL_CHANGE_VER_MAJOR,
 			      DPMAC_PROTOCOL_CHANGE_VER_MINOR) >= 0)
 		mac->features |= DPAA2_MAC_FEATURE_PROTOCOL_CHANGE;
+
+	if (dpaa2_mac_cmp_ver(mac, DPMAC_STATS_BUNDLE_VER_MAJOR,
+			      DPMAC_STATS_BUNDLE_VER_MINOR) >= 0)
+		mac->features |= DPAA2_MAC_FEATURE_STATS_BUNDLE;
 }
 
 static int phy_mode(enum dpmac_eth_if eth_if, phy_interface_t *if_mode)
@@ -79,10 +91,14 @@
 	case PHY_INTERFACE_MODE_SGMII:
 		return DPMAC_ETH_IF_SGMII;
 	case PHY_INTERFACE_MODE_10GBASER:
+	case PHY_INTERFACE_MODE_10GKR:
+	case PHY_INTERFACE_MODE_40GKR4:
 		return DPMAC_ETH_IF_XFI;
 	case PHY_INTERFACE_MODE_1000BASEX:
+	case PHY_INTERFACE_MODE_1000BASEKX:
 		return DPMAC_ETH_IF_1000BASEX;
 	case PHY_INTERFACE_MODE_25GBASER:
+	case PHY_INTERFACE_MODE_25GKR:
 		return DPMAC_ETH_IF_CAUI;
 	default:
 		return DPMAC_ETH_IF_MII;
@@ -136,12 +152,21 @@
 				 struct dpmac_attr attr)
 {
 	phy_interface_t if_mode;
+	const char *managed;
 	int err;
 
 	err = fwnode_get_phy_mode(dpmac_node);
 	if (err > 0)
 		return err;
 
+	/* Let phylink select state->interface, ignoring the dpmac eth_if value
+	 * provided by the MC firmware. That is even slightly bogus
+	 * (for 40G-KR4, the MC firmware returns DPMAC_ETH_IF_XFI)
+	 */
+	if (fwnode_property_read_string(dpmac_node, "managed", &managed) == 0 &&
+	    strcmp(managed, "c73") == 0)
+		return PHY_INTERFACE_MODE_NA;
+
 	err = phy_mode(attr.eth_if, &if_mode);
 	if (!err)
 		return if_mode;
@@ -162,6 +187,7 @@
 {
 	struct dpaa2_mac *mac = phylink_to_dpaa2_mac(config);
 	struct dpmac_link_state *dpmac_state = &mac->state;
+	size_t i;
 	int err;
 
 	if (linkmode_test_bit(ETHTOOL_LINK_MODE_Autoneg_BIT,
@@ -176,18 +202,20 @@
 		netdev_err(mac->net_dev, "%s: dpmac_set_link_state() = %d\n",
 			   __func__, err);
 
-	if (!mac->serdes_phy)
+	if (!mac->num_phys || phylink_autoneg_c73(mode))
 		return;
 
 	/* This happens only if we support changing of protocol at runtime */
 	err = dpmac_set_protocol(mac->mc_io, 0, mac->mc_dev->mc_handle,
 				 dpmac_eth_if_mode(state->interface));
 	if (err)
-		netdev_err(mac->net_dev,  "dpmac_set_protocol() = %d\n", err);
+		netdev_err(mac->net_dev, "dpmac_set_protocol(%s) = %pe\n",
+			   phy_modes(state->interface), ERR_PTR(err));
 
-	err = phy_set_mode_ext(mac->serdes_phy, PHY_MODE_ETHERNET, state->interface);
-	if (err)
-		netdev_err(mac->net_dev, "phy_set_mode_ext() = %d\n", err);
+	/* Configure any optional retimers for the current interface mode */
+	for (i = mac->num_lanes; i < mac->num_phys; i++)
+		phy_set_mode_ext(mac->phys[i], PHY_MODE_ETHERNET,
+				 state->interface);
 }
 
 static void dpaa2_mac_link_up(struct phylink_config *config,
@@ -224,6 +252,20 @@
 	if (err)
 		netdev_err(mac->net_dev, "%s: dpmac_set_link_state() = %d\n",
 			   __func__, err);
+
+	/* The resolved C73 interface is only known at mac_link_up() time.
+	 * It is not necessary to configure any retimers here, since they
+	 * are incompatible with C73 and thus will be absent in any real life
+	 * backplane setup (they block the transmission of DME base pages).
+	 */
+	if (phylink_autoneg_c73(mode)) {
+		err = dpmac_set_protocol(mac->mc_io, 0, mac->mc_dev->mc_handle,
+					 dpmac_eth_if_mode(interface));
+		if (err) {
+			netdev_err(mac->net_dev, "dpmac_set_protocol(%s) = %pe\n",
+				   phy_modes(interface), ERR_PTR(err));
+		}
+	}
 }
 
 static void dpaa2_mac_link_down(struct phylink_config *config,
@@ -262,7 +304,7 @@
 		return 0;
 	}
 
-	pcs = lynx_pcs_create_fwnode(node);
+	pcs = lynx_pcs_create_fwnode(node, mac->phys, mac->num_lanes);
 	fwnode_handle_put(node);
 
 	if (pcs == ERR_PTR(-EPROBE_DEFER)) {
@@ -298,54 +340,21 @@
 
 static void dpaa2_mac_set_supported_interfaces(struct dpaa2_mac *mac)
 {
-	int intf, err;
+	struct phylink_config *cfg = &mac->phylink_config;
 
-	/* We support the current interface mode, and if we have a PCS
-	 * similar interface modes that do not require the SerDes lane to be
-	 * reconfigured.
+	/* We support the current interface mode, and if we have a PCS, the
+	 * interface modes that the SerDes allows switching between.
 	 */
-	__set_bit(mac->if_mode, mac->phylink_config.supported_interfaces);
-	if (mac->pcs) {
-		switch (mac->if_mode) {
-		case PHY_INTERFACE_MODE_1000BASEX:
-		case PHY_INTERFACE_MODE_SGMII:
-			__set_bit(PHY_INTERFACE_MODE_1000BASEX,
-				  mac->phylink_config.supported_interfaces);
-			__set_bit(PHY_INTERFACE_MODE_SGMII,
-				  mac->phylink_config.supported_interfaces);
-			break;
-
-		default:
-			break;
-		}
-	}
-
-	if (!mac->serdes_phy)
-		return;
-
-	/* In case we have access to the SerDes phy/lane, then ask the SerDes
-	 * driver what interfaces are supported based on the current PLL
-	 * configuration.
-	 */
-	for (intf = 0; intf < PHY_INTERFACE_MODE_MAX; intf++) {
-		if (intf == PHY_INTERFACE_MODE_NA)
-			continue;
-
-		err = phy_validate(mac->serdes_phy, PHY_MODE_ETHERNET, intf, NULL);
-		if (err)
-			continue;
-
-		__set_bit(intf, mac->phylink_config.supported_interfaces);
-	}
+	__set_bit(mac->if_mode, cfg->supported_interfaces);
+	if (mac->pcs)
+		lynx_pcs_set_supported_interfaces(mac->pcs, mac->if_mode,
+						  cfg->supported_interfaces);
 }
 
 void dpaa2_mac_start(struct dpaa2_mac *mac)
 {
 	ASSERT_RTNL();
 
-	if (mac->serdes_phy)
-		phy_power_on(mac->serdes_phy);
-
 	phylink_start(mac->phylink);
 }
 
@@ -354,16 +363,68 @@
 	ASSERT_RTNL();
 
 	phylink_stop(mac->phylink);
+}
+
+static int dpaa2_mac_get_phys(struct dpaa2_mac *mac)
+{
+	struct device_node *dn = to_of_node(mac->fw_node);
+	struct device *dev = &mac->mc_dev->dev;
+	struct phy *phy;
+	size_t i;
+	int err;
+	u32 val;
+
+	err = of_count_phandle_with_args(dn, "phys", "#phy-cells");
+	if (err <= 0) {
+		mac->num_phys = 0;
+		return 0;
+	}
+	mac->num_phys = err;
+
+	if (fwnode_property_read_u32(mac->fw_node, "num-lanes", &val))
+		mac->num_lanes = 1;
+	else
+		mac->num_lanes = val;
+
+	mac->phys = devm_kcalloc(dev, mac->num_phys, sizeof(struct phy *),
+				 GFP_KERNEL);
+	if (!mac->phys)
+		return -ENOMEM;
+
+	for (i = 0; i < mac->num_phys; i++) {
+		phy = devm_of_phy_get_by_index(dev, dn, i);
+		if (IS_ERR(phy))
+			return PTR_ERR(phy);
 
-	if (mac->serdes_phy)
-		phy_power_off(mac->serdes_phy);
+		mac->phys[i] = phy;
+	}
+
+	/* PHYs 0 .. num_lanes-1 are SerDes lanes and are managed by the PCS.
+	 * The rest up to num_phys are optional retimers or other PHYs that
+	 * may be in the signal path and need to be managed. These will be
+	 * managed by the dpaa2-mac driver.
+	 */
+	for (i = mac->num_lanes; i < mac->num_phys; i++) {
+		err = phy_init(mac->phys[i]);
+		if (err)
+			return err;
+	}
+
+	return 0;
+}
+
+static void dpaa2_mac_put_phys(struct dpaa2_mac *mac)
+{
+	size_t i;
+
+	for (i = mac->num_lanes; i < mac->num_phys; i++)
+		phy_exit(mac->phys[i]);
 }
 
 int dpaa2_mac_connect(struct dpaa2_mac *mac)
 {
 	struct net_device *net_dev = mac->net_dev;
 	struct fwnode_handle *dpmac_node;
-	struct phy *serdes_phy = NULL;
 	struct phylink *phylink;
 	int err;
 
@@ -383,16 +444,10 @@
 	if (mac->features & DPAA2_MAC_FEATURE_PROTOCOL_CHANGE &&
 	    !phy_interface_mode_is_rgmii(mac->if_mode) &&
 	    is_of_node(dpmac_node)) {
-		serdes_phy = of_phy_get(to_of_node(dpmac_node), NULL);
-
-		if (serdes_phy == ERR_PTR(-ENODEV))
-			serdes_phy = NULL;
-		else if (IS_ERR(serdes_phy))
-			return PTR_ERR(serdes_phy);
-		else
-			phy_init(serdes_phy);
+		err = dpaa2_mac_get_phys(mac);
+		if (err)
+			return err;
 	}
-	mac->serdes_phy = serdes_phy;
 
 	/* The MAC does not have the capability to add RGMII delays so
 	 * error out if the interface mode requests them and there is no PHY
@@ -420,7 +475,7 @@
 
 	mac->phylink_config.mac_capabilities = MAC_SYM_PAUSE | MAC_ASYM_PAUSE |
 		MAC_10FD | MAC_100FD | MAC_1000FD | MAC_2500FD | MAC_5000FD |
-		MAC_10000FD | MAC_25000FD;
+		MAC_10000FD | MAC_25000FD | MAC_40000FD | MAC_100000FD;
 
 	dpaa2_mac_set_supported_interfaces(mac);
 
@@ -459,8 +514,7 @@
 
 	phylink_destroy(mac->phylink);
 	dpaa2_pcs_destroy(mac);
-	of_phy_put(mac->serdes_phy);
-	mac->serdes_phy = NULL;
+	dpaa2_mac_put_phys(mac);
 }
 
 int dpaa2_mac_open(struct dpaa2_mac *mac)
@@ -504,6 +558,8 @@
 	mac->fw_node = fw_node;
 	net_dev->dev.of_node = to_of_node(mac->fw_node);
 
+	dpaa2_mac_setup_stats(mac);
+
 	return 0;
 
 err_close_dpmac:
@@ -515,6 +571,8 @@
 {
 	struct fsl_mc_device *dpmac_dev = mac->mc_dev;
 
+	dpaa2_mac_clear_stats(mac);
+
 	dpmac_close(mac->mc_io, 0, dpmac_dev->mc_handle);
 	if (mac->fw_node)
 		fwnode_handle_put(mac->fw_node);
@@ -553,6 +611,80 @@
 
 #define DPAA2_MAC_NUM_STATS	ARRAY_SIZE(dpaa2_mac_ethtool_stats)
 
+#define DPAA2_MAC_STATS_INDEX_DMA_SIZE (DPAA2_MAC_NUM_STATS * sizeof(u32))
+#define DPAA2_MAC_STATS_VALUE_DMA_SIZE (DPAA2_MAC_NUM_STATS * sizeof(u64))
+
+static void dpaa2_mac_setup_stats(struct dpaa2_mac *mac)
+{
+	struct device *dev = mac->net_dev->dev.parent;
+	u32 *cnt_idx;
+	int i;
+
+	if (!(mac->features & DPAA2_MAC_FEATURE_STATS_BUNDLE))
+		return;
+
+	mac->cnt_idx_dma_mem = kzalloc(DPAA2_MAC_STATS_INDEX_DMA_SIZE, GFP_KERNEL);
+	if (!mac->cnt_idx_dma_mem)
+		goto out;
+
+	mac->cnt_values_dma_mem = kzalloc(DPAA2_MAC_STATS_VALUE_DMA_SIZE, GFP_KERNEL);
+	if (!mac->cnt_values_dma_mem)
+		goto err_alloc_values;
+
+	cnt_idx = mac->cnt_idx_dma_mem;
+	for (i = 0; i < DPAA2_MAC_NUM_STATS; i++)
+		*cnt_idx++ = cpu_to_le32((u32)i);
+
+	mac->cnt_idx_iova = dma_map_single(dev, mac->cnt_idx_dma_mem,
+					   DPAA2_MAC_STATS_INDEX_DMA_SIZE,
+					   DMA_TO_DEVICE);
+	if (dma_mapping_error(dev, mac->cnt_idx_iova))
+		goto err_dma_map_idx;
+
+	mac->cnt_values_iova = dma_map_single(dev, mac->cnt_values_dma_mem,
+					      DPAA2_MAC_STATS_VALUE_DMA_SIZE,
+					      DMA_FROM_DEVICE);
+	if (dma_mapping_error(dev, mac->cnt_values_iova))
+		goto err_dma_map_values;
+
+	return;
+
+err_dma_map_values:
+	dma_unmap_single(dev, mac->cnt_idx_iova, DPAA2_MAC_STATS_INDEX_DMA_SIZE,
+			 DMA_TO_DEVICE);
+err_dma_map_idx:
+	kfree(mac->cnt_values_dma_mem);
+err_alloc_values:
+	kfree(mac->cnt_idx_dma_mem);
+out:
+	mac->cnt_idx_dma_mem = NULL;
+	mac->cnt_values_dma_mem = NULL;
+}
+
+static void dpaa2_mac_clear_stats(struct dpaa2_mac *mac)
+{
+	struct device *dev = mac->net_dev->dev.parent;
+
+	if (!(mac->features & DPAA2_MAC_FEATURE_STATS_BUNDLE))
+		return;
+
+	if (mac->cnt_idx_dma_mem) {
+		dma_unmap_single(dev, mac->cnt_idx_iova,
+				 DPAA2_MAC_STATS_INDEX_DMA_SIZE,
+				 DMA_TO_DEVICE);
+		kfree(mac->cnt_idx_dma_mem);
+		mac->cnt_idx_dma_mem = NULL;
+	}
+
+	if (mac->cnt_values_dma_mem) {
+		dma_unmap_single(dev, mac->cnt_values_iova,
+				 DPAA2_MAC_STATS_VALUE_DMA_SIZE,
+				 DMA_FROM_DEVICE);
+		kfree(mac->cnt_values_dma_mem);
+		mac->cnt_values_dma_mem = NULL;
+	}
+}
+
 int dpaa2_mac_get_sset_count(void)
 {
 	return DPAA2_MAC_NUM_STATS;
@@ -571,10 +703,37 @@
 
 void dpaa2_mac_get_ethtool_stats(struct dpaa2_mac *mac, u64 *data)
 {
+	struct device *dev = mac->net_dev->dev.parent;
 	struct fsl_mc_device *dpmac_dev = mac->mc_dev;
+	u64 *cnt_values;
 	int i, err;
 	u64 value;
 
+	if (!(mac->features & DPAA2_MAC_FEATURE_STATS_BUNDLE))
+		goto fallback;
+
+	if (!mac->cnt_idx_dma_mem || !mac->cnt_values_dma_mem)
+		goto fallback;
+
+	err = dpmac_get_statistics(mac->mc_io, 0, dpmac_dev->mc_handle,
+				   mac->cnt_idx_iova, mac->cnt_values_iova,
+				   DPAA2_MAC_NUM_STATS);
+	if (err)
+		goto fallback;
+
+	dma_sync_single_for_cpu(dev, mac->cnt_values_iova,
+				DPAA2_MAC_STATS_VALUE_DMA_SIZE,
+				DMA_FROM_DEVICE);
+
+	cnt_values = mac->cnt_values_dma_mem;
+	for (i = 0; i < DPAA2_MAC_NUM_STATS; i++)
+		*(data + i) = le64_to_cpu(*cnt_values++);
+
+	return;
+
+fallback:
+
+	/* Fallback and retrieve each counter one by one */
 	for (i = 0; i < DPAA2_MAC_NUM_STATS; i++) {
 		err = dpmac_get_counter(mac->mc_io, 0, dpmac_dev->mc_handle,
 					i, &value);
@@ -587,3 +746,25 @@
 		*(data + i) = value;
 	}
 }
+
+void dpaa2_mac_driver_attach(struct fsl_mc_device *dpmac_dev)
+{
+	struct device_driver *drv = driver_find("fsl_dpaa2_mac", &fsl_mc_bus_type);
+	struct device *dev = &dpmac_dev->dev;
+	int err;
+
+	if (dev && dev->driver == NULL && drv) {
+		err = device_attach(dev);
+		if (err && err != -EAGAIN)
+			dev_err(dev, "Error in attaching the fsl_dpaa2_mac driver\n");
+	}
+}
+
+void dpaa2_mac_driver_detach(struct fsl_mc_device *dpmac_dev)
+{
+	struct device_driver *drv = driver_find("fsl_dpaa2_mac", &fsl_mc_bus_type);
+	struct device *dev = &dpmac_dev->dev;
+
+	if (dev && dev->driver == drv)
+		device_release_driver(dev);
+}
