--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/ethernet/freescale/dpaa2/dpaa2-mac.h	2026-04-08 18:20:08.150292422 +0200
+++ ./drivers/net/ethernet/freescale/dpaa2/dpaa2-mac.h	2026-04-08 16:21:07.516545404 +0200
@@ -1,5 +1,5 @@
 /* SPDX-License-Identifier: (GPL-2.0+ OR BSD-3-Clause) */
-/* Copyright 2019 NXP */
+/* Copyright 2019, 2024 NXP */
 #ifndef DPAA2_MAC_H
 #define DPAA2_MAC_H
 
@@ -7,7 +7,6 @@
 #include <linux/of_mdio.h>
 #include <linux/of_net.h>
 #include <linux/phylink.h>
-
 #include "dpmac.h"
 #include "dpmac-cmd.h"
 
@@ -22,12 +21,20 @@
 
 	struct phylink_config phylink_config;
 	struct phylink *phylink;
+	struct ethtool_link_ksettings kset;
 	phy_interface_t if_mode;
 	enum dpmac_link_type if_link_type;
 	struct phylink_pcs *pcs;
 	struct fwnode_handle *fw_node;
 
-	struct phy *serdes_phy;
+	struct phy **phys;
+	size_t num_phys;
+	size_t num_lanes;
+
+	int phy_req_state;
+	u32 *cnt_idx_dma_mem;
+	u64 *cnt_values_dma_mem;
+	dma_addr_t cnt_idx_iova, cnt_values_iova;
 };
 
 static inline bool dpaa2_mac_is_type_phy(struct dpaa2_mac *mac)
@@ -57,4 +64,8 @@
 
 void dpaa2_mac_stop(struct dpaa2_mac *mac);
 
+void dpaa2_mac_driver_attach(struct fsl_mc_device *dpmac_dev);
+
+void dpaa2_mac_driver_detach(struct fsl_mc_device *dpmac_dev);
+
 #endif /* DPAA2_MAC_H */
