--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/ethernet/freescale/fec_ptp.c	2026-04-08 18:20:08.154970095 +0200
+++ ./drivers/net/ethernet/freescale/fec_ptp.c	2026-04-08 16:21:07.523763279 +0200
@@ -134,6 +134,9 @@
 		return 0;
 	}
 
+	fep->pps_channel = DEFAULT_PPS_CHANNEL;
+	fep->reload_period = PPS_OUPUT_RELOAD_PERIOD;
+
 	if (enable) {
 		/* clear capture or output compare interrupt status if have.
 		 */
