--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/ethernet/freescale/fman/fman_memac.c	2026-04-08 18:20:08.155970123 +0200
+++ ./drivers/net/ethernet/freescale/fman/fman_memac.c	2026-04-08 16:21:07.524379519 +0200
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: BSD-3-Clause OR GPL-2.0-or-later
 /*
  * Copyright 2008 - 2015 Freescale Semiconductor Inc.
+ * Copyright 2020 Puresoftware Ltd.
  */
 
 #define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
@@ -9,6 +10,7 @@
 #include "fman.h"
 #include "mac.h"
 
+#include <linux/acpi.h>
 #include <linux/slab.h>
 #include <linux/io.h>
 #include <linux/pcs-lynx.h>
@@ -290,10 +292,10 @@
 	struct memac_cfg *memac_drv_param;
 	void *fm;
 	struct fman_rev_info fm_rev_info;
-	struct phy *serdes;
 	struct phylink_pcs *sgmii_pcs;
 	struct phylink_pcs *qsgmii_pcs;
 	struct phylink_pcs *xfi_pcs;
+	struct phylink_pcs *c73_pcs;
 	bool allmulti_enabled;
 	bool rgmii_no_half_duplex;
 };
@@ -524,33 +526,6 @@
 	memac->unicast_addr_hash = NULL;
 }
 
-static int memac_enable(struct fman_mac *memac)
-{
-	int ret;
-
-	ret = phy_init(memac->serdes);
-	if (ret) {
-		dev_err(memac->dev_id->dev,
-			"could not initialize serdes: %pe\n", ERR_PTR(ret));
-		return ret;
-	}
-
-	ret = phy_power_on(memac->serdes);
-	if (ret) {
-		dev_err(memac->dev_id->dev,
-			"could not power on serdes: %pe\n", ERR_PTR(ret));
-		phy_exit(memac->serdes);
-	}
-
-	return ret;
-}
-
-static void memac_disable(struct fman_mac *memac)
-{
-	phy_power_off(memac->serdes);
-	phy_exit(memac->serdes);
-}
-
 static int memac_set_promiscuous(struct fman_mac *memac, bool new_val)
 {
 	struct memac_regs __iomem *regs = memac->regs;
@@ -651,9 +626,11 @@
 		return IF_MODE_GMII | IF_MODE_RGMII;
 	case PHY_INTERFACE_MODE_SGMII:
 	case PHY_INTERFACE_MODE_1000BASEX:
+	case PHY_INTERFACE_MODE_1000BASEKX:
 	case PHY_INTERFACE_MODE_QSGMII:
 		return IF_MODE_GMII;
 	case PHY_INTERFACE_MODE_10GBASER:
+	case PHY_INTERFACE_MODE_10GKR:
 		return IF_MODE_10G;
 	default:
 		WARN_ON_ONCE(1);
@@ -666,6 +643,9 @@
 {
 	struct fman_mac *memac = fman_config_to_mac(config)->fman_mac;
 
+	if (phylink_autoneg_c73(config->cfg_link_an_mode))
+		return memac->c73_pcs;
+
 	switch (iface) {
 	case PHY_INTERFACE_MODE_SGMII:
 	case PHY_INTERFACE_MODE_1000BASEX:
@@ -679,23 +659,6 @@
 	}
 }
 
-static int memac_prepare(struct phylink_config *config, unsigned int mode,
-			 phy_interface_t iface)
-{
-	struct fman_mac *memac = fman_config_to_mac(config)->fman_mac;
-
-	switch (iface) {
-	case PHY_INTERFACE_MODE_SGMII:
-	case PHY_INTERFACE_MODE_1000BASEX:
-	case PHY_INTERFACE_MODE_QSGMII:
-	case PHY_INTERFACE_MODE_10GBASER:
-		return phy_set_mode_ext(memac->serdes, PHY_MODE_ETHERNET,
-					iface);
-	default:
-		return 0;
-	}
-}
-
 static void memac_mac_config(struct phylink_config *config, unsigned int mode,
 			     const struct phylink_link_state *state)
 {
@@ -703,6 +666,12 @@
 	struct memac_regs __iomem *regs = mac_dev->fman_mac->regs;
 	u32 tmp = ioread32be(&regs->if_mode);
 
+	/* state->interface is only known at mac_link_up() time with C73
+	 * autoneg, so delay programming the IF_MODE register
+	 */
+	if (phylink_autoneg_c73(mode))
+		return;
+
 	tmp &= ~(IF_MODE_MASK | IF_MODE_RGMII);
 	tmp |= memac_if_mode(state->interface);
 	if (phylink_autoneg_inband(mode))
@@ -778,7 +747,6 @@
 static const struct phylink_mac_ops memac_mac_ops = {
 	.validate = memac_validate,
 	.mac_select_pcs = memac_select_pcs,
-	.mac_prepare = memac_prepare,
 	.mac_config = memac_mac_config,
 	.mac_link_up = memac_link_up,
 	.mac_link_down = memac_link_down,
@@ -986,6 +954,7 @@
 {
 	free_init_resources(memac);
 
+	pcs_put(memac->c73_pcs);
 	pcs_put(memac->sgmii_pcs);
 	pcs_put(memac->qsgmii_pcs);
 	pcs_put(memac->xfi_pcs);
@@ -1035,44 +1004,78 @@
 	return memac;
 }
 
-static struct phylink_pcs *memac_pcs_create(struct device_node *mac_node,
-					    int index)
+static struct phylink_pcs *memac_pcs_create(struct fwnode_handle *mac_fwnode,
+					    int index, struct phy *serdes)
 {
-	struct device_node *node;
+	size_t num_phys = serdes ? 1 : 0;
+	struct fwnode_handle *node;
 	struct phylink_pcs *pcs;
 
-	node = of_parse_phandle(mac_node, "pcsphy-handle", index);
+	node = fwnode_find_reference(mac_fwnode, "pcsphy-handle", index);
 	if (!node)
 		return ERR_PTR(-ENODEV);
 
-	pcs = lynx_pcs_create_fwnode(of_fwnode_handle(node));
-	of_node_put(node);
+	pcs = lynx_pcs_create_fwnode(node, &serdes, num_phys);
+	fwnode_handle_put(node);
 
 	return pcs;
 }
 
-static bool memac_supports(struct mac_device *mac_dev, phy_interface_t iface)
+static int memac_get_optional_pcs(struct mac_device *mac_dev,
+				  struct fwnode_handle *mac_fwnode,
+				  struct phylink_pcs **pcsp, const char *name,
+				  struct phy *serdes)
 {
-	/* If there's no serdes device, assume that it's been configured for
-	 * whatever the default interface mode is.
-	 */
-	if (!mac_dev->fman_mac->serdes)
-		return mac_dev->phy_if == iface;
-	/* Otherwise, ask the serdes */
-	return !phy_validate(mac_dev->fman_mac->serdes, PHY_MODE_ETHERNET,
-			     iface, NULL);
+	struct phylink_pcs *pcs;
+	int err;
+
+	err = fwnode_property_match_string(mac_fwnode, "pcs-handle-names",
+					   name);
+	if (err >= 0) {
+		pcs = memac_pcs_create(mac_fwnode, err, serdes);
+		if (IS_ERR(pcs)) {
+			err = PTR_ERR(pcs);
+			return dev_err_probe(mac_dev->dev, err,
+					     "missing %s pcs\n", name);
+		}
+
+		*pcsp = pcs;
+	} else if (err != -EINVAL && err != -ENODATA) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int memac_get_default_pcs(struct mac_device *mac_dev,
+				 struct fwnode_handle *mac_fwnode,
+				 struct phylink_pcs **pcsp, struct phy *serdes)
+{
+	struct phylink_pcs *pcs;
+	int err;
+
+	pcs = memac_pcs_create(mac_fwnode, 0, serdes);
+	if (IS_ERR(pcs)) {
+		err = PTR_ERR(pcs);
+		return dev_err_probe(mac_dev->dev, err, "missing default pcs\n");
+	}
+
+	*pcsp = pcs;
+
+	return 0;
 }
 
 int memac_initialization(struct mac_device *mac_dev,
-			 struct device_node *mac_node,
 			 struct fman_mac_params *params)
 {
 	int			 err;
-	struct device_node      *fixed;
-	struct phylink_pcs	*pcs;
 	struct fman_mac		*memac;
 	unsigned long		 capabilities;
 	unsigned long		*supported;
+	struct phy		*serdes = NULL;
+	struct device		*dev = mac_dev->dev;
+	struct fwnode_handle	*mac_fwnode = dev->fwnode;
+	struct fwnode_handle	*fixed;
 
 	/* The internal connection to the serdes is XGMII, but this isn't
 	 * really correct for the phy mode (which is the external connection).
@@ -1091,8 +1094,6 @@
 	mac_dev->set_allmulti		= memac_set_allmulti;
 	mac_dev->set_tstamp		= memac_set_tstamp;
 	mac_dev->set_multi		= fman_set_multi;
-	mac_dev->enable			= memac_enable;
-	mac_dev->disable		= memac_disable;
 
 	mac_dev->fman_mac = memac_config(mac_dev, params);
 	if (!mac_dev->fman_mac)
@@ -1102,98 +1103,82 @@
 	memac->memac_drv_param->max_frame_length = fman_get_max_frm();
 	memac->memac_drv_param->reset_on_init = true;
 
-	err = of_property_match_string(mac_node, "pcs-handle-names", "xfi");
-	if (err >= 0) {
-		memac->xfi_pcs = memac_pcs_create(mac_node, err);
-		if (IS_ERR(memac->xfi_pcs)) {
-			err = PTR_ERR(memac->xfi_pcs);
-			dev_err_probe(mac_dev->dev, err, "missing xfi pcs\n");
+	if (is_of_node(mac_fwnode)) {
+		serdes = devm_of_phy_optional_get(mac_dev->dev,
+						  to_of_node(mac_fwnode),
+						  NULL);
+		if (!serdes) {
+			dev_dbg(mac_dev->dev, "could not get (optional) serdes\n");
+		} else if (IS_ERR(serdes)) {
+			err = PTR_ERR(serdes);
 			goto _return_fm_mac_free;
 		}
-	} else if (err != -EINVAL && err != -ENODATA) {
-		goto _return_fm_mac_free;
 	}
 
-	err = of_property_match_string(mac_node, "pcs-handle-names", "qsgmii");
-	if (err >= 0) {
-		memac->qsgmii_pcs = memac_pcs_create(mac_node, err);
-		if (IS_ERR(memac->qsgmii_pcs)) {
-			err = PTR_ERR(memac->qsgmii_pcs);
-			dev_err_probe(mac_dev->dev, err,
-				      "missing qsgmii pcs\n");
-			goto _return_fm_mac_free;
-		}
-	} else if (err != -EINVAL && err != -ENODATA) {
+	err = memac_get_optional_pcs(mac_dev, mac_fwnode, &memac->xfi_pcs, "xfi",
+				     serdes);
+	if (err)
 		goto _return_fm_mac_free;
-	}
 
-	/* For compatibility, if pcs-handle-names is missing, we assume this
-	 * phy is the first one in pcsphy-handle
-	 */
-	err = of_property_match_string(mac_node, "pcs-handle-names", "sgmii");
-	if (err == -EINVAL || err == -ENODATA)
-		pcs = memac_pcs_create(mac_node, 0);
-	else if (err < 0)
+	err = memac_get_optional_pcs(mac_dev, mac_fwnode, &memac->qsgmii_pcs,
+				     "qsgmii", serdes);
+	if (err)
 		goto _return_fm_mac_free;
-	else
-		pcs = memac_pcs_create(mac_node, err);
 
-	if (IS_ERR(pcs)) {
-		err = PTR_ERR(pcs);
-		dev_err_probe(mac_dev->dev, err, "missing pcs\n");
+	err = memac_get_optional_pcs(mac_dev, mac_fwnode, &memac->sgmii_pcs,
+				     "sgmii", serdes);
+	if (err)
 		goto _return_fm_mac_free;
-	}
 
-	/* If err is set here, it means that pcs-handle-names was missing above
-	 * (and therefore that xfi_pcs cannot be set). If we are defaulting to
-	 * XGMII, assume this is for XFI. Otherwise, assume it is for SGMII.
-	 */
-	if (err && mac_dev->phy_if == PHY_INTERFACE_MODE_10GBASER)
-		memac->xfi_pcs = pcs;
-	else
-		memac->sgmii_pcs = pcs;
-
-	memac->serdes = devm_of_phy_optional_get(mac_dev->dev, mac_node,
-						 "serdes");
-	if (!memac->serdes) {
-		dev_dbg(mac_dev->dev, "could not get (optional) serdes\n");
-	} else if (IS_ERR(memac->serdes)) {
-		err = PTR_ERR(memac->serdes);
+	err = memac_get_optional_pcs(mac_dev, mac_fwnode, &memac->c73_pcs,
+				     "c73", serdes);
+	if (err)
 		goto _return_fm_mac_free;
-	}
 
-	/* TODO: The following interface modes are supported by (some) hardware
-	 * but not by this driver:
-	 * - 1000BASE-KX
-	 * - 10GBASE-KR
-	 * - XAUI/HiGig
+	/* For compatibility, if pcs-handle-names is missing, we assume this
+	 * phy is the first one in pcsphy-handle
 	 */
+	if (mac_dev->phy_if == PHY_INTERFACE_MODE_10GBASER && !memac->xfi_pcs) {
+		err = memac_get_default_pcs(mac_dev, mac_fwnode,
+					    &memac->xfi_pcs, serdes);
+		if (err)
+			goto _return_fm_mac_free;
+	} else if (!memac->sgmii_pcs) {
+		err = memac_get_default_pcs(mac_dev, mac_fwnode,
+					    &memac->sgmii_pcs, serdes);
+		if (err)
+			goto _return_fm_mac_free;
+	}
+
 	supported = mac_dev->phylink_config.supported_interfaces;
 
 	/* Note that half duplex is only supported on 10/100M interfaces. */
 
-	if (memac->sgmii_pcs &&
-	    (memac_supports(mac_dev, PHY_INTERFACE_MODE_SGMII) ||
-	     memac_supports(mac_dev, PHY_INTERFACE_MODE_1000BASEX))) {
-		__set_bit(PHY_INTERFACE_MODE_SGMII, supported);
-		__set_bit(PHY_INTERFACE_MODE_1000BASEX, supported);
-	}
-
-	if (memac->sgmii_pcs &&
-	    memac_supports(mac_dev, PHY_INTERFACE_MODE_2500BASEX))
-		__set_bit(PHY_INTERFACE_MODE_2500BASEX, supported);
+	if (memac->sgmii_pcs)
+		lynx_pcs_set_supported_interfaces(memac->sgmii_pcs,
+						  mac_dev->phy_if,
+						  supported);
 
 	if (memac->qsgmii_pcs &&
-	    memac_supports(mac_dev, PHY_INTERFACE_MODE_QSGMII))
-		__set_bit(PHY_INTERFACE_MODE_QSGMII, supported);
+	    mac_dev->phy_if == PHY_INTERFACE_MODE_QSGMII)
+		lynx_pcs_set_supported_interfaces(memac->qsgmii_pcs,
+						  mac_dev->phy_if,
+						  supported);
 	else if (mac_dev->phy_if == PHY_INTERFACE_MODE_QSGMII)
 		dev_warn(mac_dev->dev, "no QSGMII pcs specified\n");
 
-	if (memac->xfi_pcs &&
-	    memac_supports(mac_dev, PHY_INTERFACE_MODE_10GBASER)) {
-		__set_bit(PHY_INTERFACE_MODE_10GBASER, supported);
-	} else {
-		/* From what I can tell, no 10g macs support RGMII. */
+	if (memac->xfi_pcs)
+		lynx_pcs_set_supported_interfaces(memac->xfi_pcs,
+						  mac_dev->phy_if,
+						  supported);
+
+	if (memac->c73_pcs)
+		lynx_pcs_set_supported_interfaces(memac->c73_pcs,
+						  mac_dev->phy_if,
+						  supported);
+
+	if (!memac->sgmii_pcs && !memac->qsgmii_pcs && !memac->xfi_pcs &&
+	    !memac->c73_pcs) {
 		phy_interface_set_rgmii(supported);
 		__set_bit(PHY_INTERFACE_MODE_MII, supported);
 	}
@@ -1229,13 +1214,13 @@
 	 * fixed link or if we are using MII or RGMII, since those
 	 * configurations modes don't use in-band autonegotiation.
 	 */
-	fixed = of_get_child_by_name(mac_node, "fixed-link");
-	if (!fixed && !of_property_read_bool(mac_node, "fixed-link") &&
-	    !of_property_read_bool(mac_node, "managed") &&
+	fixed = fwnode_get_named_child_node(mac_fwnode, "fixed-link");
+	if (!fixed && !fwnode_property_read_bool(mac_fwnode, "fixed-link") &&
+	    !fwnode_property_read_bool(mac_fwnode, "managed") &&
 	    mac_dev->phy_if != PHY_INTERFACE_MODE_MII &&
 	    !phy_interface_mode_is_rgmii(mac_dev->phy_if))
 		mac_dev->phylink_config.ovr_an_inband = true;
-	of_node_put(fixed);
+	fwnode_handle_put(fixed);
 
 	err = memac_init(mac_dev->fman_mac);
 	if (err < 0)
