--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/ethernet/mscc/ocelot_mm.c	2026-04-08 18:20:08.326974887 +0200
+++ ./drivers/net/ethernet/mscc/ocelot_mm.c	2026-04-08 16:21:07.762351052 +0200
@@ -87,20 +87,30 @@
 		       QSYS_PREEMPTION_CFG, port);
 }
 
-void ocelot_port_change_fp(struct ocelot *ocelot, int port,
-			   unsigned long preemptible_tcs)
+int ocelot_port_change_fp(struct ocelot *ocelot, int port,
+			  unsigned long preemptible_tcs)
 {
+	struct ocelot_port *ocelot_port = ocelot->ports[port];
 	struct ocelot_mm_state *mm = &ocelot->mm[port];
 
 	lockdep_assert_held(&ocelot->fwd_domain_lock);
 
+	if (ocelot_port->cut_thru_selected_by_user & preemptible_tcs) {
+		dev_err(ocelot->dev,
+			"A traffic class cannot be preemptible and cut-through at the same time.\n");
+		return -EBUSY;
+	}
+
 	if (mm->preemptible_tcs == preemptible_tcs)
-		return;
+		return 0;
 
 	mm->preemptible_tcs = preemptible_tcs;
 
 	ocelot_port_update_active_preemptible_tcs(ocelot, port);
+
+	return 0;
 }
+EXPORT_SYMBOL_GPL(ocelot_port_change_fp);
 
 static void ocelot_mm_update_port_status(struct ocelot *ocelot, int port)
 {
