--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/pcs/pcs-lynx.c	2026-04-08 18:20:08.500472369 +0200
+++ ./drivers/net/pcs/pcs-lynx.c	2026-04-08 16:21:07.973963623 +0200
@@ -4,10 +4,16 @@
  */
 
 #include <linux/mdio.h>
+#include <linux/phy/phy.h>
 #include <linux/phylink.h>
 #include <linux/pcs-lynx.h>
 #include <linux/property.h>
 
+#include "mtip_backplane.h"
+
+#define PRIMARY_LANE			0
+#define MAX_NUM_LANES			4
+
 #define SGMII_CLOCK_PERIOD_NS		8 /* PCS is clocked at 125 MHz */
 #define LINK_TIMER_VAL(ns)		((u32)((ns) / SGMII_CLOCK_PERIOD_NS))
 
@@ -23,6 +29,10 @@
 struct lynx_pcs {
 	struct phylink_pcs pcs;
 	struct mdio_device *mdio;
+	size_t num_phys;
+	struct phy *serdes[MAX_NUM_LANES];
+	struct mtip_backplane *anlt[MAX_NUM_LANES];
+	enum mtip_model model;
 };
 
 enum sgmii_speed {
@@ -85,6 +95,9 @@
 {
 	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
 
+	if (phylink_autoneg_c73(pcs->cfg_link_an_mode))
+		return mtip_backplane_get_state(lynx->anlt[PRIMARY_LANE], state);
+
 	switch (state->interface) {
 	case PHY_INTERFACE_MODE_1000BASEX:
 	case PHY_INTERFACE_MODE_SGMII:
@@ -95,9 +108,11 @@
 		lynx_pcs_get_state_2500basex(lynx->mdio, state);
 		break;
 	case PHY_INTERFACE_MODE_USXGMII:
+	case PHY_INTERFACE_MODE_10G_QXGMII:
 		lynx_pcs_get_state_usxgmii(lynx->mdio, state);
 		break;
 	case PHY_INTERFACE_MODE_10GBASER:
+	case PHY_INTERFACE_MODE_25GBASER:
 		phylink_mii_c45_pcs_get_state(lynx->mdio, state);
 		break;
 	default:
@@ -150,6 +165,7 @@
 }
 
 static int lynx_pcs_config_usxgmii(struct mdio_device *pcs,
+				   phy_interface_t interface,
 				   const unsigned long *advertising,
 				   unsigned int neg_mode)
 {
@@ -157,7 +173,8 @@
 	int addr = pcs->addr;
 
 	if (neg_mode != PHYLINK_PCS_NEG_INBAND_ENABLED) {
-		dev_err(&pcs->dev, "USXGMII only supports in-band AN for now\n");
+		dev_err(&pcs->dev, "%s only supports in-band AN for now\n",
+			phy_modes(interface));
 		return -EOPNOTSUPP;
 	}
 
@@ -168,11 +185,42 @@
 				 ADVERTISE_SGMII | ADVERTISE_LPACK);
 }
 
+static int lynx_pcs_config_c73(struct phylink_pcs *pcs, unsigned int neg_mode,
+			       const unsigned long *advertising)
+{
+	bool autoneg = !!(neg_mode & PHYLINK_PCS_NEG_ENABLED);
+	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+
+	if (!lynx->num_phys) {
+		dev_err(&lynx->mdio->dev, "C73 autoneg requires SerDes\n");
+		return -ENODEV;
+	}
+
+	return mtip_backplane_config_aneg(lynx->anlt[PRIMARY_LANE], autoneg,
+					  advertising);
+}
+
 static int lynx_pcs_config(struct phylink_pcs *pcs, unsigned int neg_mode,
 			   phy_interface_t ifmode,
 			   const unsigned long *advertising, bool permit)
 {
 	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+	size_t i;
+	int err;
+
+	if (neg_mode & PHYLINK_PCS_NEG_C73)
+		return lynx_pcs_config_c73(pcs, neg_mode, advertising);
+
+	for (i = 0; i < lynx->num_phys; i++) {
+		err = phy_set_mode_ext(lynx->serdes[i], PHY_MODE_ETHERNET,
+				       ifmode);
+		if (err) {
+			dev_err(&lynx->mdio->dev,
+				"phy_set_mode_ext() failed: %pe\n",
+				ERR_PTR(err));
+			return err;
+		}
+	}
 
 	switch (ifmode) {
 	case PHY_INTERFACE_MODE_1000BASEX:
@@ -188,10 +236,12 @@
 		}
 		break;
 	case PHY_INTERFACE_MODE_USXGMII:
-		return lynx_pcs_config_usxgmii(lynx->mdio, advertising,
+	case PHY_INTERFACE_MODE_10G_QXGMII:
+		return lynx_pcs_config_usxgmii(lynx->mdio, ifmode, advertising,
 					       neg_mode);
 	case PHY_INTERFACE_MODE_10GBASER:
-		/* Nothing to do here for 10GBASER */
+	case PHY_INTERFACE_MODE_25GBASER:
+		/* Nothing to do here for 10GBASER and 25GBASER */
 		break;
 	default:
 		return -EOPNOTSUPP;
@@ -204,6 +254,9 @@
 {
 	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
 
+	if (phylink_autoneg_c73(pcs->cfg_link_an_mode))
+		return mtip_backplane_an_restart(lynx->anlt[PRIMARY_LANE]);
+
 	phylink_mii_c22_pcs_an_restart(lynx->mdio);
 }
 
@@ -306,31 +359,246 @@
 	}
 }
 
+static int lynx_pcs_validate(struct phylink_pcs *pcs, unsigned long *supported,
+			     const struct phylink_link_state *state)
+{
+	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+
+	if (!phylink_autoneg_c73(pcs->cfg_link_an_mode))
+		return 0;
+
+	if (!lynx->num_phys) {
+		linkmode_zero(supported);
+		dev_err(&lynx->mdio->dev, "C73 autoneg requires SerDes\n");
+		return -EINVAL;
+	}
+
+	return mtip_backplane_validate(lynx->serdes[PRIMARY_LANE], supported);
+}
+
+static int lynx_pcs_c73_init(struct phylink_pcs *pcs)
+{
+	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+	size_t i;
+	int err;
+
+	if (!lynx->num_phys) {
+		dev_err(&lynx->mdio->dev, "C73 autoneg requires SerDes\n");
+		return -ENODEV;
+	}
+
+	for (i = 0; i < lynx->num_phys; i++) {
+		lynx->anlt[i] = mtip_backplane_create(lynx->mdio,
+						      lynx->serdes[i],
+						      lynx->model);
+		if (IS_ERR(lynx->anlt[i])) {
+			err = PTR_ERR(lynx->anlt[i]);
+			while (i--)
+				mtip_backplane_destroy(lynx->anlt[i]);
+			return err;
+		}
+	}
+
+	for (i = 1; i < lynx->num_phys; i++)
+		mtip_backplane_add_subordinate(lynx->anlt[PRIMARY_LANE],
+					       lynx->anlt[i]);
+
+	return mtip_backplane_resume(lynx->anlt[PRIMARY_LANE]);
+}
+
+static void lynx_pcs_c73_teardown(struct phylink_pcs *pcs)
+{
+	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+	size_t i;
+	int err;
+
+	err = mtip_backplane_suspend(lynx->anlt[PRIMARY_LANE]);
+	if (err)
+		dev_warn(&lynx->mdio->dev, "Failed to suspend: %pe\n",
+			 ERR_PTR(err));
+
+	for (i = 0; i < lynx->num_phys; i++)
+		if (lynx->anlt[i])
+			mtip_backplane_destroy(lynx->anlt[i]);
+}
+
+static int lynx_pcs_enable(struct phylink_pcs *pcs)
+{
+	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+	size_t i;
+	int err;
+
+	if (phylink_autoneg_c73(pcs->cfg_link_an_mode))
+		return lynx_pcs_c73_init(pcs);
+
+	/* The backplane AN/LT deals with lane power management */
+	for (i = 0; i < lynx->num_phys; i++) {
+		err = phy_power_on(lynx->serdes[i]);
+		if (err)
+			return err;
+	}
+
+	return 0;
+}
+
+static void lynx_pcs_disable(struct phylink_pcs *pcs)
+{
+	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+	size_t i;
+
+	if (phylink_autoneg_c73(pcs->cfg_link_an_mode))
+		return lynx_pcs_c73_teardown(pcs);
+
+	/* The backplane AN/LT deals with lane power management */
+	for (i = 0; i < lynx->num_phys; i++)
+		phy_power_off(lynx->serdes[i]);
+}
+
 static const struct phylink_pcs_ops lynx_pcs_phylink_ops = {
 	.pcs_get_state = lynx_pcs_get_state,
 	.pcs_config = lynx_pcs_config,
 	.pcs_an_restart = lynx_pcs_an_restart,
 	.pcs_link_up = lynx_pcs_link_up,
+	.pcs_validate = lynx_pcs_validate,
+	.pcs_enable = lynx_pcs_enable,
+	.pcs_disable = lynx_pcs_disable,
 };
 
-static struct phylink_pcs *lynx_pcs_create(struct mdio_device *mdio)
+void lynx_pcs_set_supported_interfaces(struct phylink_pcs *pcs,
+				       phy_interface_t default_interface,
+				       unsigned long *supported_interfaces)
+{
+	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+	phy_interface_t iface;
+	int err;
+
+	__set_bit(default_interface, supported_interfaces);
+
+	if (default_interface == PHY_INTERFACE_MODE_1000BASEX ||
+	    default_interface == PHY_INTERFACE_MODE_SGMII) {
+		__set_bit(PHY_INTERFACE_MODE_1000BASEX, supported_interfaces);
+		__set_bit(PHY_INTERFACE_MODE_SGMII, supported_interfaces);
+	}
+
+	if (!lynx->num_phys)
+		return;
+
+	/* In case we have access to the SerDes phy/lane, then ask the SerDes
+	 * driver what interfaces are supported based on the current PLL
+	 * configuration.
+	 */
+	for (iface = 0; iface < PHY_INTERFACE_MODE_MAX; iface++) {
+		if (iface == PHY_INTERFACE_MODE_NA)
+			continue;
+
+		err = phy_validate(lynx->serdes[PRIMARY_LANE],
+				   PHY_MODE_ETHERNET, iface, NULL);
+		if (err)
+			continue;
+
+		__set_bit(iface, supported_interfaces);
+	}
+}
+EXPORT_SYMBOL(lynx_pcs_set_supported_interfaces);
+
+static int lynx_pcs_validate_addr(struct mdio_device *mdiodev,
+				  struct phy *serdes)
+{
+	union phy_status_opts opts1 = {
+		.pcvt_count = {
+			.type = PHY_PCVT_ETHERNET_PCS,
+		},
+	};
+	int i, err;
+
+	err = phy_get_status(serdes, PHY_STATUS_PCVT_COUNT, &opts1);
+	if (err)
+		return err;
+
+	for (i = 0; i < opts1.pcvt_count.num_pcvt; i++) {
+		union phy_status_opts opts2 = {
+			.pcvt = {
+				.type = PHY_PCVT_ETHERNET_PCS,
+				.index = i,
+			},
+		};
+
+		err = phy_get_status(serdes, PHY_STATUS_PCVT_ADDR, &opts2);
+		if (err)
+			return err;
+
+		/* For a multi-port protocol converter, the match is
+		 * approximate, since for full confidence, we'd have to
+		 * know which port within the PCS is the consumer (MAC)
+		 * mapped to.
+		 */
+		if (opts2.pcvt.addr.mdio == mdiodev->addr)
+			return 0;
+	}
+
+	dev_err(&mdiodev->dev,
+		"Own MDIO address not found among %zu protocol converters reported by SerDes lane %s\n",
+		opts1.pcvt_count.num_pcvt,
+		dev_name(&serdes->dev));
+
+	return -ENODEV;
+}
+
+static struct phylink_pcs *lynx_pcs_create(struct mdio_device *mdio,
+					   struct phy **phys, size_t num_phys,
+					   enum mtip_model model)
 {
 	struct lynx_pcs *lynx;
+	size_t i;
+	int err;
+
+	if (num_phys > MAX_NUM_LANES)
+		return ERR_PTR(-ERANGE);
 
 	lynx = kzalloc(sizeof(*lynx), GFP_KERNEL);
 	if (!lynx)
 		return ERR_PTR(-ENOMEM);
 
+	for (i = 0; i < num_phys; i++) {
+		lynx->serdes[i] = phys[i];
+		err = phy_init(lynx->serdes[i]);
+		if (err) {
+			dev_err(&mdio->dev,
+				"Failed to initialize SerDes: %pe\n",
+				ERR_PTR(err));
+
+			while (i--)
+				phy_exit(lynx->serdes[i]);
+			goto out_free;
+		}
+	}
+
+	if (num_phys) {
+		err = lynx_pcs_validate_addr(mdio, lynx->serdes[PRIMARY_LANE]);
+		if (err && err != -EOPNOTSUPP)
+			goto out_phy_exit;
+	}
+
 	mdio_device_get(mdio);
 	lynx->mdio = mdio;
 	lynx->pcs.ops = &lynx_pcs_phylink_ops;
 	lynx->pcs.neg_mode = true;
 	lynx->pcs.poll = true;
+	lynx->num_phys = num_phys;
+	lynx->model = model;
 
 	return lynx_to_phylink_pcs(lynx);
+
+out_phy_exit:
+	for (i = 0; i < num_phys; i++)
+		phy_exit(lynx->serdes[i]);
+out_free:
+	kfree(lynx);
+	return ERR_PTR(err);
 }
 
-struct phylink_pcs *lynx_pcs_create_mdiodev(struct mii_bus *bus, int addr)
+struct phylink_pcs *lynx_pcs_create_mdiodev(struct mii_bus *bus, int addr,
+					    struct phy **phys, size_t num_phys)
 {
 	struct mdio_device *mdio;
 	struct phylink_pcs *pcs;
@@ -339,7 +607,7 @@
 	if (IS_ERR(mdio))
 		return ERR_CAST(mdio);
 
-	pcs = lynx_pcs_create(mdio);
+	pcs = lynx_pcs_create(mdio, phys, num_phys, MTIP_MODEL_AUTODETECT);
 
 	/* lynx_create() has taken a refcount on the mdiodev if it was
 	 * successful. If lynx_create() fails, this will free the mdio
@@ -363,19 +631,24 @@
  *  -ENOMEM if we fail to allocate memory
  *  pointer to a phylink_pcs on success
  */
-struct phylink_pcs *lynx_pcs_create_fwnode(struct fwnode_handle *node)
+struct phylink_pcs *lynx_pcs_create_fwnode(struct fwnode_handle *node,
+					   struct phy **phys, size_t num_phys)
 {
+	enum mtip_model model = MTIP_MODEL_AUTODETECT;
 	struct mdio_device *mdio;
 	struct phylink_pcs *pcs;
 
 	if (!fwnode_device_is_available(node))
 		return ERR_PTR(-ENODEV);
 
+	if (fwnode_device_is_compatible(node, "fsl,lx2160a-lynx-pcs"))
+		model = MTIP_MODEL_LX2160A;
+
 	mdio = fwnode_mdio_find_device(node);
 	if (!mdio)
 		return ERR_PTR(-EPROBE_DEFER);
 
-	pcs = lynx_pcs_create(mdio);
+	pcs = lynx_pcs_create(mdio, phys, num_phys, model);
 
 	/* lynx_create() has taken a refcount on the mdiodev if it was
 	 * successful. If lynx_create() fails, this will free the mdio
@@ -392,8 +665,13 @@
 void lynx_pcs_destroy(struct phylink_pcs *pcs)
 {
 	struct lynx_pcs *lynx = phylink_pcs_to_lynx(pcs);
+	size_t i;
+
+	for (i = 0; i < lynx->num_phys; i++)
+		phy_exit(lynx->serdes[i]);
 
 	mdio_device_put(lynx->mdio);
+
 	kfree(lynx);
 }
 EXPORT_SYMBOL(lynx_pcs_destroy);
