--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/pcs/pcs-xpcs.c	2026-04-08 18:20:08.500472369 +0200
+++ ./drivers/net/pcs/pcs-xpcs.c	2026-04-08 16:21:07.974808328 +0200
@@ -104,6 +104,18 @@
 	__ETHTOOL_LINK_MODE_MASK_NBITS,
 };
 
+static const int xpcs_mx95_10g_features[] = {
+	ETHTOOL_LINK_MODE_Pause_BIT,
+	ETHTOOL_LINK_MODE_Asym_Pause_BIT,
+	ETHTOOL_LINK_MODE_Autoneg_BIT,
+	ETHTOOL_LINK_MODE_1000baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_2500baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_5000baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_10000baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_10000baseR_FEC_BIT,
+	__ETHTOOL_LINK_MODE_MASK_NBITS,
+};
+
 static const phy_interface_t xpcs_usxgmii_interfaces[] = {
 	PHY_INTERFACE_MODE_USXGMII,
 };
@@ -201,6 +213,22 @@
 #define xpcs_linkmode_supported(compat, mode) \
 	__xpcs_linkmode_supported(compat, ETHTOOL_LINK_MODE_ ## mode ## _BIT)
 
+int xpcs_phy_read(struct dw_xpcs *xpcs, int dev, u32 reg)
+{
+	struct mii_bus *bus = xpcs->phydev->bus;
+	int addr = xpcs->phydev->addr;
+
+	return mdiobus_c45_read(bus, addr, dev, reg);
+}
+
+int xpcs_phy_write(struct dw_xpcs *xpcs, int dev, u32 reg, u16 val)
+{
+	struct mii_bus *bus = xpcs->phydev->bus;
+	int addr = xpcs->phydev->addr;
+
+	return mdiobus_c45_write(bus, addr, dev, reg, val);
+}
+
 int xpcs_read(struct dw_xpcs *xpcs, int dev, u32 reg)
 {
 	return mdiodev_c45_read(xpcs->mdiodev, dev, reg);
@@ -1313,6 +1341,16 @@
 	},
 };
 
+static const struct xpcs_compat nxp_mx95_xpcs_compat[DW_XPCS_INTERFACE_MAX] = {
+	[DW_XPCS_10GBASER] = {
+		.supported = xpcs_mx95_10g_features,
+		.interface = xpcs_10gbaser_interfaces,
+		.num_interfaces = ARRAY_SIZE(xpcs_10gbaser_interfaces),
+		.an_mode = DW_10GBASER,
+		.pma_config = xpcs_phy_usxgmii_pma_config,
+	},
+};
+
 static const struct xpcs_id xpcs_id_list[] = {
 	{
 		.id = SYNOPSYS_XPCS_ID,
@@ -1326,6 +1364,10 @@
 		.id = NXP_SJA1110_XPCS_ID,
 		.mask = SYNOPSYS_XPCS_MASK,
 		.compat = nxp_sja1110_xpcs_compat,
+	}, {
+		.id = NXP_MX95_XPCS_ID,
+		.mask = SYNOPSYS_XPCS_MASK,
+		.compat = nxp_mx95_xpcs_compat,
 	},
 };
 
@@ -1337,7 +1379,23 @@
 	.pcs_link_up = xpcs_link_up,
 };
 
+u32 xpcs_register_phy(struct dw_xpcs *xpcs, struct mii_bus *bus)
+{
+	u32 xpcs_phy_id;
+	int ret = 0;
+
+	xpcs_phy_reg_lock(xpcs);
+
+	xpcs_phy_id = xpcs_phy_get_id(xpcs);
+	ret = xpcs_phy_check_id(xpcs_phy_id);
+	if (!ret)
+		return 0;
+
+	return xpcs_phy_id;
+}
+
 static struct dw_xpcs *xpcs_create(struct mdio_device *mdiodev,
+				   struct mdio_device *phydev,
 				   phy_interface_t interface)
 {
 	struct dw_xpcs *xpcs;
@@ -1351,7 +1409,12 @@
 	mdio_device_get(mdiodev);
 	xpcs->mdiodev = mdiodev;
 
-	xpcs_id = xpcs_get_id(xpcs);
+	if (phydev) {
+		mdio_device_get(phydev);
+		xpcs->phydev = phydev;
+		xpcs_id = xpcs_register_phy(xpcs, mdiodev->bus);
+	} else
+		xpcs_id = xpcs_get_id(xpcs);
 
 	for (i = 0; i < ARRAY_SIZE(xpcs_id_list); i++) {
 		const struct xpcs_id *entry = &xpcs_id_list[i];
@@ -1390,15 +1453,33 @@
 
 out:
 	mdio_device_put(mdiodev);
+	if (phydev)
+		mdio_device_put(phydev);
 	kfree(xpcs);
 
 	return ERR_PTR(ret);
 }
 
+void xpcs_pcs_destroy(struct phylink_pcs *pcs)
+{
+	struct dw_xpcs *xpcs = phylink_pcs_to_xpcs(pcs);
+
+	if (xpcs) {
+		mdio_device_put(xpcs->mdiodev);
+		if (xpcs->phydev)
+			mdio_device_put(xpcs->phydev);
+	}
+	kfree(xpcs);
+}
+EXPORT_SYMBOL_GPL(xpcs_pcs_destroy);
+
 void xpcs_destroy(struct dw_xpcs *xpcs)
 {
-	if (xpcs)
+	if (xpcs) {
 		mdio_device_put(xpcs->mdiodev);
+		if (xpcs->phydev)
+			mdio_device_put(xpcs->phydev);
+	}
 	kfree(xpcs);
 }
 EXPORT_SYMBOL_GPL(xpcs_destroy);
@@ -1413,7 +1494,7 @@
 	if (IS_ERR(mdiodev))
 		return ERR_CAST(mdiodev);
 
-	xpcs = xpcs_create(mdiodev, interface);
+	xpcs = xpcs_create(mdiodev, NULL, interface);
 
 	/* xpcs_create() has taken a refcount on the mdiodev if it was
 	 * successful. If xpcs_create() fails, this will free the mdio
@@ -1427,4 +1508,48 @@
 }
 EXPORT_SYMBOL_GPL(xpcs_create_mdiodev);
 
+struct phylink_pcs *xpcs_create_mdiodev_with_phy(struct mii_bus *bus,
+						 int mdioaddr, int phyaddr,
+						 phy_interface_t interface)
+{
+	struct mdio_device *mdiodev, *phydev;
+	struct dw_xpcs *xpcs;
+	void *err_ptr;
+
+	mdiodev = mdio_device_create(bus, mdioaddr);
+	if (IS_ERR(mdiodev))
+		return ERR_CAST(mdiodev);
+
+	phydev = mdio_device_create(bus, phyaddr);
+	if (IS_ERR(phydev)) {
+		err_ptr = ERR_CAST(phydev);
+		goto err_phydev;
+	}
+
+	xpcs = xpcs_create(mdiodev, phydev, interface);
+	if (IS_ERR(xpcs)) {
+		err_ptr = ERR_CAST(xpcs);
+		goto err_xpcs;
+	}
+
+	/* xpcs_create() has taken a refcount on the mdiodev if it was
+	 * successful. If xpcs_create() fails, this will free the mdio
+	 * device here. In any case, we don't need to hold our reference
+	 * anymore, and putting it here will allow mdio_device_put() in
+	 * xpcs_destroy() to automatically free the mdio device.
+	 */
+	mdio_device_put(mdiodev);
+	mdio_device_put(phydev);
+
+	return &xpcs->pcs;
+
+err_xpcs:
+	mdio_device_free(phydev);
+err_phydev:
+	mdio_device_free(mdiodev);
+
+	return err_ptr;
+}
+EXPORT_SYMBOL_GPL(xpcs_create_mdiodev_with_phy);
+
 MODULE_LICENSE("GPL v2");
