--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/net/phy/realtek.c	2026-04-08 18:20:08.507979931 +0200
+++ ./drivers/net/phy/realtek.c	2026-04-08 16:21:07.992147550 +0200
@@ -30,6 +30,7 @@
 
 #define RTL8211F_PHYCR1				0x18
 #define RTL8211F_PHYCR2				0x19
+#define RTL8211FVD_PHYCR2			0x11
 #define RTL8211F_INSR				0x1d
 
 #define RTL8211F_TX_DELAY			BIT(8)
@@ -44,6 +45,7 @@
 #define RTL8211E_RX_DELAY			BIT(11)
 
 #define RTL8211F_CLKOUT_EN			BIT(0)
+#define RTL8211FVD_CLKOUT_EN			BIT(8)
 
 #define RTL8201F_ISR				0x1e
 #define RTL8201F_ISR_ANERR			BIT(15)
@@ -80,7 +82,6 @@
 struct rtl821x_priv {
 	u16 phycr1;
 	u16 phycr2;
-	bool has_phycr2;
 	struct clk *clk;
 };
 
@@ -99,6 +100,7 @@
 	struct device *dev = &phydev->mdio.dev;
 	struct rtl821x_priv *priv;
 	u32 phy_id = phydev->drv->phy_id;
+	bool disable_clk_out;
 	int ret;
 
 	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
@@ -118,14 +120,22 @@
 	if (of_property_read_bool(dev->of_node, "realtek,aldps-enable"))
 		priv->phycr1 |= RTL8211F_ALDPS_PLL_OFF | RTL8211F_ALDPS_ENABLE | RTL8211F_ALDPS_XTAL_OFF;
 
-	priv->has_phycr2 = !(phy_id == RTL_8211FVD_PHYID);
-	if (priv->has_phycr2) {
+	disable_clk_out = of_property_read_bool(dev->of_node, "realtek,clkout-disable");
+	if (phy_id == RTL_8211FVD_PHYID) {
+		ret = phy_read_paged(phydev, 0xd05, RTL8211FVD_PHYCR2);
+		if (ret < 0)
+			return ret;
+
+		priv->phycr2 = ret & RTL8211FVD_CLKOUT_EN;
+		if (disable_clk_out)
+			priv->phycr2 &= ~RTL8211FVD_CLKOUT_EN;
+	} else {
 		ret = phy_read_paged(phydev, 0xa43, RTL8211F_PHYCR2);
 		if (ret < 0)
 			return ret;
 
 		priv->phycr2 = ret & RTL8211F_CLKOUT_EN;
-		if (of_property_read_bool(dev->of_node, "realtek,clkout-disable"))
+		if (disable_clk_out)
 			priv->phycr2 &= ~RTL8211F_CLKOUT_EN;
 	}
 
@@ -346,6 +356,7 @@
 {
 	struct rtl821x_priv *priv = phydev->priv;
 	struct device *dev = &phydev->mdio.dev;
+	u32 phy_id = phydev->drv->phy_id;
 	u16 val_txdly, val_rxdly;
 	int ret;
 
@@ -413,19 +424,19 @@
 			val_rxdly ? "enabled" : "disabled");
 	}
 
-	if (priv->has_phycr2) {
+	if (phy_id == RTL_8211FVD_PHYID)
+		ret = phy_modify_paged(phydev, 0xd05, RTL8211FVD_PHYCR2,
+				       RTL8211FVD_CLKOUT_EN, priv->phycr2);
+	else
 		ret = phy_modify_paged(phydev, 0xa43, RTL8211F_PHYCR2,
 				       RTL8211F_CLKOUT_EN, priv->phycr2);
-		if (ret < 0) {
-			dev_err(dev, "clkout configuration failed: %pe\n",
-				ERR_PTR(ret));
-			return ret;
-		}
-
-		return genphy_soft_reset(phydev);
+	if (ret < 0) {
+		dev_err(dev, "clkout configuration failed: %pe\n",
+			ERR_PTR(ret));
+		return ret;
 	}
 
-	return 0;
+	return genphy_soft_reset(phydev);
 }
 
 static int rtl821x_suspend(struct phy_device *phydev)
