--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/phy/inphi/Kconfig	1970-01-01 01:00:00.000000000 +0100
+++ ./drivers/phy/inphi/Kconfig	2026-04-08 16:21:08.689600230 +0200
@@ -0,0 +1,10 @@
+config PHY_INPHI_IN112525_RETIMER
+	bool "Inphi IN112525 Retimer PHY support"
+	depends on OF
+	select GENERIC_PHY
+	help
+	  Enable this to add support for configuration and runtime management
+	  of Inphi IN112525 S03 Retimer PHY. The retimer is modeled as a
+	  Generic PHY and supports both 10G and 25G link speeds.
+	  The driver will configure the retimer at probe time to function in
+	  25Gbps but another configuration can be requested dynamically.
