--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/phy/Kconfig	2026-04-08 18:20:08.883990407 +0200
+++ ./drivers/phy/Kconfig	2026-04-08 16:21:08.679206705 +0200
@@ -72,6 +72,16 @@
 	  functional modes using gpios and sets the attribute max link
 	  rate, for CAN drivers.
 
+config PHY_MIXEL_LVDS
+	tristate "MIXEL LVDS PHY support"
+	depends on OF
+	select GENERIC_PHY
+
+config PHY_MIXEL_LVDS_COMBO
+	tristate "MIXEL LVDS PHY(LVDS+MIPI DSI combo PHY) support"
+	depends on OF
+	select GENERIC_PHY
+
 source "drivers/phy/allwinner/Kconfig"
 source "drivers/phy/amlogic/Kconfig"
 source "drivers/phy/broadcom/Kconfig"
@@ -98,5 +108,6 @@
 source "drivers/phy/ti/Kconfig"
 source "drivers/phy/intel/Kconfig"
 source "drivers/phy/xilinx/Kconfig"
+source "drivers/phy/inphi/Kconfig"
 
 endmenu
