--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/pmdomain/imx/imx93-blk-ctrl.c	2026-04-08 18:20:08.990056267 +0200
+++ ./drivers/pmdomain/imx/imx93-blk-ctrl.c	2026-04-08 16:21:08.893197301 +0200
@@ -216,7 +216,7 @@
 	if (IS_ERR(base))
 		return PTR_ERR(base);
 
-	bc->regmap = devm_regmap_init_mmio(dev, base, &regmap_config);
+	bc->regmap = regmap_init_mmio(NULL, base, &regmap_config);
 	if (IS_ERR(bc->regmap))
 		return dev_err_probe(dev, PTR_ERR(bc->regmap),
 				     "failed to init regmap\n");
@@ -224,15 +224,19 @@
 	bc->domains = devm_kcalloc(dev, bc_data->num_domains,
 				   sizeof(struct imx93_blk_ctrl_domain),
 				   GFP_KERNEL);
-	if (!bc->domains)
-		return -ENOMEM;
+	if (!bc->domains) {
+		ret = -ENOMEM;
+		goto free_regmap;
+	}
 
 	bc->onecell_data.num_domains = bc_data->num_domains;
 	bc->onecell_data.domains =
 		devm_kcalloc(dev, bc_data->num_domains,
 			     sizeof(struct generic_pm_domain *), GFP_KERNEL);
-	if (!bc->onecell_data.domains)
-		return -ENOMEM;
+	if (!bc->onecell_data.domains) {
+		ret = -ENOMEM;
+		goto free_regmap;
+	}
 
 	for (i = 0; i < bc_data->num_clks; i++)
 		bc->clks[i].id = bc_data->clk_names[i];
@@ -241,7 +245,7 @@
 	ret = devm_clk_bulk_get(dev, bc->num_clks, bc->clks);
 	if (ret) {
 		dev_err_probe(dev, ret, "failed to get bus clock\n");
-		return ret;
+		goto free_regmap;
 	}
 
 	for (i = 0; i < bc_data->num_domains; i++) {
@@ -303,6 +307,9 @@
 	for (i--; i >= 0; i--)
 		pm_genpd_remove(&bc->domains[i].genpd);
 
+free_regmap:
+	regmap_exit(bc->regmap);
+
 	return ret;
 }
 
@@ -428,6 +435,36 @@
 	.reg_access_table = &imx93_media_blk_ctl_access_table,
 };
 
+#ifdef CONFIG_PM_SLEEP
+static int imx93_blk_ctrl_suspend(struct device *dev)
+{
+	struct imx93_blk_ctrl *bc = dev_get_drvdata(dev);
+
+	/*
+	 * This may look strange, but is done so the generic PM_SLEEP code
+	 * can power down our domains and more importantly power them up again
+	 * after resume, without tripping over our usage of runtime PM to
+	 * control the upstream GPC domains. Things happen in the right order
+	 * in the system suspend/resume paths due to the device parent/child
+	 * hierarchy.
+	 */
+	return pm_runtime_resume_and_get(bc->dev);
+}
+
+static int imx93_blk_ctrl_resume(struct device *dev)
+{
+	struct imx93_blk_ctrl *bc = dev_get_drvdata(dev);
+
+	pm_runtime_put(bc->dev);
+
+	return 0;
+}
+#endif
+
+static const struct dev_pm_ops imx93_blk_ctrl_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(imx93_blk_ctrl_suspend, imx93_blk_ctrl_resume)
+};
+
 static const struct of_device_id imx93_blk_ctrl_of_match[] = {
 	{
 		.compatible = "fsl,imx93-media-blk-ctrl",
@@ -443,6 +480,7 @@
 	.remove = imx93_blk_ctrl_remove,
 	.driver = {
 		.name = "imx93-blk-ctrl",
+		.pm = &imx93_blk_ctrl_pm_ops,
 		.of_match_table = imx93_blk_ctrl_of_match,
 	},
 };
