--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/ptp/Kconfig	2026-04-08 18:20:09.011021742 +0200
+++ ./drivers/ptp/Kconfig	2026-04-08 16:21:08.953424660 +0200
@@ -57,7 +57,7 @@
 
 config PTP_1588_CLOCK_QORIQ
 	tristate "Freescale QorIQ 1588 timer as PTP clock"
-	depends on GIANFAR || FSL_DPAA_ETH || FSL_DPAA2_ETH || FSL_ENETC || FSL_ENETC_VF || COMPILE_TEST
+	depends on !FSL_SDK_FMAN_RTC_API && (GIANFAR || FSL_DPAA_ETH || FSL_SDK_DPAA_ETH || FSL_DPAA2_ETH || FSL_ENETC || FSL_ENETC_VF) || COMPILE_TEST
 	depends on PTP_1588_CLOCK
 	default y
 	help
@@ -211,4 +211,18 @@
 	  To compile this driver as a module, choose M here: the module
 	  will be called ptp_dfl_tod.
 
+config PTP_1588_CLOCK_NETC
+	tristate "NXP NETC Timer as PTP clock"
+	depends on FSL_NETC_PRB_IERB && (FSL_ENETC4 || FSL_ENETC_VF)
+	depends on PTP_1588_CLOCK
+	default y
+	help
+	  This driver adds support for using the NXP NETC Timer as a
+	  PTP clock. This clock is only useful if your PTP programs
+	  are getting hardware time stamps on the PTP Ethernet packets
+	  using the SO_TIMESTAMPING API.
+
+	  To compile this driver as a module, choose M here: the module
+	  will be called ptp_netc.
+
 endmenu
