--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/pwm/pwm-imx27.c	2026-04-08 18:20:09.014994057 +0200
+++ ./drivers/pwm/pwm-imx27.c	2026-04-08 16:21:08.967955032 +0200
@@ -21,6 +21,7 @@
 #include <linux/platform_device.h>
 #include <linux/pwm.h>
 #include <linux/slab.h>
+#include <linux/spinlock.h>
 
 #define MX3_PWMCR			0x00    /* PWM Control Register */
 #define MX3_PWMSR			0x04    /* PWM Status Register */
@@ -83,6 +84,7 @@
 struct pwm_imx27_chip {
 	struct clk	*clk_ipg;
 	struct clk	*clk_per;
+	struct clk	*clk_32k;
 	void __iomem	*mmio_base;
 	struct pwm_chip	chip;
 
@@ -92,6 +94,7 @@
 	 * value to return in that case.
 	 */
 	unsigned int duty_cycle;
+	spinlock_t lock;
 };
 
 #define to_pwm_imx27_chip(chip)	container_of(chip, struct pwm_imx27_chip, chip)
@@ -100,23 +103,36 @@
 {
 	int ret;
 
+	if (imx->clk_32k) {
+		ret = clk_prepare_enable(imx->clk_32k);
+		if (ret)
+			goto err1;
+	}
+
 	ret = clk_prepare_enable(imx->clk_ipg);
 	if (ret)
-		return ret;
+		goto err2;
 
 	ret = clk_prepare_enable(imx->clk_per);
-	if (ret) {
-		clk_disable_unprepare(imx->clk_ipg);
-		return ret;
-	}
+	if (ret)
+		goto err3;
 
 	return 0;
+err3:
+	clk_disable_unprepare(imx->clk_ipg);
+err2:
+	if (imx->clk_32k)
+		clk_disable_unprepare(imx->clk_32k);
+err1:
+	return ret;
 }
 
 static void pwm_imx27_clk_disable_unprepare(struct pwm_imx27_chip *imx)
 {
 	clk_disable_unprepare(imx->clk_per);
 	clk_disable_unprepare(imx->clk_ipg);
+	if (imx->clk_32k)
+		clk_disable_unprepare(imx->clk_32k);
 }
 
 static int pwm_imx27_get_state(struct pwm_chip *chip,
@@ -204,10 +220,10 @@
 
 	sr = readl(imx->mmio_base + MX3_PWMSR);
 	fifoav = FIELD_GET(MX3_PWMSR_FIFOAV, sr);
-	if (fifoav == MX3_PWMSR_FIFOAV_4WORDS) {
+	if (fifoav >= MX3_PWMSR_FIFOAV_3WORDS) {
 		period_ms = DIV_ROUND_UP_ULL(pwm_get_period(pwm),
 					 NSEC_PER_MSEC);
-		msleep(period_ms);
+		msleep(period_ms * (fifoav - 2));
 
 		sr = readl(imx->mmio_base + MX3_PWMSR);
 		if (fifoav == FIELD_GET(MX3_PWMSR_FIFOAV, sr))
@@ -218,15 +234,15 @@
 static int pwm_imx27_apply(struct pwm_chip *chip, struct pwm_device *pwm,
 			   const struct pwm_state *state)
 {
-	unsigned long period_cycles, duty_cycles, prescale, period_us, tmp;
+	unsigned long period_cycles, duty_cycles, prescale, counter_check, flags;
 	struct pwm_imx27_chip *imx = to_pwm_imx27_chip(chip);
+	void __iomem *reg_sar = imx->mmio_base + MX3_PWMSAR;
+	__force u32 sar_last, sar_current;
 	struct pwm_state cstate;
 	unsigned long long c;
 	unsigned long long clkrate;
-	unsigned long flags;
-	int val;
 	int ret;
-	u32 cr;
+	u32 cr, timeout = 1000;
 
 	pwm_get_state(pwm, &cstate);
 
@@ -267,97 +283,56 @@
 		pwm_imx27_sw_reset(chip);
 	}
 
-	val = readl(imx->mmio_base + MX3_PWMPR);
-	val = val >= MX3_PWMPR_MAX ? MX3_PWMPR_MAX : val;
-	cr = readl(imx->mmio_base + MX3_PWMCR);
-	tmp = NSEC_PER_SEC * (u64)(val + 2) * MX3_PWMCR_PRESCALER_GET(cr);
-	tmp = DIV_ROUND_UP_ULL(tmp, clkrate);
-	period_us = DIV_ROUND_UP_ULL(tmp, 1000);
-
 	/*
-	 * ERR051198:
-	 * PWM: PWM output may not function correctly if the FIFO is empty when
-	 * a new SAR value is programmed
-	 *
-	 * Description:
-	 * When the PWM FIFO is empty, a new value programmed to the PWM Sample
-	 * register (PWM_PWMSAR) will be directly applied even if the current
-	 * timer period has not expired.
-	 *
-	 * If the new SAMPLE value programmed in the PWM_PWMSAR register is
-	 * less than the previous value, and the PWM counter register
-	 * (PWM_PWMCNR) that contains the current COUNT value is greater than
-	 * the new programmed SAMPLE value, the current period will not flip
-	 * the level. This may result in an output pulse with a duty cycle of
-	 * 100%.
+	 * This is a limited workaround. When the SAR FIFO is empty, the new
+	 * write value will be directly applied to SAR even the current period
+	 * is not over.
+	 * If the new SAR value is less than the old one, and the counter is
+	 * greater than the new SAR value, the current period will not filp
+	 * the level. This will result in a pulse with a duty cycle of 100%.
+	 * So, writing the current value of the SAR to SAR here before updating
+	 * the new SAR value can avoid this issue.
+	 *
+	 * Add a spin lock and turn off the interrupt to ensure that the
+	 * real-time performance can be guaranteed as much as possible when
+	 * operating the following operations.
+	 *
+	 * 1. Add a threshold of 1.5us. If the time T between the read current
+	 * count value CNR and the end of the cycle is less than 1.5us, wait
+	 * for T to be longer than 1.5us before updating the SAR register.
+	 * This is to avoid the situation that when the first SAR is written,
+	 * the current cycle just ends and the SAR FIFO that just be written
+	 * is emptied again.
 	 *
-	 * Consider a change from
-	 *     ________
-	 *    /        \______/
-	 *    ^      *        ^
-	 * to
-	 *     ____
-	 *    /    \__________/
-	 *    ^               ^
-	 * At the time marked by *, the new write value will be directly applied
-	 * to SAR even the current period is not over if FIFO is empty.
+	 * 2. Use __raw_writel() to minimize the interval between two writes to
+	 * the SAR register to increase the fastest pwm frequency supported.
 	 *
-	 *     ________        ____________________
-	 *    /        \______/                    \__________/
-	 *    ^               ^      *        ^               ^
-	 *    |<-- old SAR -->|               |<-- new SAR -->|
-	 *
-	 * That is the output is active for a whole period.
-	 *
-	 * Workaround:
-	 * Check new SAR less than old SAR and current counter is in errata
-	 * windows, write extra old SAR into FIFO and new SAR will effect at
-	 * next period.
-	 *
-	 * Sometime period is quite long, such as over 1 second. If add old SAR
-	 * into FIFO unconditional, new SAR have to wait for next period. It
-	 * may be too long.
-	 *
-	 * Turn off the interrupt to ensure that not IRQ and schedule happen
-	 * during above operations. If any irq and schedule happen, counter
-	 * in PWM will be out of data and take wrong action.
-	 *
-	 * Add a safety margin 1.5us because it needs some time to complete
-	 * IO write.
-	 *
-	 * Use writel_relaxed() to minimize the interval between two writes to
-	 * the SAR register to increase the fastest PWM frequency supported.
-	 *
-	 * When the PWM period is longer than 2us(or <500kHz), this workaround
-	 * can solve this problem. No software workaround is available if PWM
-	 * period is shorter than IO write. Just try best to fill old data
-	 * into FIFO.
+	 * When the PWM period is longer than 2us(or <500KHz), this workaround
+	 * can solve this problem.
 	 */
-	c = clkrate * 1500;
-	do_div(c, NSEC_PER_SEC);
-
-	local_irq_save(flags);
-	val = FIELD_GET(MX3_PWMSR_FIFOAV, readl_relaxed(imx->mmio_base + MX3_PWMSR));
-
-	if (duty_cycles < imx->duty_cycle && (cr & MX3_PWMCR_EN)) {
-		if (period_us < 2) { /* 2us = 500 kHz */
-			/* Best effort attempt to fix up >500 kHz case */
-			udelay(3 * period_us);
-			writel_relaxed(imx->duty_cycle, imx->mmio_base + MX3_PWMSAR);
-			writel_relaxed(imx->duty_cycle, imx->mmio_base + MX3_PWMSAR);
-		} else if (val < MX3_PWMSR_FIFOAV_2WORDS) {
-			val = readl_relaxed(imx->mmio_base + MX3_PWMCNR);
-			/*
-			 * If counter is close to period, controller may roll over when
-			 * next IO write.
-			 */
-			if ((val + c >= duty_cycles && val < imx->duty_cycle) ||
-			    val + c >= period_cycles)
-				writel_relaxed(imx->duty_cycle, imx->mmio_base + MX3_PWMSAR);
+	if (duty_cycles < imx->duty_cycle) {
+		c = clkrate * 1500;
+		do_div(c, NSEC_PER_SEC);
+		counter_check = c;
+		sar_last = cpu_to_le32(imx->duty_cycle);
+		sar_current = cpu_to_le32(duty_cycles);
+
+		spin_lock_irqsave(&imx->lock, flags);
+		if (state->period >= 2000) {
+			while ((period_cycles -
+				readl_relaxed(imx->mmio_base + MX3_PWMCNR))
+				< counter_check) {
+				if (!--timeout)
+					break;
+			};
 		}
-	}
-	writel_relaxed(duty_cycles, imx->mmio_base + MX3_PWMSAR);
-	local_irq_restore(flags);
+		if (!(MX3_PWMSR_FIFOAV &
+		      readl_relaxed(imx->mmio_base + MX3_PWMSR)))
+			__raw_writel(sar_last, reg_sar);
+		__raw_writel(sar_current, reg_sar);
+		spin_unlock_irqrestore(&imx->lock, flags);
+	} else
+		writel(duty_cycles, imx->mmio_base + MX3_PWMSAR);
 
 	writel(period_cycles, imx->mmio_base + MX3_PWMPR);
 
@@ -419,6 +394,15 @@
 		return dev_err_probe(&pdev->dev, PTR_ERR(imx->clk_per),
 				     "failed to get peripheral clock\n");
 
+	imx->clk_32k = devm_clk_get_optional(&pdev->dev, "32k");
+	if (IS_ERR(imx->clk_32k)) {
+		dev_err(&pdev->dev, "getting 32k clock failed with %ld\n",
+				PTR_ERR(imx->clk_32k));
+		return PTR_ERR(imx->clk_32k);
+	}
+
+	spin_lock_init(&imx->lock);
+	imx->duty_cycle = 0;
 	imx->chip.ops = &pwm_imx27_ops;
 	imx->chip.dev = &pdev->dev;
 	imx->chip.npwm = 1;
