--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/regulator/Kconfig	2026-04-08 18:20:09.021894577 +0200
+++ ./drivers/regulator/Kconfig	2026-04-08 16:21:08.973460981 +0200
@@ -554,6 +554,10 @@
 	  regulator via I2C bus. The provided regulator is suitable
 	  for PXA27x chips to control VCC_CORE and VCC_USIM voltages.
 
+config REGULATOR_MAX17135
+	tristate "Maxim MAX17135 Regulator Support"
+	depends on MFD_MAX17135
+
 config REGULATOR_MAX5970
 	tristate "Maxim 5970/5978 power switch and monitor"
 	depends on I2C
@@ -728,6 +732,15 @@
 	  It includes support for control of output voltage. This
 	  regulator is found on the Samsung Galaxy S5 (klte) smartphone.
 
+config REGULATOR_FP9931
+	tristate "Fitipower FP9931 Regulator Support"
+	depends on MFD_FP9931
+	help
+	  This driver controls a Fitipower 9931 regulator via I2C bus.
+	  The regulator includes LDOs, BUCK, BUCK-BOOST, Charge-Pump
+	  and VCOM Buffer regulators. It can support for control of
+	  output voltage.
+
 config REGULATOR_MC13XXX_CORE
 	tristate
 
@@ -943,6 +956,29 @@
 	  Say y here to support the NXP PCA9450A/PCA9450B/PCA9450C PMIC
 	  regulator driver.
 
+config REGULATOR_PF9453
+	tristate "NXP PF9453 regulator driver"
+	depends on I2C
+	select REGMAP_I2C
+	help
+	  Say y here to support the NXP PF9453 PMIC regulator driver.
+
+config REGULATOR_PF0900
+	tristate "NXP PF0900/PF0901/PF09XX regulator driver"
+	depends on I2C
+	select REGMAP_I2C
+	help
+	 Say y here to support the NXP PF0900/PF0901/PF09XX PMIC
+	 regulator driver.
+
+config REGULATOR_PF5300
+	tristate "NXP PF5300/PF5301/PF5302 regulator driver"
+	depends on I2C
+	select REGMAP_I2C
+	help
+	 Say y here to support the NXP PF5300/PF5301/PF5302 PMIC
+	 regulator driver.
+
 config REGULATOR_PCAP
 	tristate "Motorola PCAP2 regulator driver"
 	depends on EZX_PCAP
@@ -969,6 +1005,14 @@
 	  on the NXP PF8100/PF8121A/PF8200 PMIC. The module will be named
 	  "pf8x00-regulator".
 
+config REGULATOR_PF1550_RPMSG
+	tristate "Freescale PF1550 rpmsg regulator driver"
+	depends on RPMSG
+	help
+	  This driver controls a PF1550 regulator which connected in M4
+	  side via RPMSG.The regulators include three switch and three
+	  ldo.
+
 config REGULATOR_PFUZE100
 	tristate "Freescale PFUZE100/200/3000/3001 regulator driver"
 	depends on I2C && OF
