--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/remoteproc/imx_dsp_rproc.c	2026-04-08 18:20:09.037166520 +0200
+++ ./drivers/remoteproc/imx_dsp_rproc.c	2026-04-08 16:21:09.013701258 +0200
@@ -25,7 +25,7 @@
 #include "remoteproc_elf_helpers.h"
 #include "remoteproc_internal.h"
 
-#define DSP_RPROC_CLK_MAX			5
+#define DSP_RPROC_CLK_MAX			(5 + 18)
 
 /*
  * Module parameters
@@ -940,6 +940,7 @@
 	.kick		= imx_dsp_rproc_kick,
 	.load		= imx_dsp_rproc_elf_load_segments,
 	.parse_fw	= imx_dsp_rproc_parse_fw,
+	.find_loaded_rsc_table = rproc_elf_find_loaded_rsc_table,
 	.sanity_check	= rproc_elf_sanity_check,
 	.get_boot_addr	= rproc_elf_get_boot_addr,
 };
@@ -1069,6 +1070,11 @@
 static const char *imx_dsp_clks_names[DSP_RPROC_CLK_MAX] = {
 	/* DSP clocks */
 	"core", "ocram", "debug", "ipg", "mu",
+
+	/* peripheral clocks */
+	"per_clk1", "per_clk2", "per_clk3", "per_clk4", "per_clk5", "per_clk6",
+	"per_clk7", "per_clk8", "per_clk9", "per_clk10", "per_clk11", "per_clk12",
+	"per_clk13", "per_clk14", "per_clk15", "per_clk16", "per_clk17", "per_clk18",
 };
 
 static int imx_dsp_rproc_clk_get(struct imx_dsp_rproc *priv)
