--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/rtc/Kconfig	2026-04-08 18:20:09.043994865 +0200
+++ ./drivers/rtc/Kconfig	2026-04-08 16:21:09.028412553 +0200
@@ -915,6 +915,21 @@
 	  This driver can also be built as a module. If so, the module
 	  will be called rtc-pcf2127.
 
+config RTC_DRV_PCF2131
+	tristate "NXP PCF2131"
+	depends on RTC_I2C_AND_SPI
+	select REGMAP_I2C if I2C
+	select REGMAP_SPI if SPI_MASTER
+	select WATCHDOG_CORE if WATCHDOG
+	help
+	  If you say yes here you get support for the NXP PCF2131 RTC
+	  chips with integrated quartz crystal for industrial applications.
+	  Both chips also have watchdog timer and tamper switch detection
+	  features.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called rtc-pcf2131.
+
 config RTC_DRV_RV3029C2
 	tristate "Micro Crystal RV3029/3049"
 	depends on RTC_I2C_AND_SPI
@@ -1799,6 +1814,21 @@
 	   If you say yes here you get support for the NXP i.MX System
 	   Controller RTC module.
 
+config RTC_DRV_IMX_SM
+	depends on IMX_SCMI_BBM_CONTROL
+	tristate "NXP i.MX RTC support through SCMI BBM protocol"
+	help
+	   If you say yes here you get support for the NXP i.MX System
+	   Controller BBM RTC module.
+
+
+config RTC_DRV_IMX_RPMSG
+	tristate "NXP RPMSG RTC support"
+	depends on OF
+	help
+	   If you say yes here you get support for the NXP RPMSG
+	   RTC module.
+
 config RTC_DRV_ST_LPC
 	tristate "STMicroelectronics LPC RTC"
 	depends on ARCH_STI
