--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/scsi/scsi_ioctl.c	2026-04-08 18:20:09.255000744 +0200
+++ ./drivers/scsi/scsi_ioctl.c	2026-04-08 16:21:09.292000846 +0200
@@ -369,6 +369,12 @@
 	return 0;
 }
 
+#ifdef CONFIG_AHCI_IMX
+extern void *sg_io_buffer_hack;
+#else
+#define sg_io_buffer_hack NULL
+#endif
+
 static int scsi_complete_sghdr_rq(struct request *rq, struct sg_io_hdr *hdr,
 		struct bio *bio)
 {
@@ -400,7 +406,12 @@
 			ret = -EFAULT;
 	}
 
-	r = blk_rq_unmap_user(bio);
+	if (sg_io_buffer_hack && !hdr->iovec_count)
+		r = copy_to_user(hdr->dxferp, sg_io_buffer_hack,
+				hdr->dxfer_len);
+	else
+		r = blk_rq_unmap_user(bio);
+
 	if (!ret)
 		ret = r;
 
@@ -424,6 +435,9 @@
 	if (hdr->dxfer_len > (queue_max_hw_sectors(sdev->request_queue) << 9))
 		return -EIO;
 
+	if (sg_io_buffer_hack && hdr->dxfer_len > 0x10000)
+		return -EIO;
+
 	if (hdr->dxfer_len)
 		switch (hdr->dxfer_direction) {
 		default:
@@ -453,9 +467,15 @@
 	if (ret < 0)
 		goto out_put_request;
 
-	ret = blk_rq_map_user_io(rq, NULL, hdr->dxferp, hdr->dxfer_len,
-			GFP_KERNEL, hdr->iovec_count && hdr->dxfer_len,
-			hdr->iovec_count, 0, rq_data_dir(rq));
+	if (sg_io_buffer_hack && !hdr->iovec_count) {
+		if (hdr->dxfer_len)
+			ret = blk_rq_map_kern(rq->q, rq, sg_io_buffer_hack,
+					hdr->dxfer_len, GFP_KERNEL);
+	} else {
+		ret = blk_rq_map_user_io(rq, NULL, hdr->dxferp, hdr->dxfer_len,
+				GFP_KERNEL, hdr->iovec_count && hdr->dxfer_len,
+				hdr->iovec_count, 0, rq_data_dir(rq));
+	}
 	if (ret)
 		goto out_put_request;
 
