--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/soc/fsl/qbman/bman_ccsr.c	2026-04-08 18:20:09.275825871 +0200
+++ ./drivers/soc/fsl/qbman/bman_ccsr.c	2026-04-08 16:21:09.320325991 +0200
@@ -1,4 +1,5 @@
 /* Copyright (c) 2009 - 2016 Freescale Semiconductor, Inc.
+ * Copyright 2020 Puresoftware Ltd.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are met:
@@ -29,6 +30,7 @@
  */
 
 #include "bman_priv.h"
+#include <linux/acpi.h>
 
 u16 bman_ip_rev;
 EXPORT_SYMBOL(bman_ip_rev);
@@ -247,7 +249,8 @@
 	 * try using the of_reserved_mem_device method
 	 */
 	if (!fbpr_a) {
-		ret = qbman_init_private_mem(dev, 0, &fbpr_a, &fbpr_sz);
+		ret = qbman_init_private_mem(dev, 0, &fbpr_a, &fbpr_sz,
+					     DPAA_BMAN_DEV);
 		if (ret) {
 			dev_err(dev, "qbman_init_private_mem() failed 0x%x\n",
 				ret);
@@ -298,6 +301,7 @@
 
 	__bman_probed = 1;
 
+	dev_dbg(dev, "Bman probed successfully [%d]\n", __bman_probed);
 	return 0;
 };
 
@@ -308,10 +312,15 @@
 	{}
 };
 
+static const struct acpi_device_id fsl_bman_acpi_ids[] = {
+	{"NXP0021", 0}
+};
+
 static struct platform_driver fsl_bman_driver = {
 	.driver = {
 		.name = KBUILD_MODNAME,
 		.of_match_table = fsl_bman_ids,
+		.acpi_match_table = ACPI_PTR(fsl_bman_acpi_ids),
 		.suppress_bind_attrs = true,
 	},
 	.probe = fsl_bman_probe,
