--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/soc/fsl/qbman/dpaa_sys.h	2026-04-08 18:20:09.275825871 +0200
+++ ./drivers/soc/fsl/qbman/dpaa_sys.h	2026-04-08 16:21:09.322136501 +0200
@@ -47,6 +47,9 @@
 #include <linux/io.h>
 #include <linux/delay.h>
 
+#define DPAA_QMAN_DEV 0
+#define DPAA_BMAN_DEV 1
+
 /* For 2-element tables related to cache-inhibited and cache-enabled mappings */
 #define DPAA_PORTAL_CE 0
 #define DPAA_PORTAL_CI 1
@@ -102,7 +105,7 @@
 
 /* Initialize the devices private memory region */
 int qbman_init_private_mem(struct device *dev, int idx, dma_addr_t *addr,
-				size_t *size);
+				size_t *size, int dev_id);
 
 /* memremap() attributes for different platforms */
 #ifdef CONFIG_PPC
