--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/soc/fsl/qbman/qman_portal.c	2026-04-08 18:20:09.276001329 +0200
+++ ./drivers/soc/fsl/qbman/qman_portal.c	2026-04-08 16:21:09.322643739 +0200
@@ -1,4 +1,5 @@
 /* Copyright 2008 - 2016 Freescale Semiconductor, Inc.
+ * Copyright 2020 Puresoftware Ltd.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are met:
@@ -28,6 +29,7 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <linux/acpi.h>
 #include "qman_priv.h"
 
 struct qman_portal *qman_dma_portal;
@@ -217,7 +219,11 @@
 		goto err_ioremap1;
 	}
 
-	err = of_property_read_u32(node, "cell-index", &val);
+	if (is_of_node(pdev->dev.fwnode))
+		err = of_property_read_u32(node, "cell-index", &val);
+	else
+		err = device_property_read_u32(&pdev->dev, "cell-index", &val);
+
 	if (err) {
 		dev_err(dev, "Can't get %pOF property 'cell-index'\n", node);
 		__qman_portals_probed = -1;
@@ -290,7 +296,8 @@
 		}
 		qman_done_cleanup();
 	}
-
+	dev_dbg(dev, "Qman : Portal[%d] probed successfully [%d]\n",
+		cpu, __qman_portals_probed);
 	return 0;
 
 err_portal_init:
@@ -311,10 +318,16 @@
 };
 MODULE_DEVICE_TABLE(of, qman_portal_ids);
 
+static const struct acpi_device_id qman_portal_acpi_ids[] = {
+	{"NXP0022", 0}
+};
+MODULE_DEVICE_TABLE(acpi, qman_portal_acpi_ids);
+
 static struct platform_driver qman_portal_driver = {
 	.driver = {
 		.name = KBUILD_MODNAME,
 		.of_match_table = qman_portal_ids,
+		.acpi_match_table = ACPI_PTR(qman_portal_acpi_ids),
 	},
 	.probe = qman_portal_probe,
 };
