--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/staging/media/imx/Makefile	2026-04-08 18:20:09.354981140 +0200
+++ ./drivers/staging/media/imx/Makefile	2026-04-08 16:21:09.465264588 +0200
@@ -8,7 +8,26 @@
 
 imx6-media-csi-objs := imx-media-csi.o imx-media-fim.o
 
+imx8-capture-objs := imx8-isi-core.o
+
+imx8-isi-capture-objs := imx8-isi-fmt.o imx8-isi-cap.o
+imx8-isi-mem2mem-objs := imx8-isi-fmt.o imx8-isi-m2m.o
 obj-$(CONFIG_VIDEO_IMX_MEDIA) += imx-media-common.o
 obj-$(CONFIG_VIDEO_IMX_MEDIA) += imx6-media.o
 obj-$(CONFIG_VIDEO_IMX_MEDIA) += imx6-media-csi.o
 obj-$(CONFIG_VIDEO_IMX_MEDIA) += imx6-mipi-csi2.o
+
+obj-$(CONFIG_IMX8_MIPI_CSI2) 	  += imx8-mipi-csi2.o
+obj-$(CONFIG_IMX8_MIPI_CSI2_SAM)  += imx8-mipi-csi2-sam.o
+obj-$(CONFIG_DWC_MIPI_CSI2_HOST) += dwc-mipi-csi2.o
+
+obj-y += hdmirx/
+
+obj-$(CONFIG_IMX8_CAPTURE_DRIVER) += imx8-capture.o
+obj-$(CONFIG_IMX8_ISI_CORE) += imx8-capture.o
+obj-$(CONFIG_IMX8_ISI_CAPTURE) 	  += imx8-isi-capture.o
+obj-$(CONFIG_IMX8_ISI_M2M)	+= imx8-isi-mem2mem.o
+obj-$(CONFIG_IMX8_ISI_HW) 	  += imx8-isi-hw.o
+obj-$(CONFIG_GMSL_MAX9286)	  += gmsl-max9286.o
+obj-$(CONFIG_IMX8_MEDIA_DEVICE)   += imx8-media-dev.o
+obj-$(CONFIG_IMX8_PARALLEL_CSI)	  += imx8-parallel-csi.o
