--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/tee/optee/smc_abi.c	2026-04-08 18:20:09.447006093 +0200
+++ ./drivers/tee/optee/smc_abi.c	2026-04-08 16:21:09.589492486 +0200
@@ -32,6 +32,9 @@
 #include <linux/kmemleak.h>
 #define CREATE_TRACE_POINTS
 #include "optee_trace.h"
+#if defined(CONFIG_HAVE_IMX_BUSFREQ)
+#include <linux/busfreq-imx.h>
+#endif
 
 /*
  * This file implement the SMC ABI used when communicating with secure world
@@ -927,6 +930,16 @@
 	}
 	/* Initialize waiter */
 	optee_cq_wait_init(&optee->call_queue, &w);
+
+#if defined(CONFIG_HAVE_IMX_BUSFREQ)
+	/*
+	 * Request Busfreq to HIGH to prevent DDR self-refresh while
+	 * executing Secure stuff.
+	 */
+	if (optee->smc.sec_caps & OPTEE_SMC_SEC_CAP_IMX_BUSFREQ)
+		request_bus_freq(BUS_FREQ_HIGH);
+#endif
+
 	while (true) {
 		struct arm_smccc_res res;
 
@@ -956,6 +969,15 @@
 	}
 
 	optee_rpc_finalize_call(&call_ctx);
+
+#if defined(CONFIG_HAVE_IMX_BUSFREQ)
+	/*
+	 * Release Busfreq from HIGH
+	 */
+	if (optee->smc.sec_caps & OPTEE_SMC_SEC_CAP_IMX_BUSFREQ)
+		release_bus_freq(BUS_FREQ_HIGH);
+#endif
+
 	/*
 	 * We're done with our thread in secure world, if there's any
 	 * thread waiters wake up one.
@@ -1698,6 +1720,11 @@
 		goto err_free_pool;
 	}
 
+#if defined(CONFIG_HAVE_IMX_BUSFREQ)
+	if (of_find_compatible_node(NULL, NULL, "fsl,imx_busfreq"))
+		sec_caps |= OPTEE_SMC_SEC_CAP_IMX_BUSFREQ;
+#endif
+
 	optee->ops = &optee_ops;
 	optee->smc.invoke_fn = invoke_fn;
 	optee->smc.sec_caps = sec_caps;
