--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/usb/cdns3/cdns3-imx.c	2026-04-08 18:20:09.503007653 +0200
+++ ./drivers/usb/cdns3/cdns3-imx.c	2026-04-08 16:21:09.634582638 +0200
@@ -360,7 +360,6 @@
 	return 0;
 }
 
-
 /* Indicate if the controller was power lost before */
 static inline bool cdns_imx_is_power_lost(struct cdns_imx *data)
 {
@@ -373,22 +372,18 @@
 		return false;
 }
 
-static int __maybe_unused cdns_imx_system_suspend(struct device *dev)
-{
-	pm_runtime_put_sync(dev);
-	return 0;
-}
-
 static int __maybe_unused cdns_imx_system_resume(struct device *dev)
 {
 	struct cdns_imx *data = dev_get_drvdata(dev);
 	int ret;
 
+	ret = pm_runtime_force_resume(dev);
+	if (ret)
+		return ret;
+
 	ret = pm_runtime_resume_and_get(dev);
-	if (ret < 0) {
-		dev_err(dev, "Could not get runtime PM.\n");
+	if (ret)
 		return ret;
-	}
 
 	if (cdns_imx_is_power_lost(data)) {
 		dev_dbg(dev, "resume from power lost\n");
@@ -397,6 +392,7 @@
 			cdns_imx_suspend(dev);
 	}
 
+	pm_runtime_put_autosuspend(dev);
 	return ret;
 }
 
@@ -411,7 +407,7 @@
 
 static const struct dev_pm_ops cdns_imx_pm_ops = {
 	SET_RUNTIME_PM_OPS(cdns_imx_suspend, cdns_imx_resume, NULL)
-	SET_SYSTEM_SLEEP_PM_OPS(cdns_imx_system_suspend, cdns_imx_system_resume)
+	SET_SYSTEM_SLEEP_PM_OPS(pm_runtime_force_suspend, cdns_imx_system_resume)
 };
 
 static const struct of_device_id cdns_imx_of_match[] = {
