--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/vhost/vhost.c	2026-04-08 18:20:09.617010829 +0200
+++ ./drivers/vhost/vhost.c	2026-04-08 16:21:09.806410945 +0200
@@ -577,6 +577,9 @@
 		vq->heads = NULL;
 		vq->dev = dev;
 		mutex_init(&vq->mutex);
+#ifdef CONFIG_VHOST_XEN
+		INIT_LIST_HEAD(&vq->desc_maps);
+#endif
 		vhost_vq_reset(dev, vq);
 		if (vq->handle_kick)
 			vhost_poll_init(&vq->poll, vq->handle_kick,
@@ -1048,6 +1051,9 @@
 		if (dev->vqs[i]->call_ctx.ctx)
 			eventfd_ctx_put(dev->vqs[i]->call_ctx.ctx);
 		vhost_vq_reset(dev, dev->vqs[i]);
+#ifdef CONFIG_VHOST_XEN
+		vhost_xen_unmap_desc_all(dev->vqs[i]);
+#endif
 	}
 	vhost_dev_free_iovecs(dev);
 	if (dev->log_ctx)
@@ -1812,6 +1818,14 @@
 	for (region = newmem->regions;
 	     region < newmem->regions + mem.nregions;
 	     region++) {
+
+#ifdef CONFIG_VHOST_XEN
+		if (region->guest_phys_addr & XEN_GRANT_DMA_ADDR_OFF) {
+			pr_err("%s: Skip pseudo memory region for Xen grant mappings\n", __func__);
+			continue;
+		}
+#endif
+
 		if (vhost_iotlb_add_range(newumem,
 					  region->guest_phys_addr,
 					  region->guest_phys_addr +
@@ -2407,6 +2421,15 @@
 	u64 s = 0, last = addr + len - 1;
 	int ret = 0;
 
+#ifdef CONFIG_VHOST_XEN
+	iov->iov_len = len;
+	iov->iov_base = vhost_xen_map_desc(vq, addr, len, access);
+	if (IS_ERR(iov->iov_base))
+		return PTR_ERR(iov->iov_base);
+	else
+		return 1;
+#endif
+
 	while ((u64)len > s) {
 		u64 size;
 		if (unlikely(ret >= iov_size)) {
@@ -2486,7 +2509,12 @@
 			vq_err(vq, "Translation failure %d in indirect.\n", ret);
 		return ret;
 	}
+
+#ifdef CONFIG_VHOST_XEN
+	iov_iter_kvec(&from, ITER_SOURCE, (struct kvec *)vq->indirect, ret, len);
+#else
 	iov_iter_init(&from, ITER_SOURCE, vq->indirect, ret, len);
+#endif
 	count = len / sizeof desc;
 	/* Buffers are chained via a 16 bit next field, so
 	 * we can have at most 2^16 of these. */
@@ -2548,6 +2576,12 @@
 			*out_num += ret;
 		}
 	} while ((i = next_desc(vq, &desc)) != -1);
+
+#ifdef CONFIG_VHOST_XEN
+	if (vq->indirect)
+		vhost_xen_unmap_desc(vq, vq->indirect->iov_base, vq->indirect->iov_len);
+#endif
+
 	return 0;
 }
 
