--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/vhost/vhost.h	2026-04-08 18:20:09.618010857 +0200
+++ ./drivers/vhost/vhost.h	2026-04-08 16:21:09.806500996 +0200
@@ -148,6 +148,15 @@
 	bool user_be;
 #endif
 	u32 busyloop_timeout;
+
+#ifdef CONFIG_VHOST_XEN
+	/*
+	 * Contains virtio descriptors mapped by using Xen specific mappings
+	 * in order to handle current request. To be unmapped once the request
+	 * is processed.
+	 */
+	struct list_head desc_maps;
+#endif
 };
 
 struct vhost_msg_node {
@@ -248,6 +257,24 @@
 void vhost_iotlb_map_free(struct vhost_iotlb *iotlb,
 			  struct vhost_iotlb_map *map);
 
+#define XEN_GRANT_DMA_ADDR_OFF    (1ULL << 63)
+
+#ifdef CONFIG_VHOST_XEN
+void vhost_xen_unmap_desc(struct vhost_virtqueue *vq, void *ptr, u32 size);
+void vhost_xen_unmap_desc_all(struct vhost_virtqueue *vq);
+void *vhost_xen_map_desc(struct vhost_virtqueue *vq, u64 addr, u32 size,
+			 int access);
+#else
+static inline void vhost_xen_unmap_desc(struct vhost_virtqueue *vq, void *ptr,
+					u32 size) {}
+static inline void vhost_xen_unmap_desc_all(struct vhost_virtqueue *vq) {}
+static inline void *vhost_xen_map_desc(struct vhost_virtqueue *vq, u64 addr,
+				       u32 size, int access)
+{
+	return ERR_PTR(-ENODEV);
+}
+#endif
+
 #define vq_err(vq, fmt, ...) do {                                  \
 		pr_debug(pr_fmt(fmt), ##__VA_ARGS__);       \
 		if ((vq)->error_ctx)                               \
