--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/vhost/vsock.c	2026-04-08 18:20:09.618010857 +0200
+++ ./drivers/vhost/vsock.c	2026-04-08 16:21:09.806665811 +0200
@@ -86,6 +86,27 @@
 	return NULL;
 }
 
+#ifdef CONFIG_VHOST_XEN
+static void vhost_vsock_unmap_desc(struct vhost_virtqueue *vq, int count)
+{
+	int i;
+
+	for (i = 0; i < count; i++) {
+		if (vq->iov[i].iov_base)
+			vhost_xen_unmap_desc(vq, vq->iov[i].iov_base, vq->iov[i].iov_len);
+	}
+
+	/*
+	 * Alternatively we could unmap *all* mapped at this point descriptors
+	 * (including indirect) in one go instead of unmapping one by one.
+	 * But we must be sure that doing that we won't end up unmapping
+	 * descriptors which are still in use. This depends on the place(s)
+	 * from which current function gets called.
+	 */
+	/*vhost_xen_unmap_desc_all(vq);*/
+}
+#endif
+
 static void
 vhost_transport_do_send_pkt(struct vhost_vsock *vsock,
 			    struct vhost_virtqueue *vq)
@@ -155,7 +176,11 @@
 			break;
 		}
 
+#ifdef CONFIG_VHOST_XEN
+		iov_iter_kvec(&iov_iter, ITER_DEST, (struct kvec *)&vq->iov[out], in, iov_len);
+#else
 		iov_iter_init(&iov_iter, ITER_DEST, &vq->iov[out], in, iov_len);
+#endif
 		payload_len = skb->len;
 		hdr = virtio_vsock_hdr(skb);
 
@@ -209,6 +234,10 @@
 		 */
 		virtio_transport_deliver_tap_pkt(skb);
 
+#ifdef CONFIG_VHOST_XEN
+		/* Descriptors must be unmapped as soon as they are not used */
+		vhost_vsock_unmap_desc(vq, out + in);
+#endif
 		vhost_add_used(vq, head, sizeof(*hdr) + payload_len);
 		added = true;
 
@@ -345,7 +374,11 @@
 	if (!skb)
 		return NULL;
 
+#ifdef CONFIG_VHOST_XEN
+	iov_iter_kvec(&iov_iter, ITER_SOURCE, (struct kvec *)vq->iov, out, len);
+#else
 	iov_iter_init(&iov_iter, ITER_SOURCE, vq->iov, out, len);
+#endif
 
 	hdr = virtio_vsock_hdr(skb);
 	nbytes = copy_from_iter(hdr, sizeof(*hdr), &iov_iter);
@@ -527,6 +560,10 @@
 		else
 			kfree_skb(skb);
 
+#ifdef CONFIG_VHOST_XEN
+		/* Descriptors must be unmapped as soon as they are not used */
+		vhost_vsock_unmap_desc(vq, out + in);
+#endif
 		vhost_add_used(vq, head, 0);
 		added = true;
 	} while(likely(!vhost_exceeds_weight(vq, ++pkts, total_len)));
