--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/video/backlight/pwm_bl.c	2026-04-08 18:20:09.621010941 +0200
+++ ./drivers/video/backlight/pwm_bl.c	2026-04-08 16:21:09.885832364 +0200
@@ -36,6 +36,7 @@
 					int brightness);
 	int			(*check_fb)(struct device *, struct fb_info *);
 	void			(*exit)(struct device *);
+	char			fb_id[16];
 };
 
 static void pwm_backlight_power_on(struct pwm_bl_data *pb)
@@ -226,6 +227,17 @@
 	return 0;
 }
 
+static int pwm_backlight_check_fb_name(struct device *dev, struct fb_info *info)
+{
+	struct backlight_device *bl = dev_get_drvdata(dev);
+	struct pwm_bl_data *pb = bl_get_data(bl);
+
+	if (strcmp(info->fix.id, pb->fb_id) == 0)
+		return true;
+
+	return false;
+}
+
 static int pwm_backlight_parse_dt(struct device *dev,
 				  struct platform_pwm_backlight_data *data)
 {
@@ -237,12 +249,18 @@
 	int length;
 	u32 value;
 	int ret;
+	const char *names;
 
 	if (!node)
 		return -ENODEV;
 
 	memset(data, 0, sizeof(*data));
 
+	if (!of_property_read_string(node, "fb-names", &names)) {
+		strncpy(data->fb_id, names, sizeof(data->fb_id));
+		data->check_fb = &pwm_backlight_check_fb_name;
+	}
+
 	/*
 	 * These values are optional and set as 0 by default, the out values
 	 * are modified only if a valid u32 value can be decoded.
@@ -354,7 +372,6 @@
 
 		data->max_brightness = num_levels - 1;
 	}
-
 	return 0;
 }
 
@@ -489,6 +506,7 @@
 	pb->enabled = false;
 	pb->post_pwm_on_delay = data->post_pwm_on_delay;
 	pb->pwm_off_delay = data->pwm_off_delay;
+	strcpy(pb->fb_id, data->fb_id);
 
 	pb->enable_gpio = devm_gpiod_get_optional(&pdev->dev, "enable",
 						  GPIOD_ASIS);
