--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/video/fbdev/core/fbmem.c	2026-04-08 18:20:09.633011275 +0200
+++ ./drivers/video/fbdev/core/fbmem.c	2026-04-08 16:21:09.813465378 +0200
@@ -48,10 +48,10 @@
 
 DEFINE_MUTEX(registration_lock);
 struct fb_info *registered_fb[FB_MAX] __read_mostly;
+EXPORT_SYMBOL(registered_fb);
+
 int num_registered_fb __read_mostly;
-#define for_each_registered_fb(i)		\
-	for (i = 0; i < FB_MAX; i++)		\
-		if (!registered_fb[i]) {} else
+EXPORT_SYMBOL(num_registered_fb);
 
 bool fb_center_logo __read_mostly;
 
@@ -978,7 +978,7 @@
 	fb_add_videomode(&mode, &fb_info->modelist);
 	registered_fb[i] = fb_info;
 
-#ifdef CONFIG_GUMSTIX_AM200EPD
+#if (defined CONFIG_GUMSTIX_AM200EPD) || (defined CONFIG_FB_MXC_HDMI) || (defined CONFIG_FB_MXS_SII902X)
 	{
 		struct fb_event event;
 		event.info = fb_info;
@@ -1024,7 +1024,7 @@
 	fb_destroy_modelist(&fb_info->modelist);
 	registered_fb[fb_info->node] = NULL;
 	num_registered_fb--;
-#ifdef CONFIG_GUMSTIX_AM200EPD
+#if (defined CONFIG_GUMSTIX_AM200EPD) || (defined CONFIG_FB_MXC_HDMI) || (defined CONFIG_FB_MXS_SII902X)
 	{
 		struct fb_event event;
 		event.info = fb_info;
@@ -1095,13 +1095,25 @@
  */
 void fb_set_suspend(struct fb_info *info, int state)
 {
+#ifdef CONFIG_FB_MXC_HDMI
+	struct fb_event event;
+#endif
 	WARN_CONSOLE_UNLOCKED();
 
+#ifdef CONFIG_FB_MXC_HDMI
+	event.info = info;
+#endif
 	if (state) {
+#ifdef CONFIG_FB_MXC_HDMI
+		fb_notifier_call_chain(FB_EVENT_SUSPEND, &event);
+#endif
 		fbcon_suspended(info);
 		info->state = FBINFO_STATE_SUSPENDED;
 	} else {
 		info->state = FBINFO_STATE_RUNNING;
+#ifdef CONFIG_FB_MXC_HDMI
+		fb_notifier_call_chain(FB_EVENT_RESUME, &event);
+#endif
 		fbcon_resumed(info);
 	}
 }
