--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/drivers/xen/swiotlb-xen.c	2026-04-08 18:20:09.690012863 +0200
+++ ./drivers/xen/swiotlb-xen.c	2026-04-08 16:21:09.930079535 +0200
@@ -383,6 +383,23 @@
 	}
 }
 
+/*swiotlb means disable IOMMU*/
+static dma_addr_t xen_swiotlb_dma_map_resource(struct device *dev, phys_addr_t phys_addr,
+					       size_t size, enum dma_data_direction dir,
+					       unsigned long attrs)
+{
+	dma_addr_t addr = DMA_MAPPING_ERROR;
+
+	addr = dma_direct_map_resource(dev, phys_addr, size, dir, attrs);
+	return addr;
+}
+
+static void xen_swiotlb_dma_unmap_resource(struct device *dev, dma_addr_t addr, size_t size,
+					   enum dma_data_direction dir, unsigned long attrs)
+{
+	/*do nothing*/
+}
+
 /*
  * Return whether the given device DMA address mask can be supported
  * properly.  For example, if your device can only drive the low 24-bits
@@ -417,4 +434,6 @@
 	.alloc_pages = dma_common_alloc_pages,
 	.free_pages = dma_common_free_pages,
 	.max_mapping_size = swiotlb_max_mapping_size,
+	.map_resource = xen_swiotlb_dma_map_resource,
+	.unmap_resource = xen_swiotlb_dma_unmap_resource,
 };
