--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/include/dt-bindings/clock/imx6qdl-clock.h	2026-04-08 18:20:09.991021249 +0200
+++ ./include/dt-bindings/clock/imx6qdl-clock.h	2026-04-08 16:21:11.165933075 +0200
@@ -275,6 +275,11 @@
 #define IMX6QDL_CLK_DCIC2			265
 #define IMX6QDL_CLK_ENET_REF_SEL		266
 #define IMX6QDL_CLK_ENET_REF_PAD		267
-#define IMX6QDL_CLK_END				268
+#define IMX6QDL_CLK_AXI_ALT_SEL			268
+#define IMX6QDL_CLK_LDB_DI0_DIV_7		269
+#define IMX6QDL_CLK_LDB_DI1_DIV_7		270
+#define IMX6QDL_CLK_LDB_DI0_DIV_SEL		271
+#define IMX6QDL_CLK_LDB_DI1_DIV_SEL		272
+#define IMX6QDL_CLK_END				273
 
 #endif /* __DT_BINDINGS_CLOCK_IMX6QDL_H */
