--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/include/dt-bindings/clock/imx93-clock.h	2026-04-08 18:20:09.992021277 +0200
+++ ./include/dt-bindings/clock/imx93-clock.h	2026-04-08 16:21:11.168283157 +0200
@@ -204,6 +204,7 @@
 #define IMX93_CLK_A55_SEL		199
 #define IMX93_CLK_A55_CORE		200
 #define IMX93_CLK_PDM_IPG		201
-#define IMX93_CLK_END			202
+#define IMX93_CLK_SPDIF_IPG		202
+#define IMX93_CLK_END			203
 
 #endif
