--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/include/linux/dma/imx-dma.h	2026-04-08 18:20:10.035859938 +0200
+++ ./include/linux/dma/imx-dma.h	2026-04-08 16:21:11.316579090 +0200
@@ -40,7 +40,8 @@
 	IMX_DMATYPE_ASRC_SP,	/* Shared ASRC */
 	IMX_DMATYPE_SAI,	/* SAI */
 	IMX_DMATYPE_MULTI_SAI,	/* MULTI FIFOs For Audio */
-	IMX_DMATYPE_HDMI,       /* HDMI Audio */
+	IMX_DMATYPE_HDMI,	/* HDMI Audio */
+	IMX_DMATYPE_I2C,	/* I2C */
 };
 
 enum imx_dma_prio {
@@ -61,6 +62,11 @@
 	return !strcmp(dev_name(chan->device->dev), "ipu-core");
 }
 
+static inline int imx_dma_is_pxp(struct dma_chan *chan)
+{
+        return strstr(dev_name(chan->device->dev), "pxp") != NULL;
+}
+
 static inline int imx_dma_is_general_purpose(struct dma_chan *chan)
 {
 	return !strcmp(chan->device->dev->driver->name, "imx-sdma") ||
@@ -82,6 +88,12 @@
  *                  and then jump to FIFO1 for next two words, and so on
  *                  after the last FIFO3 fetched, roll back to FIFO0.
  * @sw_done: Use software done. Needed for PDM (micfil)
+ * @sw_done_sel: software done selector, PDM need enable software done feature
+ *               in mcu_2_sai/sai_2_mcu script.
+ *               Bit31: sw_done eanbled or not
+ *               Bit15~Bit0: selector
+ *               For example: 0x80000000 means sw_done enabled for done0
+ *                            sector which is for PDM on i.mx8mm.
  *
  * Some i.MX Audio devices (SAI, micfil) have multiple successive FIFO
  * registers. For multichannel recording/playback the SAI/micfil have
@@ -97,6 +109,7 @@
 	int stride_fifos_dst;
 	int words_per_fifo;
 	bool sw_done;
+	unsigned int sw_done_sel;
 };
 
 #endif /* __LINUX_DMA_IMX_H */
