--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/include/linux/fb.h	2026-04-08 18:20:10.040022614 +0200
+++ ./include/linux/fb.h	2026-04-08 16:21:11.336092975 +0200
@@ -129,7 +129,19 @@
 /*	The resolution of the passed in fb_info about to change */
 #define FB_EVENT_MODE_CHANGE		0x01
 
-#ifdef CONFIG_GUMSTIX_AM200EPD
+#ifdef CONFIG_FB_MXC_HDMI
+/* only used by mxc_hdmi.c */
+/*     The display on this fb_info is being suspended, no access to the
+ *     framebuffer is allowed any more after that call returns
+ */
+#define FB_EVENT_SUSPEND               0x02
+/*     The display on this fb_info was resumed, you can restore the display
+ *     if you own it
+ */
+#define FB_EVENT_RESUME                        0x03
+#endif
+
+#if (defined CONFIG_GUMSTIX_AM200EPD) || (defined CONFIG_FB_MXC_HDMI) || (defined CONFIG_FB_MXS_SII902X)
 /* only used by mach-pxa/am200epd.c */
 #define FB_EVENT_FB_REGISTERED          0x05
 #define FB_EVENT_FB_UNREGISTERED        0x06
@@ -603,9 +615,15 @@
 extern int fb_get_options(const char *name, char **option);
 extern int fb_new_modelist(struct fb_info *info);
 
+extern struct fb_info *registered_fb[FB_MAX];
+extern int num_registered_fb;
 extern bool fb_center_logo;
 extern int fb_logo_count;
 
+#define for_each_registered_fb(i)		\
+	for (i = 0; i < FB_MAX; i++)		\
+		if (!registered_fb[i]) {} else
+
 static inline void lock_fb_info(struct fb_info *info)
 {
 	mutex_lock(&info->lock);
