--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/include/linux/fsl/enetc_mdio.h	2026-04-08 18:20:10.044695045 +0200
+++ ./include/linux/fsl/enetc_mdio.h	2026-04-08 16:21:11.353171492 +0200
@@ -33,6 +33,7 @@
 struct enetc_mdio_priv {
 	struct enetc_hw *hw;
 	int mdio_base;
+	struct regulator *regulator;
 };
 
 #if IS_REACHABLE(CONFIG_FSL_ENETC_MDIO)
