--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/include/uapi/linux/ethtool.h	2026-04-08 18:20:10.232091683 +0200
+++ ./include/uapi/linux/ethtool.h	2026-04-08 16:21:17.337974138 +0200
@@ -1787,6 +1787,8 @@
 	ETHTOOL_LINK_MODE_10baseT1S_Full_BIT		 = 99,
 	ETHTOOL_LINK_MODE_10baseT1S_Half_BIT		 = 100,
 	ETHTOOL_LINK_MODE_10baseT1S_P2MP_Half_BIT	 = 101,
+	ETHTOOL_LINK_MODE_25000baseCR_S_Full_BIT	 = 102,
+	ETHTOOL_LINK_MODE_25000baseKR_S_Full_BIT	 = 103,
 
 	/* must be last entry */
 	__ETHTOOL_LINK_MODE_MASK_NBITS
