--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/include/uapi/linux/media-bus-format.h	2026-04-08 18:20:10.241028214 +0200
+++ ./include/uapi/linux/media-bus-format.h	2026-04-08 16:21:17.350313622 +0200
@@ -71,6 +71,11 @@
 #define MEDIA_BUS_FMT_RGB888_1X36_CPADLO	0x1021
 #define MEDIA_BUS_FMT_RGB121212_1X36		0x1019
 #define MEDIA_BUS_FMT_RGB161616_1X48		0x101a
+#define MEDIA_BUS_FMT_RGB888_1X30_PADLO		0x101b
+#define MEDIA_BUS_FMT_RGB666_1X30_PADLO		0x101c
+#define MEDIA_BUS_FMT_RGB101010_1X7X5_SPWG	0x101d
+#define MEDIA_BUS_FMT_RGB101010_1X7X5_JEIDA	0x101e
+#define MEDIA_BUS_FMT_RGB565_1X30_PADLO		0x101f
 
 /* YUV (including grey) - next is	0x202f */
 #define MEDIA_BUS_FMT_Y8_1X8			0x2001
