--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/.kernel-meta/configs/standard/features/wifi/realtek-pci.cfg	1970-01-01 01:00:00.000000000 +0100
+++ ./.kernel-meta/configs/standard/features/wifi/realtek-pci.cfg	2026-04-08 16:21:19.371881218 +0200
@@ -0,0 +1,18 @@
+CONFIG_WLAN_VENDOR_REALTEK=y
+
+# rtl818x
+CONFIG_RTL8180=m
+
+# rtl8xxxu
+CONFIG_RTL8XXXU=m
+
+# rtlwifi
+CONFIG_RTL_CARDS=y
+CONFIG_RTL8192CE=m
+CONFIG_RTL8192SE=m
+CONFIG_RTL8192DE=m
+CONFIG_RTL8723AE=m
+CONFIG_RTL8723BE=m
+CONFIG_RTL8188EE=m
+CONFIG_RTL8192EE=m
+CONFIG_RTL8821AE=m
