--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/Makefile	2026-04-08 18:20:04.772875845 +0200
+++ ./Makefile	2026-04-08 18:24:50.671061093 +0200
@@ -390,7 +390,9 @@
 # Alternatively CROSS_COMPILE can be set in the environment.
 # Default value for CROSS_COMPILE is not to prefix executables
 # Note: Some architectures assign CROSS_COMPILE in their arch/*/Makefile
-ARCH		?= $(SUBARCH)
+
+ARCH		?= arm64
+CROSS_COMPILE   ?= aarch64-none-linux-gnu-
 
 # Architecture as present in compile.h
 UTS_MACHINE 	:= $(ARCH)
