--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/sound/soc/codecs/ak4458.c	2026-04-08 18:20:10.817044260 +0200
+++ ./sound/soc/codecs/ak4458.c	2026-04-08 16:21:18.236102965 +0200
@@ -634,13 +634,13 @@
 {
 	if (ak4458->reset_gpiod) {
 		gpiod_set_value_cansleep(ak4458->reset_gpiod, active);
-		usleep_range(1000, 2000);
+		usleep_range(2000, 3000);
 	} else if (!IS_ERR_OR_NULL(ak4458->reset)) {
 		if (active)
 			reset_control_assert(ak4458->reset);
 		else
 			reset_control_deassert(ak4458->reset);
-		usleep_range(1000, 2000);
+		usleep_range(2000, 3000);
 	}
 }
 
@@ -741,6 +741,7 @@
 {
 	struct ak4458_priv *ak4458;
 	int ret, i;
+	int reg;
 
 	ak4458 = devm_kzalloc(&i2c->dev, sizeof(*ak4458), GFP_KERNEL);
 	if (!ak4458)
@@ -794,6 +795,14 @@
 	regcache_cache_only(ak4458->regmap, true);
 	ak4458_reset(ak4458, false);
 
+	/* Check if first register can be read or not */
+	reg = i2c_smbus_read_byte_data(i2c, AK4458_00_CONTROL1);
+	if (reg < 0) {
+		ak4458_reset(ak4458, true);
+		pm_runtime_disable(&i2c->dev);
+		return -ENODEV;
+	}
+
 	return 0;
 }
 
