--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/sound/soc/codecs/pcm512x.c	2026-04-08 18:20:10.846045068 +0200
+++ ./sound/soc/codecs/pcm512x.c	2026-04-08 16:21:18.297260418 +0200
@@ -22,6 +22,7 @@
 
 #include "pcm512x.h"
 
+#define PCM512x_MAX_NUM_SCLK 2
 #define PCM512x_NUM_SUPPLIES 3
 static const char * const pcm512x_supply_names[PCM512x_NUM_SUPPLIES] = {
 	"AVDD",
@@ -31,7 +32,7 @@
 
 struct pcm512x_priv {
 	struct regmap *regmap;
-	struct clk *sclk;
+	struct clk *sclk[PCM512x_MAX_NUM_SCLK];
 	struct regulator_bulk_data supplies[PCM512x_NUM_SUPPLIES];
 	struct notifier_block supply_nb[PCM512x_NUM_SUPPLIES];
 	int fmt;
@@ -48,6 +49,8 @@
 	int mute;
 	struct mutex mutex;
 	unsigned int bclk_ratio;
+	int num_clocks;
+	int sclk_src;
 };
 
 /*
@@ -586,11 +589,12 @@
 	struct device *dev = dai->dev;
 	struct snd_pcm_hw_constraint_ratnums *constraints_no_pll;
 	struct snd_ratnum *rats_no_pll;
+	int ret;
 
-	if (IS_ERR(pcm512x->sclk)) {
+	if (IS_ERR(pcm512x->sclk[0])) {
 		dev_err(dev, "Need SCLK for master mode: %ld\n",
-			PTR_ERR(pcm512x->sclk));
-		return PTR_ERR(pcm512x->sclk);
+			PTR_ERR(pcm512x->sclk[0]));
+		return PTR_ERR(pcm512x->sclk[0]);
 	}
 
 	if (pcm512x->pll_out)
@@ -606,18 +610,39 @@
 	if (!constraints_no_pll)
 		return -ENOMEM;
 	constraints_no_pll->nrats = 1;
+	if (!IS_ERR(pcm512x->sclk[1]))
+		constraints_no_pll->nrats = 2;
+
 	rats_no_pll = devm_kzalloc(dev, sizeof(*rats_no_pll), GFP_KERNEL);
 	if (!rats_no_pll)
 		return -ENOMEM;
 	constraints_no_pll->rats = rats_no_pll;
-	rats_no_pll->num = clk_get_rate(pcm512x->sclk) / 64;
+	rats_no_pll->num = clk_get_rate(pcm512x->sclk[0]) / 64;
 	rats_no_pll->den_min = 1;
 	rats_no_pll->den_max = 128;
 	rats_no_pll->den_step = 1;
 
-	return snd_pcm_hw_constraint_ratnums(substream->runtime, 0,
+	ret = snd_pcm_hw_constraint_ratnums(substream->runtime, 0,
 					     SNDRV_PCM_HW_PARAM_RATE,
 					     constraints_no_pll);
+	if (ret) {
+		dev_err(dev, "Failed to set ratnums constraints\n");
+		return ret;
+	}
+
+	if (!IS_ERR(pcm512x->sclk[1])) {
+		rats_no_pll->num = clk_get_rate(pcm512x->sclk[1]) / 64;
+		ret = snd_pcm_hw_constraint_ratnums(substream->runtime, 0,
+					     SNDRV_PCM_HW_PARAM_RATE,
+					     constraints_no_pll);
+		if (ret) {
+			dev_err(dev, "Failed to set ratnums constraints\n");
+			return ret;
+		}
+	}
+
+	return 0;
+
 }
 
 static int pcm512x_dai_startup_slave(struct snd_pcm_substream *substream,
@@ -628,9 +653,9 @@
 	struct device *dev = dai->dev;
 	struct regmap *regmap = pcm512x->regmap;
 
-	if (IS_ERR(pcm512x->sclk)) {
+	if (IS_ERR(pcm512x->sclk[pcm512x->sclk_src])) {
 		dev_info(dev, "No SCLK, using BCLK: %ld\n",
-			 PTR_ERR(pcm512x->sclk));
+			 PTR_ERR(pcm512x->sclk[pcm512x->sclk_src]));
 
 		/* Disable reporting of missing SCLK as an error */
 		regmap_update_bits(regmap, PCM512x_ERROR_DETECT,
@@ -922,7 +947,7 @@
 	}
 
 	if (!pcm512x->pll_out) {
-		sck_rate = clk_get_rate(pcm512x->sclk);
+		sck_rate = clk_get_rate(pcm512x->sclk[pcm512x->sclk_src]);
 		bclk_rate = params_rate(params) * lrclk_div;
 		bclk_div = DIV_ROUND_CLOSEST(sck_rate, bclk_rate);
 
@@ -939,7 +964,7 @@
 		}
 		bclk_rate = ret;
 
-		pllin_rate = clk_get_rate(pcm512x->sclk);
+		pllin_rate = clk_get_rate(pcm512x->sclk[pcm512x->sclk_src]);
 
 		sck_rate = pcm512x_find_sck(dai, bclk_rate);
 		if (!sck_rate)
@@ -1430,6 +1455,29 @@
 	return 0;
 }
 
+static int pcm512x_set_sysclk(struct snd_soc_dai *dai, int clk_id,
+			      unsigned int freq, int dir)
+{
+	struct snd_soc_component *component = dai->component;
+	struct pcm512x_priv *pcm512x = snd_soc_component_get_drvdata(component);
+
+	if (dir == SND_SOC_CLOCK_IN) {
+		switch (clk_id) {
+		case PCM512x_SYSCLK_MCLK1:
+		case PCM512x_SYSCLK_MCLK2:
+			pcm512x->sclk_src = clk_id;
+			break;
+		default:
+			return -EINVAL;
+		}
+
+		dev_dbg(component->dev,
+			"SCLK dir: %d; sclk_src: %d\n", dir, clk_id);
+	}
+
+	return 0;
+}
+
 static int pcm512x_mute(struct snd_soc_dai *dai, int mute, int direction)
 {
 	struct snd_soc_component *component = dai->component;
@@ -1483,6 +1531,7 @@
 	.set_fmt = pcm512x_set_fmt,
 	.mute_stream = pcm512x_mute,
 	.set_bclk_ratio = pcm512x_set_bclk_ratio,
+	.set_sysclk = pcm512x_set_sysclk,
 	.no_capture_mute = 1,
 };
 
@@ -1542,6 +1591,7 @@
 int pcm512x_probe(struct device *dev, struct regmap *regmap)
 {
 	struct pcm512x_priv *pcm512x;
+	char clk_name[8];
 	int i, ret;
 
 	pcm512x = devm_kzalloc(dev, sizeof(struct pcm512x_priv), GFP_KERNEL);
@@ -1599,13 +1649,18 @@
 		goto err;
 	}
 
-	pcm512x->sclk = devm_clk_get(dev, NULL);
-	if (PTR_ERR(pcm512x->sclk) == -EPROBE_DEFER) {
+	/* default to first sclk */
+	pcm512x->num_clocks = 1;
+	pcm512x->sclk_src = PCM512x_SYSCLK_MCLK1;
+
+	pcm512x->sclk[0] = devm_clk_get(dev, NULL);
+	if (PTR_ERR(pcm512x->sclk[0]) == -EPROBE_DEFER) {
 		ret = -EPROBE_DEFER;
 		goto err;
 	}
-	if (!IS_ERR(pcm512x->sclk)) {
-		ret = clk_prepare_enable(pcm512x->sclk);
+
+	if (!IS_ERR(pcm512x->sclk[0])) {
+		ret = clk_prepare_enable(pcm512x->sclk[0]);
 		if (ret != 0) {
 			dev_err(dev, "Failed to enable SCLK: %d\n", ret);
 			goto err;
@@ -1630,6 +1685,26 @@
 		const struct device_node *np = dev->of_node;
 		u32 val;
 
+		if (of_property_read_bool(np, "clocks")) {
+			pcm512x->num_clocks =
+				of_property_count_u32_elems(np, "clocks");
+			if (pcm512x->num_clocks > PCM512x_MAX_NUM_SCLK) {
+				dev_err(dev, "Failed unsupported max sclk: %d\n",
+					pcm512x->num_clocks);
+				goto err;
+			}
+
+			for (i = 0; i < pcm512x->num_clocks; i++) {
+				sprintf(clk_name, "sclk%d", i);
+				pcm512x->sclk[i] = devm_clk_get(dev, clk_name);
+
+				if (IS_ERR(pcm512x->sclk[i])) {
+					dev_info(dev, "Failed to get sclk%d\n", i);
+					pcm512x->sclk[i] = NULL;
+				}
+			}
+		}
+
 		if (of_property_read_u32(np, "pll-in", &val) >= 0) {
 			if (val > 6) {
 				dev_err(dev, "Invalid pll-in\n");
@@ -1674,8 +1749,10 @@
 err_pm:
 	pm_runtime_disable(dev);
 err_clk:
-	if (!IS_ERR(pcm512x->sclk))
-		clk_disable_unprepare(pcm512x->sclk);
+	for (i = 0; i < pcm512x->num_clocks; i++) {
+		if (!IS_ERR(pcm512x->sclk[i]))
+			clk_disable_unprepare(pcm512x->sclk[i]);
+	}
 err:
 	regulator_bulk_disable(ARRAY_SIZE(pcm512x->supplies),
 				     pcm512x->supplies);
@@ -1686,10 +1763,13 @@
 void pcm512x_remove(struct device *dev)
 {
 	struct pcm512x_priv *pcm512x = dev_get_drvdata(dev);
+	int i;
 
 	pm_runtime_disable(dev);
-	if (!IS_ERR(pcm512x->sclk))
-		clk_disable_unprepare(pcm512x->sclk);
+	for (i = 0; i < pcm512x->num_clocks; i++) {
+		if (!IS_ERR(pcm512x->sclk[i]))
+			clk_disable_unprepare(pcm512x->sclk[i]);
+	}
 	regulator_bulk_disable(ARRAY_SIZE(pcm512x->supplies),
 			       pcm512x->supplies);
 }
@@ -1699,7 +1779,7 @@
 static int pcm512x_suspend(struct device *dev)
 {
 	struct pcm512x_priv *pcm512x = dev_get_drvdata(dev);
-	int ret;
+	int ret, i;
 
 	ret = regmap_update_bits(pcm512x->regmap, PCM512x_POWER,
 				 PCM512x_RQPD, PCM512x_RQPD);
@@ -1715,8 +1795,10 @@
 		return ret;
 	}
 
-	if (!IS_ERR(pcm512x->sclk))
-		clk_disable_unprepare(pcm512x->sclk);
+	for (i = 0; i < pcm512x->num_clocks; i++) {
+		if (!IS_ERR(pcm512x->sclk[i]))
+			clk_disable_unprepare(pcm512x->sclk[i]);
+	}
 
 	return 0;
 }
@@ -1724,13 +1806,15 @@
 static int pcm512x_resume(struct device *dev)
 {
 	struct pcm512x_priv *pcm512x = dev_get_drvdata(dev);
-	int ret;
+	int ret, i;
 
-	if (!IS_ERR(pcm512x->sclk)) {
-		ret = clk_prepare_enable(pcm512x->sclk);
-		if (ret != 0) {
-			dev_err(dev, "Failed to enable SCLK: %d\n", ret);
-			return ret;
+	for (i = 0; i < pcm512x->num_clocks; i++) {
+		if (!IS_ERR(pcm512x->sclk[i])) {
+			ret = clk_prepare_enable(pcm512x->sclk[i]);
+			if (ret != 0) {
+				dev_err(dev, "Failed to enable SCLK: %d\n", ret);
+				return ret;
+			}
 		}
 	}
 
