--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/sound/soc/fsl/imx-audio-rpmsg.c	2026-04-08 18:20:10.901046600 +0200
+++ ./sound/soc/fsl/imx-audio-rpmsg.c	2026-04-08 16:21:18.386289457 +0200
@@ -12,6 +12,7 @@
  */
 struct imx_audio_rpmsg {
 	struct platform_device *rpmsg_pdev;
+	struct platform_device *card_pdev;
 };
 
 static int imx_audio_rpmsg_cb(struct rpmsg_device *rpdev, void *data, int len,
@@ -74,6 +75,10 @@
 static int imx_audio_rpmsg_probe(struct rpmsg_device *rpdev)
 {
 	struct imx_audio_rpmsg *data;
+	struct platform_device *codec_pdev;
+	struct rpmsg_codec codec;
+	const char *model_string;
+	struct device_node *np;
 	int ret = 0;
 
 	dev_info(&rpdev->dev, "new channel: 0x%x -> 0x%x!\n",
@@ -87,14 +92,57 @@
 
 	/* Register platform driver for rpmsg routine */
 	data->rpmsg_pdev = platform_device_register_data(&rpdev->dev,
-							 IMX_PCM_DRV_NAME,
-							 PLATFORM_DEVID_AUTO,
+							 rpdev->id.name,
+							 PLATFORM_DEVID_NONE,
 							 NULL, 0);
 	if (IS_ERR(data->rpmsg_pdev)) {
 		dev_err(&rpdev->dev, "failed to register rpmsg platform.\n");
 		ret = PTR_ERR(data->rpmsg_pdev);
 	}
 
+	data->card_pdev = platform_device_register_data(&rpdev->dev,
+							"imx-audio-rpmsg",
+							PLATFORM_DEVID_AUTO,
+							rpdev->id.name,
+							strlen(rpdev->id.name) + 1);
+	if (IS_ERR(data->card_pdev)) {
+		dev_err(&rpdev->dev, "failed to register rpmsg card.\n");
+		ret = PTR_ERR(data->card_pdev);
+	}
+
+	if (!strcmp(rpdev->id.name, "rpmsg-micfil-channel"))
+		return ret;
+	np = of_find_node_by_name(NULL, "rpmsg_audio");
+	of_property_read_string(np, "model", &model_string);
+	if (np && of_device_is_compatible(np, "fsl,imx7ulp-rpmsg-audio")) {
+		codec.audioindex = 0;
+		codec.shared_lrclk = true;
+		codec.capless = false;
+		codec_pdev = platform_device_register_data(&data->rpmsg_pdev->dev,
+							   RPMSG_CODEC_DRV_NAME_WM8960,
+							   PLATFORM_DEVID_NONE,
+							   &codec,
+							   sizeof(struct rpmsg_codec));
+		if (IS_ERR(codec_pdev)) {
+			dev_err(&rpdev->dev, "failed to register rpmsg codec\n");
+			ret = PTR_ERR(codec_pdev);
+			goto fail;
+		}
+	} else if (np && of_device_is_compatible(np, "fsl,imx8mm-rpmsg-audio") &&
+			!strcmp("ak4497-audio", model_string)) {
+		codec.audioindex = 0;
+		codec_pdev = platform_device_register_data(&data->rpmsg_pdev->dev,
+							   RPMSG_CODEC_DRV_NAME_AK4497,
+							   PLATFORM_DEVID_NONE,
+							   &codec,
+							   sizeof(struct rpmsg_codec));
+		if (IS_ERR(codec_pdev)) {
+			dev_err(&rpdev->dev, "failed to register rpmsg codec\n");
+			ret = PTR_ERR(codec_pdev);
+			goto fail;
+		}
+	}
+fail:
 	return ret;
 }
 
@@ -105,6 +153,9 @@
 	if (data->rpmsg_pdev)
 		platform_device_unregister(data->rpmsg_pdev);
 
+	if (data->card_pdev)
+		platform_device_unregister(data->card_pdev);
+
 	dev_info(&rpdev->dev, "audio rpmsg driver is removed\n");
 }
 
@@ -113,6 +164,7 @@
 	{ .name = "rpmsg-micfil-channel" },
 	{ },
 };
+MODULE_DEVICE_TABLE(rpmsg, imx_audio_rpmsg_id_table);
 
 static struct rpmsg_driver imx_audio_rpmsg_driver = {
 	.drv.name	= "imx_audio_rpmsg",
@@ -126,5 +178,5 @@
 
 MODULE_DESCRIPTION("Freescale SoC Audio RPMSG interface");
 MODULE_AUTHOR("Shengjiu Wang <shengjiu.wang@nxp.com>");
-MODULE_ALIAS("platform:imx_audio_rpmsg");
+MODULE_ALIAS("rpmsg:imx_audio_rpmsg");
 MODULE_LICENSE("GPL v2");
