--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/sound/soc/fsl/imx-hdmi.c	2026-04-08 18:20:10.902046628 +0200
+++ ./sound/soc/fsl/imx-hdmi.c	2026-04-08 16:21:18.387280757 +0200
@@ -8,6 +8,8 @@
 #include <sound/hdmi-codec.h>
 #include "fsl_sai.h"
 
+#define SUPPORT_RATE_NUM 10
+
 /**
  * struct cpu_priv - CPU private data
  * @sysclk_id: SYSCLK ids for set_sysclk()
@@ -29,6 +31,55 @@
 	u32 dai_fmt;
 };
 
+static int imx_hdmi_startup(struct snd_pcm_substream *substream)
+{
+	struct snd_pcm_runtime *runtime = substream->runtime;
+	int ret;
+
+	/* Remove S20_3LE for clock issue */
+	ret = snd_pcm_hw_constraint_mask64(runtime,
+					   SNDRV_PCM_HW_PARAM_FORMAT,
+					   ~SNDRV_PCM_FMTBIT_S20_3LE);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static int imx_sii902x_startup(struct snd_pcm_substream *substream)
+{
+	struct snd_pcm_runtime *runtime = substream->runtime;
+	static struct snd_pcm_hw_constraint_list constraint_rates;
+	static u32 support_rates[SUPPORT_RATE_NUM];
+	int ret;
+
+	support_rates[0] = 32000;
+	support_rates[1] = 48000;
+	support_rates[2] = 96000;
+	support_rates[3] = 192000;
+	constraint_rates.list = support_rates;
+	constraint_rates.count = 4;
+
+	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_RATE,
+					 &constraint_rates);
+	if (ret)
+		return ret;
+
+	ret = snd_pcm_hw_constraint_minmax(runtime, SNDRV_PCM_HW_PARAM_CHANNELS,
+					   1, 2);
+	if (ret < 0)
+		return ret;
+
+	/* Remove S20_3LE for clock issue */
+	ret = snd_pcm_hw_constraint_mask64(runtime,
+					   SNDRV_PCM_HW_PARAM_FORMAT,
+					   ~SNDRV_PCM_FMTBIT_S20_3LE);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
 static int imx_hdmi_hw_params(struct snd_pcm_substream *substream,
 			      struct snd_pcm_hw_params *params)
 {
@@ -60,6 +111,12 @@
 }
 
 static const struct snd_soc_ops imx_hdmi_ops = {
+	.startup = imx_hdmi_startup,
+	.hw_params = imx_hdmi_hw_params,
+};
+
+static struct snd_soc_ops imx_sii902x_ops = {
+	.startup = imx_sii902x_startup,
 	.hw_params = imx_hdmi_hw_params,
 };
 
@@ -157,6 +214,7 @@
 	if (of_device_is_compatible(np, "fsl,imx-audio-sii902x")) {
 		data->dai_fmt = SND_SOC_DAIFMT_LEFT_J;
 		data->cpu_priv.slot_width = 24;
+		data->dai.ops = &imx_sii902x_ops;
 	} else {
 		data->dai_fmt = SND_SOC_DAIFMT_I2S;
 		data->cpu_priv.slot_width = 32;
