--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/sound/soc/fsl/imx-pcm-rpmsg.h	2026-04-08 18:20:10.902046628 +0200
+++ ./sound/soc/fsl/imx-pcm-rpmsg.h	2026-04-08 16:21:18.388518722 +0200
@@ -507,6 +507,19 @@
 	struct stream_timer      stream_timer[2];
 };
 
+struct rpmsg_codec {
+        int audioindex;
+
+        /*property for wm8960*/
+        bool capless;
+        bool shared_lrclk;
+};
+
+#define RPMSG_CODEC_WM8960 1
+#define RPMSG_CODEC_AK4497 2
+
+#define RPMSG_CODEC_DRV_NAME_WM8960 "rpmsg-codec-wm8960"
+#define RPMSG_CODEC_DRV_NAME_AK4497 "rpmsg-codec-ak4497"
 #define IMX_PCM_DRV_NAME "imx_pcm_rpmsg"
 
 #endif /* IMX_PCM_RPMSG_H */
