--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/sound/soc/fsl/imx-spdif.c	2026-04-08 18:20:10.902046628 +0200
+++ ./sound/soc/fsl/imx-spdif.c	2026-04-08 16:21:18.389546812 +0200
@@ -6,9 +6,43 @@
 #include <linux/of_platform.h>
 #include <sound/soc.h>
 
+#define SUPPORT_RATE_NUM 10
+
 struct imx_spdif_data {
 	struct snd_soc_dai_link dai;
 	struct snd_soc_card card;
+	u32 support_rates[SUPPORT_RATE_NUM];
+	u32 support_rates_num;
+};
+
+#define CLK_8K_FREQ    24576000
+#define CLK_11K_FREQ   22579200
+
+static int imx_spdif_startup(struct snd_pcm_substream *substream)
+{
+	struct snd_pcm_runtime *runtime = substream->runtime;
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_card *card = rtd->card;
+	struct imx_spdif_data *data = snd_soc_card_get_drvdata(card);
+	static struct snd_pcm_hw_constraint_list constraint_rates;
+	int ret;
+
+	if (!data->support_rates_num)
+		return 0;
+
+	constraint_rates.list = data->support_rates;
+	constraint_rates.count = data->support_rates_num;
+
+	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_RATE,
+					 &constraint_rates);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static struct snd_soc_ops imx_spdif_ops = {
+	.startup = imx_spdif_startup,
 };
 
 static int imx_spdif_audio_probe(struct platform_device *pdev)
@@ -16,7 +50,7 @@
 	struct device_node *spdif_np, *np = pdev->dev.of_node;
 	struct imx_spdif_data *data;
 	struct snd_soc_dai_link_component *comp;
-	int ret = 0;
+	int ret = 0, i;
 
 	spdif_np = of_parse_phandle(np, "spdif-controller", 0);
 	if (!spdif_np) {
@@ -49,6 +83,7 @@
 	data->dai.cpus->of_node = spdif_np;
 	data->dai.playback_only = true;
 	data->dai.capture_only = true;
+	data->dai.ops = &imx_spdif_ops;
 
 	if (of_property_read_bool(np, "spdif-out"))
 		data->dai.capture_only = false;
@@ -61,6 +96,16 @@
 		goto end;
 	}
 
+	for (i = 0; i < SUPPORT_RATE_NUM; i++) {
+		ret = of_property_read_u32_index(pdev->dev.of_node,
+						 "fsl,constraint-rate",
+						 i, &data->support_rates[i]);
+		if (!ret)
+			data->support_rates_num = i + 1;
+		else
+			break;
+	}
+
 	data->card.dev = &pdev->dev;
 	data->card.dai_link = &data->dai;
 	data->card.num_links = 1;
@@ -70,6 +115,7 @@
 	if (ret)
 		goto end;
 
+	snd_soc_card_set_drvdata(&data->card, data);
 	ret = devm_snd_soc_register_card(&pdev->dev, &data->card);
 	if (ret)
 		dev_err_probe(&pdev->dev, ret, "snd_soc_register_card failed\n");
