--- /home/rossierd/edgemtech/products/edgem1/edgem1/30/build/tmp/work/linux-6.6-verdin-r0/linux-6.6-verdin/tools/virtio/Makefile	2026-04-08 18:20:11.263126816 +0200
+++ ./tools/virtio/Makefile	2026-04-08 16:21:19.358642789 +0200
@@ -1,8 +1,10 @@
 # SPDX-License-Identifier: GPL-2.0
-all: test mod
+all: test mod virtio-ivshmem-block virtio-ivshmem-console
 test: virtio_test vringh_test
 virtio_test: virtio_ring.o virtio_test.o
 vringh_test: vringh_test.o vringh.o virtio_ring.o
+virtio-ivshmem-block: virtio-ivshmem-block.o
+virtio-ivshmem-console: virtio-ivshmem-console.o
 
 try-run = $(shell set -e;		\
 	if ($(1)) >/dev/null 2>&1;	\
@@ -50,5 +52,6 @@
 .PHONY: all test mod clean vhost oot oot-clean oot-build
 clean:
 	${RM} *.o vringh_test virtio_test vhost_test/*.o vhost_test/.*.cmd \
-              vhost_test/Module.symvers vhost_test/modules.order *.d
+              vhost_test/Module.symvers vhost_test/modules.order *.d \
+              virtio-ivshmem-block virtio-ivshmem-console
 -include *.d
